
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../study_index/">
      
      
        <link rel="next" href="../study_locus_overlap/">
      
      
      <link rel="icon" href="../../../assets/imgs/favicon.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.26">
    
    
      
        <title>Study Locus - Open Targets Gentropy</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/timeago.css">
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../../../assets/stylesheets/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#gentropy.dataset.study_locus.StudyLocus" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Open Targets Gentropy" class="md-header__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Open Targets Gentropy
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Study Locus
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="blue-grey" data-md-color-accent="light-blue"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
    
  
  Open Targets Gentropy

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../howto/_howto/" class="md-tabs__link">
        
  
    
  
  How-to

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../roadmap/" class="md-tabs__link">
        
  
    
  
  Roadmap

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../development/_development/" class="md-tabs__link">
        
  
    
  
  Development

      </a>
    </li>
  

      
        
  
  
    
  
  
    <li class="md-tabs__item md-tabs__item--active">
      <a href="../../_python_api/" class="md-tabs__link">
        
  
    
  
  Python API

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Open Targets Gentropy" class="md-nav__button md-logo" aria-label="Open Targets Gentropy" data-md-component="logo">
      
  <img src="../../../assets/imgs/otlogo.png" alt="logo">

    </a>
    Open Targets Gentropy
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/opentargets/gentropy" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    opentargets/gentropy
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Open Targets Gentropy
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/_howto/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    How-to
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            How-to
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/command_line/_command_line/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Command line
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_1" id="__nav_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_1">
            <span class="md-nav__icon md-icon"></span>
            Command line
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_in_cli/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step in CLI
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/command_line/run_step_using_config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Run step using YAML config
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../howto/python_api/_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python api
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Python api
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/a_creating_spark_session/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Creating a Spark Session
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/b_create_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Create a dataset
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/c_applying_methods/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Applying methods
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../howto/python_api/d_inspect_dataset/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Inspect a dataset
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../roadmap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Roadmap
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../development/_development/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Development
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Development
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/airflow/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Airflow configuration
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/contributing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Contributing guidelines
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/troubleshooting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Troubleshooting
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../development/workflows/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pipeline workflows
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../_python_api/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Python API
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6" id="__nav_6_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            Python API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../common/_common/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Common
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_1" id="__nav_6_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_1">
            <span class="md-nav__icon md-icon"></span>
            Common
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/types/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Literal Types
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../common/version_engine/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    VersionEngine
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_2" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../_datasets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_2" id="__nav_6_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_6_2">
            <span class="md-nav__icon md-icon"></span>
            Datasets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../gene_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Gene Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../intervals/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_feature/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_feature_matrix/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../l2g_prediction/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Prediction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Study Locus
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocus
    </span>
  </a>
  
    <nav class="md-nav" aria-label="StudyLocus">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_credible_sets
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_ld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_locus_statistics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="md-nav__link">
    <span class="md-ellipsis">
      assign_study_locus_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_credible_set_log10bf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.clump" class="md-nav__link">
    <span class="md-ellipsis">
      clump
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_by_study_type" class="md-nav__link">
    <span class="md-ellipsis">
      filter_by_study_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="md-nav__link">
    <span class="md-ellipsis">
      filter_credible_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="md-nav__link">
    <span class="md-ellipsis">
      find_overlaps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="md-nav__link">
    <span class="md-ellipsis">
      neglog_pvalue
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="md-nav__link">
    <span class="md-ellipsis">
      unique_variants_in_locus
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.update_quality_flag" class="md-nav__link">
    <span class="md-ellipsis">
      update_quality_flag
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocusQualityCheck" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocusQualityCheck
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.CredibleInterval" class="md-nav__link">
    <span class="md-ellipsis">
      CredibleInterval
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      Schema
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../study_locus_overlap/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Locus Overlap
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_annotation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../variant_to_gene/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variant-to-gene
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/_datasources/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Datasources
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3" id="__nav_6_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3">
            <span class="md-nav__icon md-icon"></span>
            Datasources
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/eqtl_catalogue/_eqtl_catalogue/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Eqtl catalogue
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_1" id="__nav_6_3_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_1">
            <span class="md-nav__icon md-icon"></span>
            Eqtl catalogue
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fine mapping results
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/eqtl_catalogue/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Stats
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_2" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/finngen/_finngen/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Finngen
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_2" id="__nav_6_3_2_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_2">
            <span class="md-nav__icon md-icon"></span>
            Finngen
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/finemapping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Finemapping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/finngen/summary_stats/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gnomad/_gnomad/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gnomad
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_3" id="__nav_6_3_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_3">
            <span class="md-nav__icon md-icon"></span>
            Gnomad
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_ld/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LD Matrix
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gnomad/gnomad_variants/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Variants
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/gwas_catalog/_gwas_catalog/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Gwas catalog
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_4" id="__nav_6_3_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_4">
            <span class="md-nav__icon md-icon"></span>
            Gwas catalog
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/associations/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Associations
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/study_splitter/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Study Splitter
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/gwas_catalog/summary_statistics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/intervals/_intervals/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Intervals
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_5" id="__nav_6_3_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_5">
            <span class="md-nav__icon md-icon"></span>
            Intervals
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/andersson/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Andersson et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/javierre/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Javierre et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/jung/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Jung et al.
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/intervals/thurman/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Thurman et al.
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_3_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../datasources/open_targets/_open_targets/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Open targets
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_3_6" id="__nav_6_3_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_3_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_3_6">
            <span class="md-nav__icon md-icon"></span>
            Open targets
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/open_targets/l2g_gold_standard/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Gold Standard
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../datasources/open_targets/target/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Target
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../methods/_methods/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Methods
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_4" id="__nav_6_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_4">
            <span class="md-nav__icon md-icon"></span>
            Methods
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/carma/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CARMA
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/coloc/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Coloc
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/ecaviar/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eCAVIAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/ld_annotator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    LDAnnotator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    PICS
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/sumstat_imputation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Summary Statistics Imputation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/sumstat_quality_controls/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    QC of GWAS Summary Statistics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/susie_inf/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SuSiE-inf
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_4_10" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../methods/l2g/_l2g/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    L2g
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_4_10" id="__nav_6_4_10_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_6_4_10_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_4_10">
            <span class="md-nav__icon md-icon"></span>
            L2g
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/evaluator/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    W&B evaluator
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/feature_factory/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Feature Factory
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/model/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Model
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../methods/l2g/trainer/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    L2G Trainer
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_6_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../steps/_steps/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Steps
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6_5" id="__nav_6_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_6_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6_5">
            <span class="md-nav__icon md-icon"></span>
            Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/colocalisation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    colocalisation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/eqtl_catalogue/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    eQTL Catalogue
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_studies/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_studies
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/finngen_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    finngen_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gene_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gene_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_curation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_curation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_inclusion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_study_inclusion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_ingestion/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_ingestion
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/gwas_catalog_sumstat_preprocess/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    gwas_catalog_sumstat_preprocess
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/l2g/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    locus_to_gene
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_clump/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/ld_index/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    ld_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/pics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    pics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_annotation_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_annotation
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_index_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_index
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/variant_to_gene_step/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    variant_to_gene
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../steps/window_based_clumping/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    window_based_clumping
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocus
    </span>
  </a>
  
    <nav class="md-nav" aria-label="StudyLocus">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_credible_sets
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_ld
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="md-nav__link">
    <span class="md-ellipsis">
      annotate_locus_statistics
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="md-nav__link">
    <span class="md-ellipsis">
      assign_study_locus_id
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_credible_set_log10bf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.clump" class="md-nav__link">
    <span class="md-ellipsis">
      clump
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_by_study_type" class="md-nav__link">
    <span class="md-ellipsis">
      filter_by_study_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="md-nav__link">
    <span class="md-ellipsis">
      filter_credible_set
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="md-nav__link">
    <span class="md-ellipsis">
      find_overlaps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.get_schema" class="md-nav__link">
    <span class="md-ellipsis">
      get_schema
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="md-nav__link">
    <span class="md-ellipsis">
      neglog_pvalue
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="md-nav__link">
    <span class="md-ellipsis">
      unique_variants_in_locus
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocus.update_quality_flag" class="md-nav__link">
    <span class="md-ellipsis">
      update_quality_flag
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.StudyLocusQualityCheck" class="md-nav__link">
    <span class="md-ellipsis">
      StudyLocusQualityCheck
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#gentropy.dataset.study_locus.CredibleInterval" class="md-nav__link">
    <span class="md-ellipsis">
      CredibleInterval
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#schema" class="md-nav__link">
    <span class="md-ellipsis">
      Schema
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


  <h1>Study Locus</h1>

<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.StudyLocus" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.StudyLocus</code>


  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-dataclass"><code>dataclass</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="gentropy.dataset.dataset.Dataset" href="../_datasets/#gentropy.dataset.dataset.Dataset">Dataset</a></code></p>


      <p>Study-Locus dataset.</p>
<p>This dataset captures associations between study/traits and a genetic loci as provided by finemapping methods.</p>

              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@dataclass</span>
<span class="k">class</span> <span class="nc">StudyLocus</span><span class="p">(</span><span class="n">Dataset</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Study-Locus dataset.</span>

<span class="sd">    This dataset captures associations between study/traits and a genetic loci as provided by finemapping methods.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_overlapping_peaks</span><span class="p">(</span>
        <span class="n">credset_to_overlap</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate overlapping signals (study-locus) between GWAS-GWAS and GWAS-Molecular trait.</span>

<span class="sd">        Args:</span>
<span class="sd">            credset_to_overlap (DataFrame): DataFrame containing at least `studyLocusId`, `studyType`, `chromosome` and `tagVariantId` columns.</span>
<span class="sd">            intra_study_overlap (bool): When True, finds intra-study overlaps for credible set deduplication. Default is False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: containing `leftStudyLocusId`, `rightStudyLocusId` and `chromosome` columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Reduce columns to the minimum to reduce the size of the dataframe</span>
        <span class="n">credset_to_overlap</span> <span class="o">=</span> <span class="n">credset_to_overlap</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;region&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tagVariantId&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="c1"># Define join condition - if intra_study_overlap is True, finds overlaps within the same study. Otherwise finds gwas vs everything overlaps for coloc.</span>
        <span class="n">join_condition</span> <span class="o">=</span> <span class="p">(</span>
            <span class="p">[</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.tagVariantId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.tagVariantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyLocusId&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyLocusId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.region&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.region&quot;</span><span class="p">),</span>
            <span class="p">]</span>
            <span class="k">if</span> <span class="n">intra_study_overlap</span>
            <span class="k">else</span> <span class="p">[</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.chromosome&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.tagVariantId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.tagVariantId&quot;</span><span class="p">),</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyType&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="s2">&quot;gwas&quot;</span><span class="p">)</span>
                <span class="o">|</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyLocusId&quot;</span><span class="p">)</span> <span class="o">&gt;</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyLocusId&quot;</span><span class="p">)),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="s2">&quot;gwas&quot;</span><span class="p">),</span>
            <span class="p">]</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="p">(</span>
            <span class="n">credset_to_overlap</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;left&quot;</span><span class="p">)</span>
            <span class="c1"># Self join with complex condition.</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">credset_to_overlap</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;right&quot;</span><span class="p">),</span>
                <span class="n">on</span><span class="o">=</span><span class="n">join_condition</span><span class="p">,</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;right.studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;left.chromosome&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
            <span class="o">.</span><span class="n">repartition</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">_align_overlapping_tags</span><span class="p">(</span>
        <span class="n">loci_to_overlap</span><span class="p">:</span> <span class="n">DataFrame</span><span class="p">,</span> <span class="n">peak_overlaps</span><span class="p">:</span> <span class="n">DataFrame</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Align overlapping tags in pairs of overlapping study-locus, keeping all tags in both loci.</span>

<span class="sd">        Args:</span>
<span class="sd">            loci_to_overlap (DataFrame): containing `studyLocusId`, `studyType`, `chromosome`, `tagVariantId`, `logBF` and `posteriorProbability` columns.</span>
<span class="sd">            peak_overlaps (DataFrame): containing `leftStudyLocusId`, `rightStudyLocusId` and `chromosome` columns.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocusOverlap: Pairs of overlapping study-locus with aligned tags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Complete information about all tags in the left study-locus of the overlap</span>
        <span class="n">stats_cols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;logBF&quot;</span><span class="p">,</span>
            <span class="s2">&quot;posteriorProbability&quot;</span><span class="p">,</span>
            <span class="s2">&quot;beta&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pValueMantissa&quot;</span><span class="p">,</span>
            <span class="s2">&quot;pValueExponent&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">overlapping_left</span> <span class="o">=</span> <span class="n">loci_to_overlap</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">),</span>
            <span class="o">*</span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">col</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;left_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">peak_overlaps</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>

        <span class="c1"># Complete information about all tags in the right study-locus of the overlap</span>
        <span class="n">overlapping_right</span> <span class="o">=</span> <span class="n">loci_to_overlap</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">),</span>
            <span class="o">*</span><span class="p">[</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="n">col</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;right_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">],</span>
        <span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">peak_overlaps</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>

        <span class="c1"># Include information about all tag variants in both study-locus aligned by tag variant id</span>
        <span class="n">overlaps</span> <span class="o">=</span> <span class="n">overlapping_left</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">overlapping_right</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;tagVariantId&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">,</span>
        <span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;leftStudyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;rightStudyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;tagVariantId&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;left_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">]</span> <span class="o">+</span> <span class="p">[</span><span class="sa">f</span><span class="s2">&quot;right_</span><span class="si">{</span><span class="n">e</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">e</span> <span class="ow">in</span> <span class="n">stats_cols</span><span class="p">]</span>
            <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;statistics&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyLocusOverlap</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">overlaps</span><span class="p">,</span>
            <span class="n">_schema</span><span class="o">=</span><span class="n">StudyLocusOverlap</span><span class="o">.</span><span class="n">get_schema</span><span class="p">(),</span>
        <span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">update_quality_flag</span><span class="p">(</span>
        <span class="n">qc</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flag_condition</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flag_text</span><span class="p">:</span> <span class="n">StudyLocusQualityCheck</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Update the provided quality control list with a new flag if condition is met.</span>

<span class="sd">        Args:</span>
<span class="sd">            qc (Column): Array column with the current list of qc flags.</span>
<span class="sd">            flag_condition (Column): This is a column of booleans, signing which row should be flagged</span>
<span class="sd">            flag_text (StudyLocusQualityCheck): Text for the new quality control flag</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Array column with the updated list of qc flags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">qc</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">qc</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">qc</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">flag_condition</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">array_union</span><span class="p">(</span><span class="n">qc</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">flag_text</span><span class="o">.</span><span class="n">value</span><span class="p">))),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">qc</span><span class="p">)</span>

    <span class="nd">@staticmethod</span>
    <span class="k">def</span> <span class="nf">assign_study_locus_id</span><span class="p">(</span><span class="n">study_id_col</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">variant_id_col</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Hashes a column with a variant ID and a study ID to extract a consistent studyLocusId.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_id_col (Column): column name with a study ID</span>
<span class="sd">            variant_id_col (Column): column name with a variant ID</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: column with a study locus ID</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; df = spark.createDataFrame([(&quot;GCST000001&quot;, &quot;1_1000_A_C&quot;), (&quot;GCST000002&quot;, &quot;1_1000_A_C&quot;)]).toDF(&quot;studyId&quot;, &quot;variantId&quot;)</span>
<span class="sd">            &gt;&gt;&gt; df.withColumn(&quot;study_locus_id&quot;, StudyLocus.assign_study_locus_id(f.col(&quot;studyId&quot;), f.col(&quot;variantId&quot;))).show()</span>
<span class="sd">            +----------+----------+-------------------+</span>
<span class="sd">            |   studyId| variantId|     study_locus_id|</span>
<span class="sd">            +----------+----------+-------------------+</span>
<span class="sd">            |GCST000001|1_1000_A_C|1553357789130151995|</span>
<span class="sd">            |GCST000002|1_1000_A_C|-415050894682709184|</span>
<span class="sd">            +----------+----------+-------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">variant_id_col</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">variant_id_col</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">rand</span><span class="p">()</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="s2">&quot;string&quot;</span><span class="p">))</span>
        <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">xxhash64</span><span class="p">(</span><span class="n">study_id_col</span><span class="p">,</span> <span class="n">variant_id_col</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">calculate_credible_set_log10bf</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">],</span> <span class="n">logbfs</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate Bayes factor for the entire credible set. The Bayes factor is calculated as the logsumexp of the logBF values of the variants in the locus.</span>

<span class="sd">        Args:</span>
<span class="sd">            logbfs (Column): Array column with the logBF values of the variants in the locus.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: log10 Bayes factor for the entire credible set.</span>

<span class="sd">        Examples:</span>
<span class="sd">            &gt;&gt;&gt; spark.createDataFrame([([0.2, 0.1, 0.05, 0.0],)]).toDF(&quot;logBF&quot;).select(f.round(StudyLocus.calculate_credible_set_log10bf(f.col(&quot;logBF&quot;)), 7).alias(&quot;credibleSetlog10BF&quot;)).show()</span>
<span class="sd">            +------------------+</span>
<span class="sd">            |credibleSetlog10BF|</span>
<span class="sd">            +------------------+</span>
<span class="sd">            |         1.4765565|</span>
<span class="sd">            +------------------+</span>
<span class="sd">            &lt;BLANKLINE&gt;</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">logsumexp_udf</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">udf</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">get_logsum</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">FloatType</span><span class="p">())</span>
        <span class="k">return</span> <span class="n">logsumexp_udf</span><span class="p">(</span><span class="n">logbfs</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="s2">&quot;double&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;credibleSetlog10BF&quot;</span><span class="p">)</span>

    <span class="nd">@classmethod</span>
    <span class="k">def</span> <span class="nf">get_schema</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">StructType</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Provides the schema for the StudyLocus dataset.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StructType: schema for the StudyLocus dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;study_locus.json&quot;</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">filter_by_study_type</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Creates a new StudyLocus dataset filtered by study type.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_type (str): Study type to filter for. Can be one of `gwas`, `eqtl`, `pqtl`, `eqtl`.</span>
<span class="sd">            study_index (StudyIndex): Study index to resolve study types.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Filtered study-locus dataset.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If study type is not supported.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="n">study_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gwas&quot;</span><span class="p">,</span> <span class="s2">&quot;eqtl&quot;</span><span class="p">,</span> <span class="s2">&quot;pqtl&quot;</span><span class="p">,</span> <span class="s2">&quot;sqtl&quot;</span><span class="p">]:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;Study type </span><span class="si">{</span><span class="n">study_type</span><span class="si">}</span><span class="s2"> not supported. Supported types are: gwas, eqtl, pqtl, sqtl.&quot;</span>
            <span class="p">)</span>
        <span class="n">new_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">study_index</span><span class="o">.</span><span class="n">study_type_lut</span><span class="p">(),</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">study_type</span><span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
            <span class="n">_df</span><span class="o">=</span><span class="n">new_df</span><span class="p">,</span>
            <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_schema</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">filter_credible_set</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">credible_interval</span><span class="p">:</span> <span class="n">CredibleInterval</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Filter study-locus tag variants based on given credible interval.</span>

<span class="sd">        Args:</span>
<span class="sd">            credible_interval (CredibleInterval): Credible interval to filter for.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Filtered study-locus dataset.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                <span class="k">lambda</span> <span class="n">tag</span><span class="p">:</span> <span class="p">(</span><span class="n">tag</span><span class="p">[</span><span class="n">credible_interval</span><span class="o">.</span><span class="n">value</span><span class="p">]),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">find_overlaps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate overlapping study-locus.</span>

<span class="sd">        Find overlapping study-locus that share at least one tagging variant. All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always</span>
<span class="sd">        appearing on the right side.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_index (StudyIndex): Study index to resolve study types.</span>
<span class="sd">            intra_study_overlap (bool): If True, finds intra-study overlaps for credible set deduplication. Default is False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocusOverlap: Pairs of overlapping study-locus with aligned tags.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">loci_to_overlap</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">study_index</span><span class="o">.</span><span class="n">study_type_lut</span><span class="p">(),</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
                <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
                <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
                <span class="s2">&quot;region&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.logBF&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="p">)</span>

        <span class="c1"># overlapping study-locus</span>
        <span class="n">peak_overlaps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_overlapping_peaks</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">)</span>

        <span class="c1"># study-locus overlap by aligning overlapping variants</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_align_overlapping_tags</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">peak_overlaps</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">unique_variants_in_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;All unique variants collected in a `StudyLocus` dataframe.</span>

<span class="sd">        Returns:</span>
<span class="sd">            DataFrame: A dataframe containing `variantId` and `chromosome` columns.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
                <span class="c1"># Joint array of variants in that studylocus. Locus can be null</span>
                <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array_union</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()),</span>
                    <span class="p">)</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">select</span><span class="p">(</span>
                <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">neglog_pvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Returns the negative log p-value.</span>

<span class="sd">        Returns:</span>
<span class="sd">            Column: Negative log p-value</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="n">calculate_neglog_pvalue</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
        <span class="p">)</span>

    <span class="k">def</span> <span class="nf">annotate_credible_sets</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotate study-locus dataset with credible set flags.</span>

<span class="sd">        Sorts the array in the `locus` column elements by their `posteriorProbability` values in descending order and adds</span>
<span class="sd">        `is95CredibleSet` and `is99CredibleSet` fields to the elements, indicating which are the tagging variants whose cumulative sum</span>
<span class="sd">        of their `posteriorProbability` values is below 0.95 and 0.99, respectively.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: including annotation on `is95CredibleSet` and `is99CredibleSet`.</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: If `locus` column is not available.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">if</span> <span class="s2">&quot;locus&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Locus column not available.&quot;</span><span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="c1"># Sort credible set by posterior probability in descending order</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">order_array_of_structs_by_field</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="c1"># Calculate array of cumulative sums of posterior probabilities to determine which variants are in the 95% and 99% credible sets</span>
            <span class="c1"># and zip the cumulative sums array with the credible set array to add the flags</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">zip_with</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))),</span>
                        <span class="k">lambda</span> <span class="n">index</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span>
                                <span class="c1"># By using `index - 1` we introduce a value of `0.0` in the cumulative sums array. to ensure that the last variant</span>
                                <span class="c1"># that exceeds the 0.95 threshold is included in the cumulative sum, as its probability is necessary to satisfy the threshold.</span>
                                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">),</span>
                                <span class="mi">1</span><span class="p">,</span>
                                <span class="n">index</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                            <span class="p">),</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
                            <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">el</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">el</span><span class="p">,</span>
                        <span class="p">),</span>
                    <span class="p">),</span>
                    <span class="k">lambda</span> <span class="n">struct_e</span><span class="p">,</span> <span class="n">acc</span><span class="p">:</span> <span class="n">struct_e</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                        <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS95</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.95</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                    <span class="p">)</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                        <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS99</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.99</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                    <span class="p">),</span>
                <span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">annotate_locus_statistics</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
        <span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span>
        <span class="n">collect_locus_distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotates study locus with summary statistics in the specified distance around the position.</span>

<span class="sd">        Args:</span>
<span class="sd">            summary_statistics (SummaryStatistics): Summary statistics to be used for annotation.</span>
<span class="sd">            collect_locus_distance (int): distance from variant defining window for inclusion of variants in locus.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Study locus annotated with summary statistics in `locus` column. If no statistics are found, the `locus` column will be empty.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># The clumps will be used several times (persisting)</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
        <span class="c1"># Renaming columns:</span>
        <span class="n">sumstats_renamed</span> <span class="o">=</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span>
            <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> as tag_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
        <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstat&quot;</span><span class="p">)</span>

        <span class="n">locus_df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">sumstats_renamed</span>
            <span class="c1"># Joining the two datasets together:</span>
            <span class="o">.</span><span class="n">join</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;clumped&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                        <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;studyLocusId&quot;</span>
                    <span class="p">)</span>
                <span class="p">),</span>
                <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                    <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.studyId&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.chromosome&quot;</span><span class="p">))</span>
                    <span class="o">&amp;</span> <span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                        <span class="o">&gt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                    <span class="p">)</span>
                    <span class="o">&amp;</span> <span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                        <span class="o">&lt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">],</span>
                <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
            <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">locus_df</span><span class="p">,</span>
            <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
        <span class="p">)</span>

        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">annotate_ld</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">:</span> <span class="n">LDIndex</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Annotate LD information to study-locus.</span>

<span class="sd">        Args:</span>
<span class="sd">            study_index (StudyIndex): Study index to resolve ancestries.</span>
<span class="sd">            ld_index (LDIndex): LD index to resolve LD information.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Study locus annotated with ld information from LD index.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="kn">from</span> <span class="nn">gentropy.method.ld</span> <span class="kn">import</span> <span class="n">LDAnnotator</span>

        <span class="k">return</span> <span class="n">LDAnnotator</span><span class="o">.</span><span class="n">ld_annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_index</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">)</span>

    <span class="k">def</span> <span class="nf">clump</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Perform LD clumping of the studyLocus.</span>

<span class="sd">        Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: with empty credible sets for linked variants and QC flag.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;is_lead_linked&quot;</span><span class="p">,</span>
                <span class="n">LDclumping</span><span class="o">.</span><span class="n">_is_lead_linked</span><span class="p">(</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyId</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">variantId</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
                    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">ldSet</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;ldSet&quot;</span><span class="p">,</span>
                <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ldSet&quot;</span><span class="p">)),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
                <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
                <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span>
                    <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">LD_CLUMPED</span><span class="p">,</span>
                <span class="p">),</span>
            <span class="p">)</span>
            <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>

    <span class="k">def</span> <span class="nf">_qc_no_population</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Flag associations where the study doesn&#39;t have population information to resolve LD.</span>

<span class="sd">        Returns:</span>
<span class="sd">            StudyLocus: Updated study locus.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># If the tested column is not present, return self unchanged:</span>
        <span class="k">if</span> <span class="s2">&quot;ldPopulationStructure&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">return</span> <span class="bp">self</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ldPopulationStructure&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">NO_POPULATION</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">









<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_credible_sets</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_credible_sets" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Annotate study-locus dataset with credible set flags.</p>
<p>Sorts the array in the <code>locus</code> column elements by their <code>posteriorProbability</code> values in descending order and adds
<code>is95CredibleSet</code> and <code>is99CredibleSet</code> fields to the elements, indicating which are the tagging variants whose cumulative sum
of their <code>posteriorProbability</code> values is below 0.95 and 0.99, respectively.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>including annotation on <code>is95CredibleSet</code> and <code>is99CredibleSet</code>.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If <code>locus</code> column is not available.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_credible_sets</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate study-locus dataset with credible set flags.</span>

<span class="sd">    Sorts the array in the `locus` column elements by their `posteriorProbability` values in descending order and adds</span>
<span class="sd">    `is95CredibleSet` and `is99CredibleSet` fields to the elements, indicating which are the tagging variants whose cumulative sum</span>
<span class="sd">    of their `posteriorProbability` values is below 0.95 and 0.99, respectively.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: including annotation on `is95CredibleSet` and `is99CredibleSet`.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If `locus` column is not available.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="s2">&quot;locus&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Locus column not available.&quot;</span><span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="c1"># Sort credible set by posterior probability in descending order</span>
        <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
            <span class="n">order_array_of_structs_by_field</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="c1"># Calculate array of cumulative sums of posterior probabilities to determine which variants are in the 95% and 99% credible sets</span>
        <span class="c1"># and zip the cumulative sums array with the credible set array to add the flags</span>
        <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span> <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">zip_with</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">transform</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">sequence</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">size</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))),</span>
                    <span class="k">lambda</span> <span class="n">index</span><span class="p">:</span> <span class="n">f</span><span class="o">.</span><span class="n">aggregate</span><span class="p">(</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">slice</span><span class="p">(</span>
                            <span class="c1"># By using `index - 1` we introduce a value of `0.0` in the cumulative sums array. to ensure that the last variant</span>
                            <span class="c1"># that exceeds the 0.95 threshold is included in the cumulative sum, as its probability is necessary to satisfy the threshold.</span>
                            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">),</span>
                            <span class="mi">1</span><span class="p">,</span>
                            <span class="n">index</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span>
                        <span class="p">),</span>
                        <span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="mf">0.0</span><span class="p">),</span>
                        <span class="k">lambda</span> <span class="n">acc</span><span class="p">,</span> <span class="n">el</span><span class="p">:</span> <span class="n">acc</span> <span class="o">+</span> <span class="n">el</span><span class="p">,</span>
                    <span class="p">),</span>
                <span class="p">),</span>
                <span class="k">lambda</span> <span class="n">struct_e</span><span class="p">,</span> <span class="n">acc</span><span class="p">:</span> <span class="n">struct_e</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                    <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS95</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.95</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                <span class="p">)</span><span class="o">.</span><span class="n">withField</span><span class="p">(</span>
                    <span class="n">CredibleInterval</span><span class="o">.</span><span class="n">IS99</span><span class="o">.</span><span class="n">value</span><span class="p">,</span> <span class="p">(</span><span class="n">acc</span> <span class="o">&lt;</span> <span class="mf">0.99</span><span class="p">)</span> <span class="o">&amp;</span> <span class="n">acc</span><span class="o">.</span><span class="n">isNotNull</span><span class="p">()</span>
                <span class="p">),</span>
            <span class="p">),</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_ld</span><span class="p">(</span><span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">:</span> <span class="n">LDIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_ld" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Annotate LD information to study-locus.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>study_index</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study index to resolve ancestries.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>ld_index</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.ld_index.LDIndex" href="../ld_index/#gentropy.dataset.ld_index.LDIndex">LDIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>LD index to resolve LD information.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus annotated with ld information from LD index.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_ld</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">:</span> <span class="n">LDIndex</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotate LD information to study-locus.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_index (StudyIndex): Study index to resolve ancestries.</span>
<span class="sd">        ld_index (LDIndex): LD index to resolve LD information.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Study locus annotated with ld information from LD index.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="kn">from</span> <span class="nn">gentropy.method.ld</span> <span class="kn">import</span> <span class="n">LDAnnotator</span>

    <span class="k">return</span> <span class="n">LDAnnotator</span><span class="o">.</span><span class="n">ld_annotate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">study_index</span><span class="p">,</span> <span class="n">ld_index</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">annotate_locus_statistics</span><span class="p">(</span><span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span> <span class="n">collect_locus_distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.annotate_locus_statistics" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Annotates study locus with summary statistics in the specified distance around the position.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>summary_statistics</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.summary_statistics.SummaryStatistics" href="../summary_statistics/#gentropy.dataset.summary_statistics.SummaryStatistics">SummaryStatistics</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Summary statistics to be used for annotation.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>collect_locus_distance</code></td>
            <td>
                  <code>int</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>distance from variant defining window for inclusion of variants in locus.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study locus annotated with summary statistics in <code>locus</code> column. If no statistics are found, the <code>locus</code> column will be empty.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">annotate_locus_statistics</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">summary_statistics</span><span class="p">:</span> <span class="n">SummaryStatistics</span><span class="p">,</span>
    <span class="n">collect_locus_distance</span><span class="p">:</span> <span class="nb">int</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Annotates study locus with summary statistics in the specified distance around the position.</span>

<span class="sd">    Args:</span>
<span class="sd">        summary_statistics (SummaryStatistics): Summary statistics to be used for annotation.</span>
<span class="sd">        collect_locus_distance (int): distance from variant defining window for inclusion of variants in locus.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Study locus annotated with summary statistics in `locus` column. If no statistics are found, the `locus` column will be empty.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># The clumps will be used several times (persisting)</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="c1"># Renaming columns:</span>
    <span class="n">sumstats_renamed</span> <span class="o">=</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">selectExpr</span><span class="p">(</span>
        <span class="o">*</span><span class="p">[</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> as tag_</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">summary_statistics</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>
    <span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;sumstat&quot;</span><span class="p">)</span>

    <span class="n">locus_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">sumstats_renamed</span>
        <span class="c1"># Joining the two datasets together:</span>
        <span class="o">.</span><span class="n">join</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">broadcast</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;clumped&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span>
                    <span class="s2">&quot;position&quot;</span><span class="p">,</span> <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;studyLocusId&quot;</span>
                <span class="p">)</span>
            <span class="p">),</span>
            <span class="n">on</span><span class="o">=</span><span class="p">[</span>
                <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_studyId&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.studyId&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_chromosome&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.chromosome&quot;</span><span class="p">))</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                    <span class="o">&gt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">-</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="o">&amp;</span> <span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;sumstat.tag_position&quot;</span><span class="p">)</span>
                    <span class="o">&lt;=</span> <span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;clumped.position&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="n">collect_locus_distance</span><span class="p">)</span>
                <span class="p">)</span>
            <span class="p">],</span>
            <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">struct</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;tag_standardError&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;standardError&quot;</span><span class="p">),</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">groupBy</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">agg</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">collect_list</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
        <span class="p">)</span>
    <span class="p">)</span>

    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">locus_df</span><span class="p">,</span>
        <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
        <span class="n">how</span><span class="o">=</span><span class="s2">&quot;left&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">assign_study_locus_id</span><span class="p">(</span><span class="n">study_id_col</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">variant_id_col</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.assign_study_locus_id" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Hashes a column with a variant ID and a study ID to extract a consistent studyLocusId.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>study_id_col</code></td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>column name with a study ID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>variant_id_col</code></td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>column name with a variant ID</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>column with a study locus ID</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span> <span class="o">=</span> <span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">([(</span><span class="s2">&quot;GCST000001&quot;</span><span class="p">,</span> <span class="s2">&quot;1_1000_A_C&quot;</span><span class="p">),</span> <span class="p">(</span><span class="s2">&quot;GCST000002&quot;</span><span class="p">,</span> <span class="s2">&quot;1_1000_A_C&quot;</span><span class="p">)])</span><span class="o">.</span><span class="n">toDF</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="s2">&quot;variantId&quot;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;study_locus_id&quot;</span><span class="p">,</span> <span class="n">StudyLocus</span><span class="o">.</span><span class="n">assign_study_locus_id</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)))</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+----------+----------+-------------------+</span>
<span class="go">|   studyId| variantId|     study_locus_id|</span>
<span class="go">+----------+----------+-------------------+</span>
<span class="go">|GCST000001|1_1000_A_C|1553357789130151995|</span>
<span class="go">|GCST000002|1_1000_A_C|-415050894682709184|</span>
<span class="go">+----------+----------+-------------------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">assign_study_locus_id</span><span class="p">(</span><span class="n">study_id_col</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">variant_id_col</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Hashes a column with a variant ID and a study ID to extract a consistent studyLocusId.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_id_col (Column): column name with a study ID</span>
<span class="sd">        variant_id_col (Column): column name with a variant ID</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: column with a study locus ID</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; df = spark.createDataFrame([(&quot;GCST000001&quot;, &quot;1_1000_A_C&quot;), (&quot;GCST000002&quot;, &quot;1_1000_A_C&quot;)]).toDF(&quot;studyId&quot;, &quot;variantId&quot;)</span>
<span class="sd">        &gt;&gt;&gt; df.withColumn(&quot;study_locus_id&quot;, StudyLocus.assign_study_locus_id(f.col(&quot;studyId&quot;), f.col(&quot;variantId&quot;))).show()</span>
<span class="sd">        +----------+----------+-------------------+</span>
<span class="sd">        |   studyId| variantId|     study_locus_id|</span>
<span class="sd">        +----------+----------+-------------------+</span>
<span class="sd">        |GCST000001|1_1000_A_C|1553357789130151995|</span>
<span class="sd">        |GCST000002|1_1000_A_C|-415050894682709184|</span>
<span class="sd">        +----------+----------+-------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variant_id_col</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">variant_id_col</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">rand</span><span class="p">()</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="s2">&quot;string&quot;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">xxhash64</span><span class="p">(</span><span class="n">study_id_col</span><span class="p">,</span> <span class="n">variant_id_col</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;studyLocusId&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">calculate_credible_set_log10bf</span><span class="p">(</span><span class="n">logbfs</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.calculate_credible_set_log10bf" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Calculate Bayes factor for the entire credible set. The Bayes factor is calculated as the logsumexp of the logBF values of the variants in the locus.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>logbfs</code></td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array column with the logBF values of the variants in the locus.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>log10 Bayes factor for the entire credible set.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Examples:</span></p>
    <div class="highlight"><pre><span></span><code><span class="gp">&gt;&gt;&gt; </span><span class="n">spark</span><span class="o">.</span><span class="n">createDataFrame</span><span class="p">([([</span><span class="mf">0.2</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.05</span><span class="p">,</span> <span class="mf">0.0</span><span class="p">],)])</span><span class="o">.</span><span class="n">toDF</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">select</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">round</span><span class="p">(</span><span class="n">StudyLocus</span><span class="o">.</span><span class="n">calculate_credible_set_log10bf</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">)),</span> <span class="mi">7</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;credibleSetlog10BF&quot;</span><span class="p">))</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
<span class="go">+------------------+</span>
<span class="go">|credibleSetlog10BF|</span>
<span class="go">+------------------+</span>
<span class="go">|         1.4765565|</span>
<span class="go">+------------------+</span>
</code></pre></div>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">calculate_credible_set_log10bf</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">],</span> <span class="n">logbfs</span><span class="p">:</span> <span class="n">Column</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate Bayes factor for the entire credible set. The Bayes factor is calculated as the logsumexp of the logBF values of the variants in the locus.</span>

<span class="sd">    Args:</span>
<span class="sd">        logbfs (Column): Array column with the logBF values of the variants in the locus.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: log10 Bayes factor for the entire credible set.</span>

<span class="sd">    Examples:</span>
<span class="sd">        &gt;&gt;&gt; spark.createDataFrame([([0.2, 0.1, 0.05, 0.0],)]).toDF(&quot;logBF&quot;).select(f.round(StudyLocus.calculate_credible_set_log10bf(f.col(&quot;logBF&quot;)), 7).alias(&quot;credibleSetlog10BF&quot;)).show()</span>
<span class="sd">        +------------------+</span>
<span class="sd">        |credibleSetlog10BF|</span>
<span class="sd">        +------------------+</span>
<span class="sd">        |         1.4765565|</span>
<span class="sd">        +------------------+</span>
<span class="sd">        &lt;BLANKLINE&gt;</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logsumexp_udf</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">udf</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">get_logsum</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="n">FloatType</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">logsumexp_udf</span><span class="p">(</span><span class="n">logbfs</span><span class="p">)</span><span class="o">.</span><span class="n">cast</span><span class="p">(</span><span class="s2">&quot;double&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;credibleSetlog10BF&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.clump" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">clump</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.clump" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Perform LD clumping of the studyLocus.</p>
<p>Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>with empty credible sets for linked variants and QC flag.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">clump</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Perform LD clumping of the studyLocus.</span>

<span class="sd">    Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: with empty credible sets for linked variants and QC flag.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;is_lead_linked&quot;</span><span class="p">,</span>
            <span class="n">LDclumping</span><span class="o">.</span><span class="n">_is_lead_linked</span><span class="p">(</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">studyId</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">variantId</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
                <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">ldSet</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;ldSet&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;ldSet&quot;</span><span class="p">)),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;qualityControls&quot;</span><span class="p">,</span>
            <span class="n">StudyLocus</span><span class="o">.</span><span class="n">update_quality_flag</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;qualityControls&quot;</span><span class="p">),</span>
                <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">),</span>
                <span class="n">StudyLocusQualityCheck</span><span class="o">.</span><span class="n">LD_CLUMPED</span><span class="p">,</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;is_lead_linked&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.filter_by_study_type" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_by_study_type</span><span class="p">(</span><span class="n">study_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.filter_by_study_type" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Creates a new StudyLocus dataset filtered by study type.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>study_type</code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study type to filter for. Can be one of <code>gwas</code>, <code>eqtl</code>, <code>pqtl</code>, <code>eqtl</code>.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>study_index</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study index to resolve study types.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered study-locus dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


<p><span class="doc-section-title">Raises:</span></p>
    <table>
      <thead>
        <tr>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                  <code>ValueError</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If study type is not supported.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_by_study_type</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_type</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates a new StudyLocus dataset filtered by study type.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_type (str): Study type to filter for. Can be one of `gwas`, `eqtl`, `pqtl`, `eqtl`.</span>
<span class="sd">        study_index (StudyIndex): Study index to resolve study types.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Filtered study-locus dataset.</span>

<span class="sd">    Raises:</span>
<span class="sd">        ValueError: If study type is not supported.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">study_type</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;gwas&quot;</span><span class="p">,</span> <span class="s2">&quot;eqtl&quot;</span><span class="p">,</span> <span class="s2">&quot;pqtl&quot;</span><span class="p">,</span> <span class="s2">&quot;sqtl&quot;</span><span class="p">]:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;Study type </span><span class="si">{</span><span class="n">study_type</span><span class="si">}</span><span class="s2"> not supported. Supported types are: gwas, eqtl, pqtl, sqtl.&quot;</span>
        <span class="p">)</span>
    <span class="n">new_df</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">study_index</span><span class="o">.</span><span class="n">study_type_lut</span><span class="p">(),</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span> <span class="o">==</span> <span class="n">study_type</span><span class="p">)</span>
        <span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;studyType&quot;</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">StudyLocus</span><span class="p">(</span>
        <span class="n">_df</span><span class="o">=</span><span class="n">new_df</span><span class="p">,</span>
        <span class="n">_schema</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_schema</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_credible_set</span><span class="p">(</span><span class="n">credible_interval</span><span class="p">:</span> <span class="n">CredibleInterval</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.filter_credible_set" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Filter study-locus tag variants based on given credible interval.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>credible_interval</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.CredibleInterval" href="#gentropy.dataset.study_locus.CredibleInterval">CredibleInterval</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Credible interval to filter for.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocus</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocus" href="#gentropy.dataset.study_locus.StudyLocus">StudyLocus</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Filtered study-locus dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">filter_credible_set</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span>
    <span class="n">credible_interval</span><span class="p">:</span> <span class="n">CredibleInterval</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocus</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filter study-locus tag variants based on given credible interval.</span>

<span class="sd">    Args:</span>
<span class="sd">        credible_interval (CredibleInterval): Credible interval to filter for.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocus: Filtered study-locus dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
        <span class="s2">&quot;locus&quot;</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">),</span>
            <span class="k">lambda</span> <span class="n">tag</span><span class="p">:</span> <span class="p">(</span><span class="n">tag</span><span class="p">[</span><span class="n">credible_interval</span><span class="o">.</span><span class="n">value</span><span class="p">]),</span>
        <span class="p">),</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">find_overlaps</span><span class="p">(</span><span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.find_overlaps" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Calculate overlapping study-locus.</p>
<p>Find overlapping study-locus that share at least one tagging variant. All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always
appearing on the right side.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>study_index</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_index.StudyIndex" href="../study_index/#gentropy.dataset.study_index.StudyIndex">StudyIndex</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study index to resolve study types.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>intra_study_overlap</code></td>
            <td>
                  <code>bool</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>If True, finds intra-study overlaps for credible set deduplication. Default is False.</p>
              </div>
            </td>
            <td>
                  <code>False</code>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StudyLocusOverlap</code></td>            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus_overlap.StudyLocusOverlap" href="../study_locus_overlap/#gentropy.dataset.study_locus_overlap.StudyLocusOverlap">StudyLocusOverlap</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Pairs of overlapping study-locus with aligned tags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">find_overlaps</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">,</span> <span class="n">study_index</span><span class="p">:</span> <span class="n">StudyIndex</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">StudyLocusOverlap</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate overlapping study-locus.</span>

<span class="sd">    Find overlapping study-locus that share at least one tagging variant. All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always</span>
<span class="sd">    appearing on the right side.</span>

<span class="sd">    Args:</span>
<span class="sd">        study_index (StudyIndex): Study index to resolve study types.</span>
<span class="sd">        intra_study_overlap (bool): If True, finds intra-study overlaps for credible set deduplication. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StudyLocusOverlap: Pairs of overlapping study-locus with aligned tags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">loci_to_overlap</span> <span class="o">=</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">study_index</span><span class="o">.</span><span class="n">study_type_lut</span><span class="p">(),</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;studyId&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;inner&quot;</span><span class="p">)</span>
        <span class="o">.</span><span class="n">withColumn</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span><span class="s2">&quot;locus&quot;</span><span class="p">))</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;studyLocusId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyId&quot;</span><span class="p">,</span>
            <span class="s2">&quot;studyType&quot;</span><span class="p">,</span>
            <span class="s2">&quot;chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;region&quot;</span><span class="p">,</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;tagVariantId&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.logBF&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;logBF&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.posteriorProbability&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;posteriorProbability&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueMantissa&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueMantissa&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.pValueExponent&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;pValueExponent&quot;</span><span class="p">),</span>
            <span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.beta&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;beta&quot;</span><span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">persist</span><span class="p">()</span>
    <span class="p">)</span>

    <span class="c1"># overlapping study-locus</span>
    <span class="n">peak_overlaps</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_overlapping_peaks</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">intra_study_overlap</span><span class="p">)</span>

    <span class="c1"># study-locus overlap by aligning overlapping variants</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_align_overlapping_tags</span><span class="p">(</span><span class="n">loci_to_overlap</span><span class="p">,</span> <span class="n">peak_overlaps</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.get_schema" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_schema</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">StructType</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.get_schema" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Provides the schema for the StudyLocus dataset.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>StructType</code></td>            <td>
                  <code><span title="pyspark.sql.types.StructType">StructType</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>schema for the StudyLocus dataset.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@classmethod</span>
<span class="k">def</span> <span class="nf">get_schema</span><span class="p">(</span><span class="bp">cls</span><span class="p">:</span> <span class="nb">type</span><span class="p">[</span><span class="n">StudyLocus</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">StructType</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Provides the schema for the StudyLocus dataset.</span>

<span class="sd">    Returns:</span>
<span class="sd">        StructType: schema for the StudyLocus dataset.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">parse_spark_schema</span><span class="p">(</span><span class="s2">&quot;study_locus.json&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">neglog_pvalue</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.neglog_pvalue" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Returns the negative log p-value.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Negative log p-value</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">neglog_pvalue</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Returns the negative log p-value.</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Negative log p-value</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">calculate_neglog_pvalue</span><span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueMantissa</span><span class="p">,</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">pValueExponent</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">unique_variants_in_locus</span><span class="p">()</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span></code>

<a href="#gentropy.dataset.study_locus.StudyLocus.unique_variants_in_locus" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>All unique variants collected in a <code>StudyLocus</code> dataframe.</p>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>DataFrame</code></td>            <td>
                  <code><span title="pyspark.sql.DataFrame">DataFrame</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>A dataframe containing <code>variantId</code> and <code>chromosome</code> columns.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span> <span class="nf">unique_variants_in_locus</span><span class="p">(</span><span class="bp">self</span><span class="p">:</span> <span class="n">StudyLocus</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;All unique variants collected in a `StudyLocus` dataframe.</span>

<span class="sd">    Returns:</span>
<span class="sd">        DataFrame: A dataframe containing `variantId` and `chromosome` columns.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="p">(</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">df</span><span class="o">.</span><span class="n">withColumn</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span>
            <span class="c1"># Joint array of variants in that studylocus. Locus can be null</span>
            <span class="n">f</span><span class="o">.</span><span class="n">explode</span><span class="p">(</span>
                <span class="n">f</span><span class="o">.</span><span class="n">array_union</span><span class="p">(</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">)),</span>
                    <span class="n">f</span><span class="o">.</span><span class="n">coalesce</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;locus.variantId&quot;</span><span class="p">),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">()),</span>
                <span class="p">)</span>
            <span class="p">),</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">select</span><span class="p">(</span>
            <span class="s2">&quot;variantId&quot;</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">col</span><span class="p">(</span><span class="s2">&quot;variantId&quot;</span><span class="p">),</span> <span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">alias</span><span class="p">(</span><span class="s2">&quot;chromosome&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="o">.</span><span class="n">distinct</span><span class="p">()</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h3 id="gentropy.dataset.study_locus.StudyLocus.update_quality_flag" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">update_quality_flag</span><span class="p">(</span><span class="n">qc</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flag_condition</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flag_text</span><span class="p">:</span> <span class="n">StudyLocusQualityCheck</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-staticmethod"><code>staticmethod</code></small>
  </span>

<a href="#gentropy.dataset.study_locus.StudyLocus.update_quality_flag" class="headerlink" title="Permanent link">&para;</a></h3>


    <div class="doc doc-contents ">

      <p>Update the provided quality control list with a new flag if condition is met.</p>


<p><span class="doc-section-title">Parameters:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
          <th>Default</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code>qc</code></td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array column with the current list of qc flags.</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>flag_condition</code></td>
            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>This is a column of booleans, signing which row should be flagged</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code>flag_text</code></td>
            <td>
                  <code><a class="autorefs autorefs-internal" title="gentropy.dataset.study_locus.StudyLocusQualityCheck" href="#gentropy.dataset.study_locus.StudyLocusQualityCheck">StudyLocusQualityCheck</a></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Text for the new quality control flag</p>
              </div>
            </td>
            <td>
                <em>required</em>
            </td>
          </tr>
      </tbody>
    </table>


    <p><span class="doc-section-title">Returns:</span></p>
    <table>
      <thead>
        <tr>
<th>Name</th>          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
<td><code>Column</code></td>            <td>
                  <code><span title="pyspark.sql.Column">Column</span></code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Array column with the updated list of qc flags.</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

            <details class="quote">
              <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nd">@staticmethod</span>
<span class="k">def</span> <span class="nf">update_quality_flag</span><span class="p">(</span>
    <span class="n">qc</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flag_condition</span><span class="p">:</span> <span class="n">Column</span><span class="p">,</span> <span class="n">flag_text</span><span class="p">:</span> <span class="n">StudyLocusQualityCheck</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Column</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Update the provided quality control list with a new flag if condition is met.</span>

<span class="sd">    Args:</span>
<span class="sd">        qc (Column): Array column with the current list of qc flags.</span>
<span class="sd">        flag_condition (Column): This is a column of booleans, signing which row should be flagged</span>
<span class="sd">        flag_text (StudyLocusQualityCheck): Text for the new quality control flag</span>

<span class="sd">    Returns:</span>
<span class="sd">        Column: Array column with the updated list of qc flags.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">qc</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span><span class="n">qc</span><span class="o">.</span><span class="n">isNull</span><span class="p">(),</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">())</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">qc</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">f</span><span class="o">.</span><span class="n">when</span><span class="p">(</span>
        <span class="n">flag_condition</span><span class="p">,</span>
        <span class="n">f</span><span class="o">.</span><span class="n">array_union</span><span class="p">(</span><span class="n">qc</span><span class="p">,</span> <span class="n">f</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">f</span><span class="o">.</span><span class="n">lit</span><span class="p">(</span><span class="n">flag_text</span><span class="o">.</span><span class="n">value</span><span class="p">))),</span>
    <span class="p">)</span><span class="o">.</span><span class="n">otherwise</span><span class="p">(</span><span class="n">qc</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div><hr />


<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.StudyLocusQualityCheck" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.StudyLocusQualityCheck</code>


<a href="#gentropy.dataset.study_locus.StudyLocusQualityCheck" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


      <p>Study-Locus quality control options listing concerns on the quality of the association.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.SUBSIGNIFICANT_FLAG">SUBSIGNIFICANT_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>p-value below significance threshold</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NO_GENOMIC_LOCATION_FLAG">NO_GENOMIC_LOCATION_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Incomplete genomic mapping</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.COMPOSITE_FLAG">COMPOSITE_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Composite association due to variant x variant interactions</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.INCONSISTENCY_FLAG">INCONSISTENCY_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Inconsistencies in the reported variants</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NON_MAPPED_VARIANT_FLAG">NON_MAPPED_VARIANT_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant not mapped to GnomAd</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.PALINDROMIC_ALLELE_FLAG">PALINDROMIC_ALLELE_FLAG</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Alleles are palindromic - cannot harmonize</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.AMBIGUOUS_STUDY">AMBIGUOUS_STUDY</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Association with ambiguous study</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.UNRESOLVED_LD">UNRESOLVED_LD</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Variant not found in LD reference</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.LD_CLUMPED">LD_CLUMPED</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Explained by a more significant variant in high LD (clumped)</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NO_POPULATION">NO_POPULATION</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>Study does not have population annotation to resolve LD</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.StudyLocusQualityCheck.NOT_QUALIFYING_LD_BLOCK">NOT_QUALIFYING_LD_BLOCK</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>LD block does not contain variants at the required R^2 threshold</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span>
<span class="normal">56</span>
<span class="normal">57</span>
<span class="normal">58</span>
<span class="normal">59</span>
<span class="normal">60</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">StudyLocusQualityCheck</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Study-Locus quality control options listing concerns on the quality of the association.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        SUBSIGNIFICANT_FLAG (str): p-value below significance threshold</span>
<span class="sd">        NO_GENOMIC_LOCATION_FLAG (str): Incomplete genomic mapping</span>
<span class="sd">        COMPOSITE_FLAG (str): Composite association due to variant x variant interactions</span>
<span class="sd">        INCONSISTENCY_FLAG (str): Inconsistencies in the reported variants</span>
<span class="sd">        NON_MAPPED_VARIANT_FLAG (str): Variant not mapped to GnomAd</span>
<span class="sd">        PALINDROMIC_ALLELE_FLAG (str): Alleles are palindromic - cannot harmonize</span>
<span class="sd">        AMBIGUOUS_STUDY (str): Association with ambiguous study</span>
<span class="sd">        UNRESOLVED_LD (str): Variant not found in LD reference</span>
<span class="sd">        LD_CLUMPED (str): Explained by a more significant variant in high LD (clumped)</span>
<span class="sd">        NO_POPULATION (str): Study does not have population annotation to resolve LD</span>
<span class="sd">        NOT_QUALIFYING_LD_BLOCK (str): LD block does not contain variants at the required R^2 threshold</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">SUBSIGNIFICANT_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Subsignificant p-value&quot;</span>
    <span class="n">NO_GENOMIC_LOCATION_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Incomplete genomic mapping&quot;</span>
    <span class="n">COMPOSITE_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Composite association&quot;</span>
    <span class="n">INCONSISTENCY_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Variant inconsistency&quot;</span>
    <span class="n">NON_MAPPED_VARIANT_FLAG</span> <span class="o">=</span> <span class="s2">&quot;No mapping in GnomAd&quot;</span>
    <span class="n">PALINDROMIC_ALLELE_FLAG</span> <span class="o">=</span> <span class="s2">&quot;Palindrome alleles - cannot harmonize&quot;</span>
    <span class="n">AMBIGUOUS_STUDY</span> <span class="o">=</span> <span class="s2">&quot;Association with ambiguous study&quot;</span>
    <span class="n">UNRESOLVED_LD</span> <span class="o">=</span> <span class="s2">&quot;Variant not found in LD reference&quot;</span>
    <span class="n">LD_CLUMPED</span> <span class="o">=</span> <span class="s2">&quot;Explained by a more significant variant in high LD (clumped)&quot;</span>
    <span class="n">NO_POPULATION</span> <span class="o">=</span> <span class="s2">&quot;Study does not have population annotation to resolve LD&quot;</span>
    <span class="n">NOT_QUALIFYING_LD_BLOCK</span> <span class="o">=</span> <span class="p">(</span>
        <span class="s2">&quot;LD block does not contain variants at the required R^2 threshold&quot;</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div><hr />


<div class="doc doc-object doc-class">



<h2 id="gentropy.dataset.study_locus.CredibleInterval" class="doc doc-heading">
            <code>gentropy.dataset.study_locus.CredibleInterval</code>


<a href="#gentropy.dataset.study_locus.CredibleInterval" class="headerlink" title="Permanent link">&para;</a></h2>


    <div class="doc doc-contents first">
            <p class="doc doc-class-bases">
              Bases: <code><span title="enum.Enum">Enum</span></code></p>


      <p>Credible interval enum.</p>
<p>Interval within which an unobserved parameter value falls with a particular probability.</p>


<p><span class="doc-section-title">Attributes:</span></p>
    <table>
      <thead>
        <tr>
          <th>Name</th>
          <th>Type</th>
          <th>Description</th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.CredibleInterval.IS95">IS95</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>95% credible interval</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td><code><span title="gentropy.dataset.study_locus.CredibleInterval.IS99">IS99</span></code></td>
            <td>
                  <code>str</code>
            </td>
            <td>
              <div class="doc-md-description">
                <p>99% credible interval</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>

              <details class="quote">
                <summary>Source code in <code>src/gentropy/dataset/study_locus.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span> <span class="nc">CredibleInterval</span><span class="p">(</span><span class="n">Enum</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Credible interval enum.</span>

<span class="sd">    Interval within which an unobserved parameter value falls with a particular probability.</span>

<span class="sd">    Attributes:</span>
<span class="sd">        IS95 (str): 95% credible interval</span>
<span class="sd">        IS99 (str): 99% credible interval</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">IS95</span> <span class="o">=</span> <span class="s2">&quot;is95CredibleSet&quot;</span>
    <span class="n">IS99</span> <span class="o">=</span> <span class="s2">&quot;is99CredibleSet&quot;</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">











  </div>

    </div>

</div><h2 id="schema">Schema<a class="headerlink" href="#schema" title="Permanent link">&para;</a></h2>
<div class="highlight"><pre><span></span><code>root
 |-- studyLocusId: long (nullable = false)
 |-- variantId: string (nullable = false)
 |-- chromosome: string (nullable = true)
 |-- position: integer (nullable = true)
 |-- region: string (nullable = true)
 |-- studyId: string (nullable = false)
 |-- beta: double (nullable = true)
 |-- zScore: double (nullable = true)
 |-- pValueMantissa: float (nullable = true)
 |-- pValueExponent: integer (nullable = true)
 |-- effectAlleleFrequencyFromSource: float (nullable = true)
 |-- standardError: double (nullable = true)
 |-- subStudyDescription: string (nullable = true)
 |-- qualityControls: array (nullable = true)
 |    |-- element: string (containsNull = false)
 |-- finemappingMethod: string (nullable = true)
 |-- credibleSetIndex: integer (nullable = true)
 |-- credibleSetlog10BF: double (nullable = true)
 |-- purityMeanR2: double (nullable = true)
 |-- purityMinR2: double (nullable = true)
 |-- locusStart: integer (nullable = true)
 |-- locusEnd: integer (nullable = true)
 |-- sampleSize: integer (nullable = true)
 |-- ldSet: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- tagVariantId: string (nullable = true)
 |    |    |-- r2Overall: double (nullable = true)
 |-- locus: array (nullable = true)
 |    |-- element: struct (containsNull = true)
 |    |    |-- is95CredibleSet: boolean (nullable = true)
 |    |    |-- is99CredibleSet: boolean (nullable = true)
 |    |    |-- logBF: double (nullable = true)
 |    |    |-- posteriorProbability: double (nullable = true)
 |    |    |-- variantId: string (nullable = true)
 |    |    |-- pValueMantissa: float (nullable = true)
 |    |    |-- pValueExponent: integer (nullable = true)
 |    |    |-- beta: double (nullable = true)
 |    |    |-- standardError: double (nullable = true)
 |    |    |-- r2Overall: double (nullable = true)
</code></pre></div>
<hr />
<div class="md-source-file">
  <small>
    <ul
      class="metadata page-metadata"
      data-bi-name="page info"
      lang="en-us"
      dir="ltr"
    >
      <!-- mkdocs-git-revision-date-plugin -->
      
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2023-02-10T22:28:52+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-02-10</span> </span>
      </li>
       
      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg>
        </span>
        <span class="label"> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class="timeago" datetime="2024-01-18T11:17:17+00:00" locale="en"></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2024-01-18</span> </span>
      </li>
      

      <li>
        <span class="icon twemoji">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg>
        </span>
        <span class="label"> Contributors </span>
      </li>
      <li class="contributors-holder">
        <ul class="contributors" data-bi-name="contributors"><li>
            <a
              href="https://github.com/d0choa"
              title="d0choa"
              data-bi-name="contributorprofile"
              target="_blank"
              ><img src="https://avatars.githubusercontent.com/u/5097586?v=4" alt="d0choa"
            /></a>
          </li></ul>
      </li>
    </ul>
  </small>
</div>




                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
      
      
    
    <a href="https://github.com/gentropy" target="_blank" rel="noopener" title="github.com" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
        <script src="../../../js/timeago.min.js"></script>
      
        <script src="../../../js/timeago_mkdocs_material.js"></script>
      
        <script src="../../../assets/javascripts/extra.js"></script>
      
    
  </body>
</html>