<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link href=../_method/ rel=prev><link href=../coloc/ rel=next><link rel=icon href=../../../assets/imgs/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-9.1.6"><title>Clumping - Open Targets Genetics</title><link rel=stylesheet href=../../../assets/stylesheets/main.ded33207.min.css><link rel=stylesheet href=../../../assets/stylesheets/palette.a0c5b2b5.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../../css/timeago.css><link rel=stylesheet href=../../../assets/_mkdocstrings.css><link rel=stylesheet href=../../../assets/stylesheets/extra.css><script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#clumping class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class="md-header md-header--shadow md-header--lifted" data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../../.. title="Open Targets Genetics" class="md-header__button md-logo" aria-label="Open Targets Genetics" data-md-component=logo> <img src=../../../assets/imgs/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Open Targets Genetics </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Clumping </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list role=presentation></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class=md-grid> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../../../installation/ class=md-tabs__link> Installation </a> </li> <li class=md-tabs__item> <a href=../../../usage/ class=md-tabs__link> How-to </a> </li> <li class=md-tabs__item> <a href=../../../contributing/_contributing/ class=md-tabs__link> Contributing </a> </li> <li class=md-tabs__item> <a href=../../_python_api/ class="md-tabs__link md-tabs__link--active"> Python API </a> </li> </ul> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../../.. title="Open Targets Genetics" class="md-nav__button md-logo" aria-label="Open Targets Genetics" data-md-component=logo> <img src=../../../assets/imgs/otlogo.png alt=logo> </a> Open Targets Genetics </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../installation/ class=md-nav__link> Installation </a> </li> <li class=md-nav__item> <a href=../../../usage/ class=md-nav__link> How-to </a> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_3> <div class="md-nav__link md-nav__link--index "> <a href=../../../contributing/_contributing/ >Contributing</a> <label for=__nav_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_3_label aria-expanded=false> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Contributing </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../../contributing/guidelines/ class=md-nav__link> Guidelines </a> </li> <li class=md-nav__item> <a href=../../../contributing/troubleshooting/ class=md-nav__link> Troubleshooting </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4 checked> <div class="md-nav__link md-nav__link--index "> <a href=../../_python_api/ >Python API</a> <label for=__nav_4> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=1 aria-labelledby=__nav_4_label aria-expanded=true> <label class=md-nav__title for=__nav_4> <span class="md-nav__icon md-icon"></span> Python API </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_1> <div class="md-nav__link md-nav__link--index "> <a href=../../dataset/_dataset/ >Dataset</a> <label for=__nav_4_1> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_1_label aria-expanded=false> <label class=md-nav__title for=__nav_4_1> <span class="md-nav__icon md-icon"></span> Dataset </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../dataset/colocalisation/ class=md-nav__link> Colocalisation </a> </li> <li class=md-nav__item> <a href=../../dataset/gene_index/ class=md-nav__link> GeneIndex </a> </li> <li class=md-nav__item> <a href=../../dataset/intervals/ class=md-nav__link> Intervals </a> </li> <li class=md-nav__item> <a href=../../dataset/ld_index/ class=md-nav__link> LDIndex </a> </li> <li class=md-nav__item> <a href=../../dataset/study_index/ class=md-nav__link> StudyIndex </a> </li> <li class=md-nav__item> <a href=../../dataset/study_locus/ class=md-nav__link> StudyLocus </a> </li> <li class=md-nav__item> <a href=../../dataset/study_locus_overlap/ class=md-nav__link> StudyLocusOverlap </a> </li> <li class=md-nav__item> <a href=../../dataset/summary_statistics/ class=md-nav__link> SummaryStatistics </a> </li> <li class=md-nav__item> <a href=../../dataset/variant_annotation/ class=md-nav__link> VariantAnnotation </a> </li> <li class=md-nav__item> <a href=../../dataset/variant_index/ class=md-nav__link> VariantIndex </a> </li> <li class=md-nav__item> <a href=../../dataset/variant_to_gene/ class=md-nav__link> V2G </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/_datasource/ >Datasource</a> <label for=__nav_4_2> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_2_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2> <span class="md-nav__icon md-icon"></span> Datasource </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_1> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/finngen/_finngen/ >Finngen</a> <label for=__nav_4_2_1> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_1_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_1> <span class="md-nav__icon md-icon"></span> Finngen </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../datasource/finngen/study_index/ class=md-nav__link> Study index </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_2> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/gnomad/_gnomad/ >Gnomad</a> <label for=__nav_4_2_2> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_2_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_2> <span class="md-nav__icon md-icon"></span> Gnomad </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../datasource/gnomad/gnomad_ld/ class=md-nav__link> Gnomad ld </a> </li> <li class=md-nav__item> <a href=../../datasource/gnomad/gnomad_variants/ class=md-nav__link> Gnomad variants </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_3> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/gwas_catalog/_gwas_catalog/ >Gwas catalog</a> <label for=__nav_4_2_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_3_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_3> <span class="md-nav__icon md-icon"></span> Gwas catalog </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../datasource/gwas_catalog/associations/ class=md-nav__link> Associations </a> </li> <li class=md-nav__item> <a href=../../datasource/gwas_catalog/study_index/ class=md-nav__link> Study index </a> </li> <li class=md-nav__item> <a href=../../datasource/gwas_catalog/study_splitter/ class=md-nav__link> Study splitter </a> </li> <li class=md-nav__item> <a href=../../datasource/gwas_catalog/summary_statistics/ class=md-nav__link> Summary statistics </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_4> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/intervals/_intervals/ >Intervals</a> <label for=__nav_4_2_4> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_4> <span class="md-nav__icon md-icon"></span> Intervals </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../datasource/intervals/andersson/ class=md-nav__link> Andersson </a> </li> <li class=md-nav__item> <a href=../../datasource/intervals/javierre/ class=md-nav__link> Javierre </a> </li> <li class=md-nav__item> <a href=../../datasource/intervals/jung/ class=md-nav__link> Jung </a> </li> <li class=md-nav__item> <a href=../../datasource/intervals/thurman/ class=md-nav__link> Thurman </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_5> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/open_targets/_open_targets/ >Open targets</a> <label for=__nav_4_2_5> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_5_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_5> <span class="md-nav__icon md-icon"></span> Open targets </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../datasource/open_targets/target/ class=md-nav__link> Target </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_2_6> <div class="md-nav__link md-nav__link--index "> <a href=../../datasource/ukbiobank/_ukbiobank/ >Ukbiobank</a> <label for=__nav_4_2_6> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=3 aria-labelledby=__nav_4_2_6_label aria-expanded=false> <label class=md-nav__title for=__nav_4_2_6> <span class="md-nav__icon md-icon"></span> Ukbiobank </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../datasource/ukbiobank/study_index/ class=md-nav__link> Study index </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_3 checked> <div class="md-nav__link md-nav__link--index "> <a href=../_method/ >Method</a> <label for=__nav_4_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_3_label aria-expanded=true> <label class=md-nav__title for=__nav_4_3> <span class="md-nav__icon md-icon"></span> Method </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Clumping <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Clumping </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#clumping-based-on-linkage-disequilibrium-ld class=md-nav__link> Clumping based on Linkage Disequilibrium (LD) </a> </li> <li class=md-nav__item> <a href=#otg.method.clump.LDclumping class=md-nav__link> otg.method.clump.LDclumping </a> </li> <li class=md-nav__item> <a href=#otg.method.clump.LDclumping.clump class=md-nav__link> clump() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../coloc/ class=md-nav__link> coloc </a> </li> <li class=md-nav__item> <a href=../ecaviar/ class=md-nav__link> eCAVIAR </a> </li> <li class=md-nav__item> <a href=../ld_annotator/ class=md-nav__link> LD annotator </a> </li> <li class=md-nav__item> <a href=../pics/ class=md-nav__link> PICS </a> </li> <li class=md-nav__item> <a href=../window_based_clumping/ class=md-nav__link> Window-based clumping </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle " type=checkbox id=__nav_4_4> <div class="md-nav__link md-nav__link--index "> <a href=../../step/_step/ >Step</a> <label for=__nav_4_4> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav data-md-level=2 aria-labelledby=__nav_4_4_label aria-expanded=false> <label class=md-nav__title for=__nav_4_4> <span class="md-nav__icon md-icon"></span> Step </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../step/colocalisation/ class=md-nav__link> Colocalisation </a> </li> <li class=md-nav__item> <a href=../../step/finngen/ class=md-nav__link> FinnGen </a> </li> <li class=md-nav__item> <a href=../../step/gene_index/ class=md-nav__link> Gene index </a> </li> <li class=md-nav__item> <a href=../../step/gwas_catalog/ class=md-nav__link> GWAS Catalog </a> </li> <li class=md-nav__item> <a href=../../step/gwas_catalog_sumstat_preprocess/ class=md-nav__link> GWAS Catalog sumstat preprocess </a> </li> <li class=md-nav__item> <a href=../../step/ld_index/ class=md-nav__link> LD index </a> </li> <li class=md-nav__item> <a href=../../step/ukbiobank/ class=md-nav__link> UKBiobank </a> </li> <li class=md-nav__item> <a href=../../step/variant_annotation_step/ class=md-nav__link> Variant annotation </a> </li> <li class=md-nav__item> <a href=../../step/variant_index_step/ class=md-nav__link> Variant index </a> </li> <li class=md-nav__item> <a href=../../step/variant_to_gene_step/ class=md-nav__link> V2G </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#clumping-based-on-linkage-disequilibrium-ld class=md-nav__link> Clumping based on Linkage Disequilibrium (LD) </a> </li> <li class=md-nav__item> <a href=#otg.method.clump.LDclumping class=md-nav__link> otg.method.clump.LDclumping </a> </li> <li class=md-nav__item> <a href=#otg.method.clump.LDclumping.clump class=md-nav__link> clump() </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <h1 id=clumping>Clumping<a class=headerlink href=#clumping title="Permanent link">&para;</a></h1> <p>Clumping is a commonly used post-processing method that allows for identification of independent association signals from GWAS summary statistics and curated associations. This process is critical because of the complex linkage disequilibrium (LD) structure in human populations, which can result in multiple statistically significant associations within the same genomic region. Clumping methods help reduce redundancy in GWAS results and ensure that each reported association represents an independent signal.</p> <p>We have implemented 2 clumping methods:</p> <h2 id=clumping-based-on-linkage-disequilibrium-ld>Clumping based on Linkage Disequilibrium (LD)<a class=headerlink href=#clumping-based-on-linkage-disequilibrium-ld title="Permanent link">&para;</a></h2> <div class="doc doc-object doc-class"> <a id=otg.method.clump.LDclumping></a> <div class="doc doc-contents first"> <p>LD clumping reports the most significant genetic associations in a region in terms of a smaller number of “clumps” of genetically linked SNPs.</p> <details class=quote> <summary>Source code in <code>src/otg/method/clump.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>17</span>
<span class=normal>18</span>
<span class=normal>19</span>
<span class=normal>20</span>
<span class=normal>21</span>
<span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>class</span> <span class=nc>LDclumping</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;LD clumping reports the most significant genetic associations in a region in terms of a smaller number of “clumps” of genetically linked SNPs.&quot;&quot;&quot;</span>

    <span class=nd>@staticmethod</span>
    <span class=k>def</span> <span class=nf>_is_lead_linked</span><span class=p>(</span>
        <span class=n>study_id</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>variant_id</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_exponent</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>p_value_mantissa</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
        <span class=n>ld_set</span><span class=p>:</span> <span class=n>Column</span><span class=p>,</span>
    <span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Evaluates whether a lead variant is linked to a tag (with lowest p-value) in the same studyLocus dataset.</span>

<span class=sd>        Args:</span>
<span class=sd>            study_id (Column): studyId</span>
<span class=sd>            variant_id (Column): Lead variant id</span>
<span class=sd>            p_value_exponent (Column): p-value exponent</span>
<span class=sd>            p_value_mantissa (Column): p-value mantissa</span>
<span class=sd>            locus (Column): Credible set &lt;array of structs&gt;</span>

<span class=sd>        Returns:</span>
<span class=sd>            Column: Boolean in which True indicates that the lead is linked to another tag in the same dataset.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=n>leads_in_study</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>collect_set</span><span class=p>(</span><span class=n>variant_id</span><span class=p>)</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=n>study_id</span><span class=p>))</span>
        <span class=n>tags_in_studylocus</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>array_union</span><span class=p>(</span>
            <span class=c1># Get all tag variants from the credible set per studyLocusId</span>
            <span class=n>f</span><span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>ld_set</span><span class=p>,</span> <span class=k>lambda</span> <span class=n>x</span><span class=p>:</span> <span class=n>x</span><span class=o>.</span><span class=n>tagVariantId</span><span class=p>),</span>
            <span class=c1># And append the lead variant so that the intersection is the same for all studyLocusIds in a study</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span><span class=n>variant_id</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=n>intersect_lead_tags</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>array_sort</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array_intersect</span><span class=p>(</span><span class=n>leads_in_study</span><span class=p>,</span> <span class=n>tags_in_studylocus</span><span class=p>)</span>
        <span class=p>)</span>
        <span class=k>return</span> <span class=p>(</span>
            <span class=c1># If the lead is in the credible set, we rank the peaks by p-value</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>intersect_lead_tags</span><span class=p>)</span> <span class=o>&gt;</span> <span class=mi>0</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>row_number</span><span class=p>()</span><span class=o>.</span><span class=n>over</span><span class=p>(</span>
                    <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=n>study_id</span><span class=p>,</span> <span class=n>intersect_lead_tags</span><span class=p>)</span><span class=o>.</span><span class=n>orderBy</span><span class=p>(</span>
                        <span class=n>p_value_exponent</span><span class=p>,</span> <span class=n>p_value_mantissa</span>
                    <span class=p>)</span>
                <span class=p>)</span>
                <span class=o>&gt;</span> <span class=mi>1</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># If the intersection is empty (lead is not in the credible set or cred set is empty), the association is not linked</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>False</span><span class=p>))</span>
        <span class=p>)</span>

    <span class=nd>@classmethod</span>
    <span class=k>def</span> <span class=nf>clump</span><span class=p>(</span><span class=bp>cls</span><span class=p>:</span> <span class=nb>type</span><span class=p>[</span><span class=n>LDclumping</span><span class=p>],</span> <span class=n>associations</span><span class=p>:</span> <span class=n>StudyLocus</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>StudyLocus</span><span class=p>:</span>
<span class=w>        </span><span class=sd>&quot;&quot;&quot;Perform clumping on studyLocus dataset.</span>

<span class=sd>        Args:</span>
<span class=sd>            associations (StudyLocus): StudyLocus dataset</span>

<span class=sd>        Returns:</span>
<span class=sd>            StudyLocus: including flag and removing locus information for LD clumped loci.</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>associations</span><span class=o>.</span><span class=n>clump</span><span class=p>()</span>
</code></pre></div></td></tr></table></div> </details> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=otg.method.clump.LDclumping.clump class="doc doc-heading"> <code class="highlight language-python"><span class=n>clump</span><span class=p>(</span><span class=n>associations</span><span class=p>)</span></code> <span class="doc doc-labels"> <small class="doc doc-label doc-label-classmethod"><code>classmethod</code></small> </span> <a href=#otg.method.clump.LDclumping.clump class=headerlink title="Permanent link">&para;</a></h2> <div class="doc doc-contents "> <p>Perform clumping on studyLocus dataset.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>associations</code></td> <td> <code><a class="autorefs autorefs-internal" title="otg.dataset.study_locus.StudyLocus" href="../../dataset/study_locus/#otg.dataset.study_locus.StudyLocus">StudyLocus</a></code> </td> <td><p>StudyLocus dataset</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>StudyLocus</code></td> <td> <code><a class="autorefs autorefs-internal" title="otg.dataset.study_locus.StudyLocus" href="../../dataset/study_locus/#otg.dataset.study_locus.StudyLocus">StudyLocus</a></code> </td> <td><p>including flag and removing locus information for LD clumped loci.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>src/otg/method/clump.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span></pre></div></td><td class=code><div><pre><span></span><code><span class=nd>@classmethod</span>
<span class=k>def</span> <span class=nf>clump</span><span class=p>(</span><span class=bp>cls</span><span class=p>:</span> <span class=nb>type</span><span class=p>[</span><span class=n>LDclumping</span><span class=p>],</span> <span class=n>associations</span><span class=p>:</span> <span class=n>StudyLocus</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>StudyLocus</span><span class=p>:</span>
<span class=w>    </span><span class=sd>&quot;&quot;&quot;Perform clumping on studyLocus dataset.</span>

<span class=sd>    Args:</span>
<span class=sd>        associations (StudyLocus): StudyLocus dataset</span>

<span class=sd>    Returns:</span>
<span class=sd>        StudyLocus: including flag and removing locus information for LD clumped loci.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>associations</span><span class=o>.</span><span class=n>clump</span><span class=p>()</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> <ul class="metadata page-metadata" data-bi-name="page info" lang=en-us dir=ltr> <!-- mkdocs-git-revision-date-plugin --> <li> <span class="icon twemoji"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M14.47 15.08 11 13V7h1.5v5.25l3.08 1.83c-.41.28-.79.62-1.11 1m-1.39 4.84c-.36.05-.71.08-1.08.08-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8c0 .37-.03.72-.08 1.08.69.1 1.33.32 1.92.64.1-.56.16-1.13.16-1.72 0-5.5-4.5-10-10-10S2 6.5 2 12s4.47 10 10 10c.59 0 1.16-.06 1.72-.16-.32-.59-.54-1.23-.64-1.92M18 15v3h-3v2h3v3h2v-3h3v-2h-3v-3h-2Z"/></svg> </span> <span class=label> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class=timeago datetime=2023-07-10T14:20:31+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-07-10</span> </span> </li> <li> <span class="icon twemoji"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M21 13.1c-.1 0-.3.1-.4.2l-1 1 2.1 2.1 1-1c.2-.2.2-.6 0-.8l-1.3-1.3c-.1-.1-.2-.2-.4-.2m-1.9 1.8-6.1 6V23h2.1l6.1-6.1-2.1-2M12.5 7v5.2l4 2.4-1 1L11 13V7h1.5M11 21.9c-5.1-.5-9-4.8-9-9.9C2 6.5 6.5 2 12 2c5.3 0 9.6 4.1 10 9.3-.3-.1-.6-.2-1-.2s-.7.1-1 .2C19.6 7.2 16.2 4 12 4c-4.4 0-8 3.6-8 8 0 4.1 3.1 7.5 7.1 7.9l-.1.2v1.8Z"/></svg> </span> <span class=label> <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-timeago"><span class=timeago datetime=2023-10-16T08:22:37+00:00 locale=en></span></span><span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-iso_date">2023-10-16</span> </span> </li> <li> <span class="icon twemoji"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 2A10 10 0 0 0 2 12c0 4.42 2.87 8.17 6.84 9.5.5.08.66-.23.66-.5v-1.69c-2.77.6-3.36-1.34-3.36-1.34-.46-1.16-1.11-1.47-1.11-1.47-.91-.62.07-.6.07-.6 1 .07 1.53 1.03 1.53 1.03.87 1.52 2.34 1.07 2.91.83.09-.65.35-1.09.63-1.34-2.22-.25-4.55-1.11-4.55-4.92 0-1.11.38-2 1.03-2.71-.1-.25-.45-1.29.1-2.64 0 0 .84-.27 2.75 1.02.79-.22 1.65-.33 2.5-.33.85 0 1.71.11 2.5.33 1.91-1.29 2.75-1.02 2.75-1.02.55 1.35.2 2.39.1 2.64.65.71 1.03 1.6 1.03 2.71 0 3.82-2.34 4.66-4.57 4.91.36.31.69.92.69 1.85V21c0 .27.16.59.67.5C19.14 20.16 22 16.42 22 12A10 10 0 0 0 12 2Z"/></svg> </span> <span class=label> Contributors </span> </li> <li class=contributors-holder> <ul class=contributors data-bi-name=contributors><li><a href=https://github.com/d0choa title=d0choa data-bi-name=contributorprofile target=_blank><img src=https://avatars.githubusercontent.com/u/5097586 alt=d0choa></a></li></ul> </li> </ul> </small> </div> </article> </div> </div> <button type=button class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </button> </main> <footer class=md-footer> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../../..", "features": ["navigation.instant", "navigation.top", "navigation.tabs", "navigation.tabs.sticky", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.208ed371.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script> <script src=../../../assets/javascripts/bundle.51198bba.min.js></script> <script src=../../../js/timeago.min.js></script> <script src=../../../js/timeago_mkdocs_material.js></script> <script src=../../../assets/javascripts/extra.js></script> </body> </html>