<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=../../assets/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-8.5.10"><title>Variants - Genetics Portal Pipeline</title><link rel=stylesheet href=../../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#summary-of-the-logic class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="Genetics Portal Pipeline" class="md-header__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Genetics Portal Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Variants </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="Genetics Portal Pipeline" class="md-nav__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> Genetics Portal Pipeline </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Home </a> </li> <li class=md-nav__item> <a href=../../roadmap/ class=md-nav__link> Roadmap </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <div class="md-nav__link md-nav__link--index "> <a href=../colocalisation/ >Modules</a> <label for=__nav_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav aria-label=Modules data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Modules </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../gwas_ingest/ class=md-nav__link> Gwas ingest </a> </li> <li class=md-nav__item> <a href=../schemas/ class=md-nav__link> Schemas </a> </li> <li class=md-nav__item> <a href=../variant_to_gene/ class=md-nav__link> Variant to gene </a> </li> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Variants <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Variants </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#summary-of-the-logic class=md-nav__link> Summary of the logic </a> <nav class=md-nav aria-label="Summary of the logic"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#variant-annotation class=md-nav__link> Variant annotation </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.variants.variant_annotation class=md-nav__link> etl.variants.variant_annotation </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_annotation.generate_variant_annotation class=md-nav__link> generate_variant_annotation() </a> <nav class=md-nav aria-label=generate_variant_annotation()> <ul class=md-nav__list> <li class=md-nav__item> <a href=#variant-index class=md-nav__link> Variant index </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index class=md-nav__link> etl.variants.variant_index </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index.get_variants_from_credset class=md-nav__link> get_variants_from_credset() </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index.join_variants_w_credset class=md-nav__link> join_variants_w_credset() </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index.read_variant_annotation class=md-nav__link> read_variant_annotation() </a> </li> </ul> </nav> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#summary-of-the-logic class=md-nav__link> Summary of the logic </a> <nav class=md-nav aria-label="Summary of the logic"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#variant-annotation class=md-nav__link> Variant annotation </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.variants.variant_annotation class=md-nav__link> etl.variants.variant_annotation </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_annotation.generate_variant_annotation class=md-nav__link> generate_variant_annotation() </a> <nav class=md-nav aria-label=generate_variant_annotation()> <ul class=md-nav__list> <li class=md-nav__item> <a href=#variant-index class=md-nav__link> Variant index </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index class=md-nav__link> etl.variants.variant_index </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index.get_variants_from_credset class=md-nav__link> get_variants_from_credset() </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index.join_variants_w_credset class=md-nav__link> join_variants_w_credset() </a> </li> <li class=md-nav__item> <a href=#etl.variants.variant_index.read_variant_annotation class=md-nav__link> read_variant_annotation() </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/opentargets/genetics_etl_python/edit/master/docs/modules/variants.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1>Variants</h1> <p>This workflow produces two outputs from the variant dataset of GnomAD.</p> <ol> <li>Variant annotation. The dataset is derived from the GnomAD 3.1 release, with some modification. This dataset is used in other pipelines to generate annotation for all the variants the Portal processes.</li> <li>Variant index. Based on the variant annotation dataset, the dataset has been filtered to only contain variants that have association data.</li> </ol> <p>Schemas for each dataset are defined in the <code>json.schemas</code> module.</p> <h2 id=summary-of-the-logic>Summary of the logic</h2> <h3 id=variant-annotation>Variant annotation</h3> <ol> <li>The variant dataset from GnomAD is processed with Hail to extract relevant information about a variant.</li> <li>The transcript consequences features provided by VEP are filtered to only refer to only refer to the canonical transcript.</li> <li>Genome coordinates are liftovered from GRCh38 to GRCh37.</li> <li>Field names are converted to camel case, to follow the same conventions as other pipelines.</li> </ol> <div class="doc doc-object doc-module"> <a id=etl.variants.variant_annotation></a> <div class="doc doc-contents first"> <p>Step to generate variant annotation dataset.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.variants.variant_annotation.generate_variant_annotation class="doc doc-heading"> <code class="highlight language-python"><span class=n>generate_variant_annotation</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gnomad_variants_path</span><span class=p>,</span> <span class=n>chain_file_path</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Creates a dataset with several annotations derived from GnomAD.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gnomad_variants_path</code></td> <td> <code>str</code> </td> <td><p>Path to the GnomAD variants dataset</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>chain_file_path</code></td> <td> <code>str</code> </td> <td><p>Chain to liftover from grch38 to grch37</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Subset of variant annotations derived from GnomAD</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/variants/variant_annotation.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 32</span>
<span class=normal> 33</span>
<span class=normal> 34</span>
<span class=normal> 35</span>
<span class=normal> 36</span>
<span class=normal> 37</span>
<span class=normal> 38</span>
<span class=normal> 39</span>
<span class=normal> 40</span>
<span class=normal> 41</span>
<span class=normal> 42</span>
<span class=normal> 43</span>
<span class=normal> 44</span>
<span class=normal> 45</span>
<span class=normal> 46</span>
<span class=normal> 47</span>
<span class=normal> 48</span>
<span class=normal> 49</span>
<span class=normal> 50</span>
<span class=normal> 51</span>
<span class=normal> 52</span>
<span class=normal> 53</span>
<span class=normal> 54</span>
<span class=normal> 55</span>
<span class=normal> 56</span>
<span class=normal> 57</span>
<span class=normal> 58</span>
<span class=normal> 59</span>
<span class=normal> 60</span>
<span class=normal> 61</span>
<span class=normal> 62</span>
<span class=normal> 63</span>
<span class=normal> 64</span>
<span class=normal> 65</span>
<span class=normal> 66</span>
<span class=normal> 67</span>
<span class=normal> 68</span>
<span class=normal> 69</span>
<span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>generate_variant_annotation</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>gnomad_variants_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>chain_file_path</span><span class=p>:</span> <span class=nb>str</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Creates a dataset with several annotations derived from GnomAD.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        gnomad_variants_path (str): Path to the GnomAD variants dataset</span>
<span class=sd>        chain_file_path (str): Chain to liftover from grch38 to grch37</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Subset of variant annotations derived from GnomAD</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Generating variant annotation...&quot;</span><span class=p>)</span>

    <span class=n>hl</span><span class=o>.</span><span class=n>init</span><span class=p>(</span><span class=n>sc</span><span class=o>=</span><span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>sparkContext</span><span class=p>)</span>

    <span class=c1># Load variants dataset</span>
    <span class=n>ht</span> <span class=o>=</span> <span class=n>hl</span><span class=o>.</span><span class=n>read_table</span><span class=p>(</span>
        <span class=n>gnomad_variants_path</span><span class=p>,</span>
        <span class=n>_load_refs</span><span class=o>=</span><span class=kc>False</span><span class=p>,</span>
    <span class=p>)</span>
    <span class=c1># Drop non biallelic variants</span>
    <span class=n>ht</span> <span class=o>=</span> <span class=n>ht</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>ht</span><span class=o>.</span><span class=n>alleles</span><span class=o>.</span><span class=n>length</span><span class=p>()</span> <span class=o>==</span> <span class=mi>2</span><span class=p>)</span>

    <span class=c1># Generate struct for alt. allele frequency in selected populations:</span>
    <span class=n>population_indices</span> <span class=o>=</span> <span class=n>ht</span><span class=o>.</span><span class=n>globals</span><span class=o>.</span><span class=n>freq_index_dict</span><span class=o>.</span><span class=n>collect</span><span class=p>()[</span><span class=mi>0</span><span class=p>]</span>
    <span class=n>population_indices</span> <span class=o>=</span> <span class=p>{</span><span class=n>pop</span><span class=p>:</span> <span class=n>population_indices</span><span class=p>[</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>pop</span><span class=si>}</span><span class=s2>-adj&quot;</span><span class=p>]</span> <span class=k>for</span> <span class=n>pop</span> <span class=ow>in</span> <span class=n>POPULATIONS</span><span class=p>}</span>
    <span class=n>ht</span> <span class=o>=</span> <span class=n>ht</span><span class=o>.</span><span class=n>annotate</span><span class=p>(</span>
        <span class=n>alleleFrequenciesRaw</span><span class=o>=</span><span class=n>hl</span><span class=o>.</span><span class=n>struct</span><span class=p>(</span>
            <span class=o>**</span><span class=p>{</span><span class=n>pop</span><span class=p>:</span> <span class=n>ht</span><span class=o>.</span><span class=n>freq</span><span class=p>[</span><span class=n>index</span><span class=p>]</span><span class=o>.</span><span class=n>AF</span> <span class=k>for</span> <span class=n>pop</span><span class=p>,</span> <span class=n>index</span> <span class=ow>in</span> <span class=n>population_indices</span><span class=o>.</span><span class=n>items</span><span class=p>()}</span>
        <span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Liftover</span>
    <span class=n>grch37</span> <span class=o>=</span> <span class=n>hl</span><span class=o>.</span><span class=n>get_reference</span><span class=p>(</span><span class=s2>&quot;GRCh37&quot;</span><span class=p>)</span>
    <span class=n>grch38</span> <span class=o>=</span> <span class=n>hl</span><span class=o>.</span><span class=n>get_reference</span><span class=p>(</span><span class=s2>&quot;GRCh38&quot;</span><span class=p>)</span>
    <span class=n>grch38</span><span class=o>.</span><span class=n>add_liftover</span><span class=p>(</span><span class=n>chain_file_path</span><span class=p>,</span> <span class=n>grch37</span><span class=p>)</span>
    <span class=n>ht</span> <span class=o>=</span> <span class=n>ht</span><span class=o>.</span><span class=n>annotate</span><span class=p>(</span><span class=n>locus_GRCh37</span><span class=o>=</span><span class=n>hl</span><span class=o>.</span><span class=n>liftover</span><span class=p>(</span><span class=n>ht</span><span class=o>.</span><span class=n>locus</span><span class=p>,</span> <span class=s2>&quot;GRCh37&quot;</span><span class=p>))</span>

    <span class=c1># Adding build-specific coordinates to the table:</span>
    <span class=n>ht</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>ht</span><span class=o>.</span><span class=n>annotate</span><span class=p>(</span>
            <span class=n>chromosome</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>locus</span><span class=o>.</span><span class=n>contig</span><span class=o>.</span><span class=n>replace</span><span class=p>(</span><span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span>
            <span class=n>position</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>locus</span><span class=o>.</span><span class=n>position</span><span class=p>,</span>
            <span class=n>chromosomeB37</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>locus_GRCh37</span><span class=o>.</span><span class=n>contig</span><span class=o>.</span><span class=n>replace</span><span class=p>(</span><span class=s2>&quot;chr&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span>
            <span class=n>positionB37</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>locus_GRCh37</span><span class=o>.</span><span class=n>position</span><span class=p>,</span>
            <span class=n>referenceAllele</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>alleles</span><span class=p>[</span><span class=mi>0</span><span class=p>],</span>
            <span class=n>alternateAllele</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>alleles</span><span class=p>[</span><span class=mi>1</span><span class=p>],</span>
            <span class=n>alleleType</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>allele_info</span><span class=o>.</span><span class=n>allele_type</span><span class=p>,</span>
            <span class=n>cadd</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>cadd</span><span class=o>.</span><span class=n>rename</span><span class=p>({</span><span class=s2>&quot;raw_score&quot;</span><span class=p>:</span> <span class=s2>&quot;raw&quot;</span><span class=p>})</span><span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;has_duplicate&quot;</span><span class=p>),</span>
            <span class=n>vepRaw</span><span class=o>=</span><span class=n>ht</span><span class=o>.</span><span class=n>vep</span><span class=o>.</span><span class=n>drop</span><span class=p>(</span>
                <span class=s2>&quot;assembly_name&quot;</span><span class=p>,</span>
                <span class=s2>&quot;allele_string&quot;</span><span class=p>,</span>
                <span class=s2>&quot;ancestral&quot;</span><span class=p>,</span>
                <span class=s2>&quot;context&quot;</span><span class=p>,</span>
                <span class=s2>&quot;end&quot;</span><span class=p>,</span>
                <span class=s2>&quot;id&quot;</span><span class=p>,</span>
                <span class=s2>&quot;input&quot;</span><span class=p>,</span>
                <span class=s2>&quot;intergenic_consequences&quot;</span><span class=p>,</span>
                <span class=s2>&quot;seq_region_name&quot;</span><span class=p>,</span>
                <span class=s2>&quot;start&quot;</span><span class=p>,</span>
                <span class=s2>&quot;strand&quot;</span><span class=p>,</span>
                <span class=s2>&quot;variant_class&quot;</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>rename</span><span class=p>({</span><span class=s2>&quot;rsid&quot;</span><span class=p>:</span> <span class=s2>&quot;rsIds&quot;</span><span class=p>})</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;vep&quot;</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>ht</span><span class=o>.</span><span class=n>select_globals</span><span class=p>()</span>
        <span class=o>.</span><span class=n>to_spark</span><span class=p>(</span><span class=n>flatten</span><span class=o>=</span><span class=kc>False</span><span class=p>)</span>
        <span class=c1># Creating new column based on the transcript_consequences</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;gnomadVariantId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span>
                <span class=s2>&quot;-&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=s2>&quot;position&quot;</span><span class=p>,</span> <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span> <span class=s2>&quot;alternateAllele&quot;</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;ensembl_position&quot;</span><span class=p>,</span>
            <span class=n>convert_gnomad_position_to_ensembl</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span>
                <span class=s2>&quot;_&quot;</span><span class=p>,</span>
                <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
                <span class=s2>&quot;ensembl_position&quot;</span><span class=p>,</span>
                <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
                <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;id&quot;</span><span class=p>),</span>
            <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ensembl_position&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span>
            <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;chromosomeB37&quot;</span><span class=p>,</span>
            <span class=s2>&quot;positionB37&quot;</span><span class=p>,</span>
            <span class=s2>&quot;gnomadVariantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;alleleType&quot;</span><span class=p>,</span>
            <span class=s2>&quot;rsIds&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span>
                <span class=o>*</span><span class=p>[</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>struct</span><span class=p>(</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;alleleFrequenciesRaw.</span><span class=si>{</span><span class=n>pop</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;alleleFrequency&quot;</span><span class=p>),</span>
                        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>pop</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;populationName&quot;</span><span class=p>),</span>
                    <span class=p>)</span>
                    <span class=k>for</span> <span class=n>pop</span> <span class=ow>in</span> <span class=n>POPULATIONS</span>
                <span class=p>]</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;alleleFrequencies&quot;</span><span class=p>),</span>
            <span class=s2>&quot;cadd&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>struct</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;vepRaw.most_severe_consequence&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;mostSevereConsequence&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;vepRaw.motif_feature_consequences&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                    <span class=s2>&quot;motifFeatureConsequences&quot;</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;vepRaw.regulatory_feature_consequences&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                    <span class=s2>&quot;regulatoryFeatureConsequences&quot;</span>
                <span class=p>),</span>
                <span class=c1># Non canonical transcripts and gene IDs other than ensembl are filtered out</span>
                <span class=n>f</span><span class=o>.</span><span class=n>expr</span><span class=p>(</span>
                    <span class=s2>&quot;filter(vepRaw.transcript_consequences, array -&gt; (array.canonical == 1) and (array.gene_symbol_source == &#39;HGNC&#39;))&quot;</span>
                <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;transcriptConsequences&quot;</span><span class=p>),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;vep&quot;</span><span class=p>),</span>
            <span class=s2>&quot;filters&quot;</span><span class=p>,</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div><h3 id=variant-index>Variant index</h3> <ol> <li>The variant annotation dataset is further processed to follow our variant model definition.</li> <li>The dataset is filtered to only include variants that are present in the credible set. The variants in the credible set that are filtered out are written in the invalid variants file.</li> </ol> <div class="doc doc-object doc-module"> <a id=etl.variants.variant_index></a> <div class="doc doc-contents first"> <p>Helper functions to generate the variant index.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.variants.variant_index.get_variants_from_credset class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_variants_from_credset</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>credible_sets_path</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>It reads the credible sets from the given path, extracts the lead and tag variants.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETLSession</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>credible_sets_path</code></td> <td> <code>str</code> </td> <td><p>the path to the credible sets</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe with all variants contained in the credible sets</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/variants/variant_index.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_variants_from_credset</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>credible_sets_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;It reads the credible sets from the given path, extracts the lead and tag variants.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETLSession</span>
<span class=sd>        credible_sets_path (str): the path to the credible sets</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: A dataframe with all variants contained in the credible sets</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>credset</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>parquet</span><span class=p>(</span><span class=n>credible_sets_path</span><span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;leadVariantId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;tagVariantId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;leadVariantId&quot;</span><span class=p>),</span> <span class=s2>&quot;_&quot;</span><span class=p>)[</span><span class=mi>0</span><span class=p>]</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>repartition</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>credset</span><span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span><span class=s2>&quot;leadVariantId as id&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>union</span><span class=p>(</span><span class=n>credset</span><span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span><span class=s2>&quot;tagVariantId as id&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>dropDuplicates</span><span class=p>([</span><span class=s2>&quot;id&quot;</span><span class=p>])</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.variants.variant_index.join_variants_w_credset class="doc doc-heading"> <code class="highlight language-python"><span class=n>join_variants_w_credset</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>,</span> <span class=n>credible_sets_path</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Returns a dataframe with the variants from the credible sets and their annotation, if the variant is in gnomad.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETLSession</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation_path</code></td> <td> <code>str</code> </td> <td><p>The path to the variant annotation file</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>credible_sets_path</code></td> <td> <code>str</code> </td> <td><p>the path to the credible sets file</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>variant_idx</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe with all the variants of interest and their annotation</p></td> </tr> </tbody> </table> <details class=the-join-is-performed-in-a-2-stage-process-to-minimise-data-shuffling> <summary>The join is performed in a 2 stage process to minimise data shuffling</summary> <ul> <li><code>credset</code> is broadcasted to all executors to join it with the variant annotation.</li> <li>left join with <code>credset</code> to bring all the variants of the credible sets.</li> </ul> </details> <details class=quote> <summary>Source code in <code>etl/variants/variant_index.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>16</span>
<span class=normal>17</span>
<span class=normal>18</span>
<span class=normal>19</span>
<span class=normal>20</span>
<span class=normal>21</span>
<span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>join_variants_w_credset</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>variant_annotation_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>credible_sets_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Returns a dataframe with the variants from the credible sets and their annotation, if the variant is in gnomad.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETLSession</span>
<span class=sd>        variant_annotation_path (str): The path to the variant annotation file</span>
<span class=sd>        credible_sets_path (str): the path to the credible sets file</span>

<span class=sd>    Returns:</span>
<span class=sd>        variant_idx (DataFrame): A dataframe with all the variants of interest and their annotation</span>

<span class=sd>    The join is performed in a 2 stage process to minimise data shuffling:</span>
<span class=sd>        - `credset` is broadcasted to all executors to join it with the variant annotation.</span>
<span class=sd>        - left join with `credset` to bring all the variants of the credible sets.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>va</span> <span class=o>=</span> <span class=n>read_variant_annotation</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>)</span>
    <span class=n>credset</span> <span class=o>=</span> <span class=n>get_variants_from_credset</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>credible_sets_path</span><span class=p>)</span>

    <span class=n>credset_gnomad_overlap</span> <span class=o>=</span> <span class=n>va</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>broadcast</span><span class=p>(</span><span class=n>credset</span><span class=p>),</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;id&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=n>credset</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>credset_gnomad_overlap</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;id&quot;</span><span class=p>,</span> <span class=s2>&quot;chromosome&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
    <span class=p>)</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;variantInGnomad&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantInGnomad&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>False</span><span class=p>)))</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.variants.variant_index.read_variant_annotation class="doc doc-heading"> <code class="highlight language-python"><span class=n>read_variant_annotation</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>It reads the variant annotation parquet file and formats it to follow the OTG variant model.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETLSession</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation_path</code></td> <td> <code>str</code> </td> <td><p>path to the variant annotation parquet file</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe of variants and their annotation</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/variants/variant_index.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>read_variant_annotation</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;It reads the variant annotation parquet file and formats it to follow the OTG variant model.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETLSession</span>
<span class=sd>        variant_annotation_path (str): path to the variant annotation parquet file</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: A dataframe of variants and their annotation</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>unchanged_cols</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;id&quot;</span><span class=p>,</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;position&quot;</span><span class=p>,</span>
        <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;chromosomeB37&quot;</span><span class=p>,</span>
        <span class=s2>&quot;positionB37&quot;</span><span class=p>,</span>
        <span class=s2>&quot;alleleType&quot;</span><span class=p>,</span>
        <span class=s2>&quot;alleleFrequencies&quot;</span><span class=p>,</span>
        <span class=s2>&quot;cadd&quot;</span><span class=p>,</span>
    <span class=p>]</span>

    <span class=k>return</span> <span class=n>etl</span><span class=o>.</span><span class=n>read_parquet</span><span class=p>(</span><span class=n>variant_annotation_path</span><span class=p>,</span> <span class=s2>&quot;variant_annotation.json&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=o>*</span><span class=n>unchanged_cols</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;vep.mostSevereConsequence&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;mostSevereConsequence&quot;</span><span class=p>),</span>
        <span class=c1># filters/rsid are arrays that can be empty, in this case we convert them to null</span>
        <span class=n>nullify_empty_array</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;filters&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;filters&quot;</span><span class=p>),</span>
        <span class=n>nullify_empty_array</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIds&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;rsIds&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;variantInGnomad&quot;</span><span class=p>),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">November 1, 2022</span> </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../variant_to_gene/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Variant to gene" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Variant to gene </div> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.instant", "navigation.top", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../../assets/javascripts/bundle.5a2dcb6a.min.js></script> </body> </html>