<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=../../assets/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-8.5.10"><title>Gwas ingest - Genetics Portal Pipeline</title><link rel=stylesheet href=../../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#etl.gwas_ingest class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="Genetics Portal Pipeline" class="md-header__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Genetics Portal Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Gwas ingest </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="Genetics Portal Pipeline" class="md-nav__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> Genetics Portal Pipeline </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Home </a> </li> <li class=md-nav__item> <a href=../../roadmap/ class=md-nav__link> Roadmap </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <div class="md-nav__link md-nav__link--index "> <a href=../colocalisation/ >Modules</a> <label for=__nav_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav aria-label=Modules data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Modules </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Gwas ingest <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Gwas ingest </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#etl.gwas_ingest class=md-nav__link> etl.gwas_ingest </a> <nav class=md-nav aria-label=etl.gwas_ingest> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.gwas_ingest--summary-of-the-logic class=md-nav__link> Summary of the logic </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion class=md-nav__link> etl.gwas_ingest.study_ingestion </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.column2camel_case class=md-nav__link> column2camel_case() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.extract_discovery_sample_sizes class=md-nav__link> extract_discovery_sample_sizes() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.generate_study_table class=md-nav__link> generate_study_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.get_sumstats_location class=md-nav__link> get_sumstats_location() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.ingest_gwas_catalog_studies class=md-nav__link> ingest_gwas_catalog_studies() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_ancestries class=md-nav__link> parse_ancestries() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_efos class=md-nav__link> parse_efos() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.read_study_table class=md-nav__link> read_study_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.spliting_gwas_studies class=md-nav__link> spliting_gwas_studies() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations class=md-nav__link> etl.gwas_ingest.process_associations </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.deconvolute_variants class=md-nav__link> deconvolute_variants() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.generate_association_table class=md-nav__link> generate_association_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.ingest_gwas_catalog_associations class=md-nav__link> ingest_gwas_catalog_associations() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.read_associations_data class=md-nav__link> read_associations_data() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.splitting_association class=md-nav__link> splitting_association() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization class=md-nav__link> etl.gwas_ingest.effect_harmonization </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonize_effect class=md-nav__link> harmonize_effect() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../schemas/ class=md-nav__link> Schemas </a> </li> <li class=md-nav__item> <a href=../variant_to_gene/ class=md-nav__link> Variant to gene </a> </li> <li class=md-nav__item> <a href=../variants/ class=md-nav__link> Variants </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#etl.gwas_ingest class=md-nav__link> etl.gwas_ingest </a> <nav class=md-nav aria-label=etl.gwas_ingest> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.gwas_ingest--summary-of-the-logic class=md-nav__link> Summary of the logic </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion class=md-nav__link> etl.gwas_ingest.study_ingestion </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.column2camel_case class=md-nav__link> column2camel_case() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.extract_discovery_sample_sizes class=md-nav__link> extract_discovery_sample_sizes() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.generate_study_table class=md-nav__link> generate_study_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.get_sumstats_location class=md-nav__link> get_sumstats_location() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.ingest_gwas_catalog_studies class=md-nav__link> ingest_gwas_catalog_studies() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_ancestries class=md-nav__link> parse_ancestries() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_efos class=md-nav__link> parse_efos() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.read_study_table class=md-nav__link> read_study_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.spliting_gwas_studies class=md-nav__link> spliting_gwas_studies() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations class=md-nav__link> etl.gwas_ingest.process_associations </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.deconvolute_variants class=md-nav__link> deconvolute_variants() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.generate_association_table class=md-nav__link> generate_association_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.ingest_gwas_catalog_associations class=md-nav__link> ingest_gwas_catalog_associations() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.read_associations_data class=md-nav__link> read_associations_data() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.splitting_association class=md-nav__link> splitting_association() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization class=md-nav__link> etl.gwas_ingest.effect_harmonization </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonize_effect class=md-nav__link> harmonize_effect() </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/opentargets/genetics_etl_python/edit/master/docs/modules/gwas_ingest.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1>Gwas ingest</h1> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest></a> <div class="doc doc-contents first"> <p>This pipeline ingest the curated GWAS Catalog associations and studies. Prepares top-loci table and study table.</p> <h3 id=etl.gwas_ingest--summary-of-the-logic>Summary of the logic</h3> <ol> <li>Reading association data set. Applying some filters and do basic processing.</li> <li>Map associated variants to GnomAD variants based on the annotated chromosome:position (on GRCh38).</li> <li>Harmonize effect size, calculate Z-score + direction of effect.</li> <li>Read and process study table (parse ancestry, sample size etc).</li> <li>Join study table with associations on <code>study_accession</code>. Split studies when necessary on the basis that one study describes one trait only.</li> <li>Split data into top-loci and study tables and save.</li> </ol> <div class="doc doc-children"> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest.study_ingestion></a> <div class="doc doc-contents first"> <p>GWAS Catalog study ingestion.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.column2camel_case class="doc doc-heading"> <code class="highlight language-python"><span class=n>column2camel_case</span><span class=p>(</span><span class=n>col_name</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>A helper function to convert column names to camel cases.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>col_name</code></td> <td> <code>str</code> </td> <td><p>a single column name</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>str</code></td> <td> <code>str</code> </td> <td><p>spark expression to select and rename the column</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>250</span>
<span class=normal>251</span>
<span class=normal>252</span>
<span class=normal>253</span>
<span class=normal>254</span>
<span class=normal>255</span>
<span class=normal>256</span>
<span class=normal>257</span>
<span class=normal>258</span>
<span class=normal>259</span>
<span class=normal>260</span>
<span class=normal>261</span>
<span class=normal>262</span>
<span class=normal>263</span>
<span class=normal>264</span>
<span class=normal>265</span>
<span class=normal>266</span>
<span class=normal>267</span>
<span class=normal>268</span>
<span class=normal>269</span>
<span class=normal>270</span>
<span class=normal>271</span>
<span class=normal>272</span>
<span class=normal>273</span>
<span class=normal>274</span>
<span class=normal>275</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>column2camel_case</span><span class=p>(</span><span class=n>col_name</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=nb>str</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;A helper function to convert column names to camel cases.</span>

<span class=sd>    Args:</span>
<span class=sd>        col_name (str): a single column name</span>

<span class=sd>    Returns:</span>
<span class=sd>        str: spark expression to select and rename the column</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=k>def</span> <span class=nf>string2camelcase</span><span class=p>(</span><span class=n>col_name</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=nb>str</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Converting a string to camelcase.</span>

<span class=sd>        Args:</span>
<span class=sd>            col_name (str): a random string</span>

<span class=sd>        Returns:</span>
<span class=sd>            str: Camel cased string</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># Removing a bunch of unwanted characters from the column names:</span>
        <span class=n>col_name</span> <span class=o>=</span> <span class=n>re</span><span class=o>.</span><span class=n>sub</span><span class=p>(</span><span class=sa>r</span><span class=s2>&quot;[\/\(\)\-]+&quot;</span><span class=p>,</span> <span class=s2>&quot; &quot;</span><span class=p>,</span> <span class=n>col_name</span><span class=p>)</span>

        <span class=n>first</span><span class=p>,</span> <span class=o>*</span><span class=n>rest</span> <span class=o>=</span> <span class=n>col_name</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=s2>&quot; &quot;</span><span class=p>)</span>
        <span class=k>return</span> <span class=s2>&quot;&quot;</span><span class=o>.</span><span class=n>join</span><span class=p>([</span><span class=n>first</span><span class=o>.</span><span class=n>lower</span><span class=p>(),</span> <span class=o>*</span><span class=nb>map</span><span class=p>(</span><span class=nb>str</span><span class=o>.</span><span class=n>capitalize</span><span class=p>,</span> <span class=n>rest</span><span class=p>)])</span>

    <span class=k>return</span> <span class=sa>f</span><span class=s2>&quot;`</span><span class=si>{</span><span class=n>col_name</span><span class=si>}</span><span class=s2>` as </span><span class=si>{</span><span class=n>string2camelcase</span><span class=p>(</span><span class=n>col_name</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.extract_discovery_sample_sizes class="doc doc-heading"> <code class="highlight language-python"><span class=n>extract_discovery_sample_sizes</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extract the sample size of the discovery stage of the study as annotated in the GWAS Catalog.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>gwas studies table with a column called <code>initialSampleSize</code></p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>df with columns <code>nCases</code>, <code>nControls</code>, and <code>nSamples</code> per <code>studyAccession</code></p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>extract_discovery_sample_sizes</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extract the sample size of the discovery stage of the study as annotated in the GWAS Catalog.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): gwas studies table with a column called `initialSampleSize`</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: df with columns `nCases`, `nControls`, and `nSamples` per `studyAccession`</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode_outer</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleSize&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;,\s+&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;samples&quot;</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Extracting the sample size from the string:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;sampleSize&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_extract</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;samples&quot;</span><span class=p>),</span> <span class=s2>&quot;,&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;[0-9,]+&quot;</span><span class=p>,</span> <span class=mi>0</span>
            <span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=s2>&quot;sampleSize&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;samples&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;cases&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nCases&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;samples&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;controls&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nControls&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Aggregating sample sizes for all ancestries:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=s2>&quot;nCases&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nCases&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=s2>&quot;nControls&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nControls&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nSamples&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.generate_study_table class="doc doc-heading"> <code class="highlight language-python"><span class=n>generate_study_table</span><span class=p>(</span><span class=n>association_study</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extracting studies from the joined study/association table.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>association_study</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame with both associations and studies.</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe with the columns specified in the study_columns list.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>435</span>
<span class=normal>436</span>
<span class=normal>437</span>
<span class=normal>438</span>
<span class=normal>439</span>
<span class=normal>440</span>
<span class=normal>441</span>
<span class=normal>442</span>
<span class=normal>443</span>
<span class=normal>444</span>
<span class=normal>445</span>
<span class=normal>446</span>
<span class=normal>447</span>
<span class=normal>448</span>
<span class=normal>449</span>
<span class=normal>450</span>
<span class=normal>451</span>
<span class=normal>452</span>
<span class=normal>453</span>
<span class=normal>454</span>
<span class=normal>455</span>
<span class=normal>456</span>
<span class=normal>457</span>
<span class=normal>458</span>
<span class=normal>459</span>
<span class=normal>460</span>
<span class=normal>461</span>
<span class=normal>462</span>
<span class=normal>463</span>
<span class=normal>464</span>
<span class=normal>465</span>
<span class=normal>466</span>
<span class=normal>467</span>
<span class=normal>468</span>
<span class=normal>469</span>
<span class=normal>470</span>
<span class=normal>471</span>
<span class=normal>472</span>
<span class=normal>473</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>generate_study_table</span><span class=p>(</span><span class=n>association_study</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extracting studies from the joined study/association table.</span>

<span class=sd>    Args:</span>
<span class=sd>        association_study (DataFrame): DataFrame with both associations and studies.</span>

<span class=sd>    Returns:</span>
<span class=sd>        A dataframe with the columns specified in the study_columns list.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>study_columns</span> <span class=o>=</span> <span class=p>[</span>
        <span class=c1># Study id and type:</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;id&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;gwas&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;type&quot;</span><span class=p>),</span>
        <span class=c1># Publication level information:</span>
        <span class=s2>&quot;pubmedId&quot;</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;firstAuthor&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;publicationFirstAuthor&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;journal&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;publicationJournal&quot;</span><span class=p>),</span>
        <span class=s2>&quot;publicationDate&quot;</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;study&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;publicationTitle&quot;</span><span class=p>),</span>
        <span class=c1># Trait level information:</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;diseaseTrait&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;traitFromSource&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;efos&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;traitFromSourceMappedIds&quot;</span><span class=p>),</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;backgroundEfos&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;backgroundTraitFromSourceMappedIds&quot;</span><span class=p>),</span>
        <span class=c1># Sample descriptions:</span>
        <span class=c1># ancestryInitial</span>
        <span class=c1># ancestryReplication</span>
        <span class=s2>&quot;initialSampleSize&quot;</span><span class=p>,</span>
        <span class=s2>&quot;discoverySamples&quot;</span><span class=p>,</span>
        <span class=s2>&quot;replicationSamples&quot;</span><span class=p>,</span>
        <span class=c1># Sample counts:</span>
        <span class=s2>&quot;nCases&quot;</span><span class=p>,</span>
        <span class=s2>&quot;nControls&quot;</span><span class=p>,</span>
        <span class=s2>&quot;nSamples&quot;</span><span class=p>,</span>
        <span class=c1># Summary stats related info:</span>
        <span class=s2>&quot;hasSumstats&quot;</span><span class=p>,</span>
        <span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>,</span>
    <span class=p>]</span>

    <span class=k>return</span> <span class=n>association_study</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=o>*</span><span class=n>study_columns</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.get_sumstats_location class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_sumstats_location</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>summarystats_list</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Get summary stat locations.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>summarystats_list</code></td> <td> <code>str</code> </td> <td><p>filepath of table listing summary stats</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>dataframe with each GCST with summary stats and its location</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span>
<span class=normal>76</span>
<span class=normal>77</span>
<span class=normal>78</span>
<span class=normal>79</span>
<span class=normal>80</span>
<span class=normal>81</span>
<span class=normal>82</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_sumstats_location</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>summarystats_list</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get summary stat locations.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        summarystats_list (str): filepath of table listing summary stats</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: dataframe with each GCST with summary stats and its location</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>gwas_sumstats_base_uri</span> <span class=o>=</span> <span class=s2>&quot;ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics&quot;</span>

    <span class=n>sumstats</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>summarystats_list</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>False</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>gwas_sumstats_base_uri</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;_c0&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;^\.\/&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_extract</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;\/(GCST\d+)\/&quot;</span><span class=p>,</span> <span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;studyAccession&quot;</span>
            <span class=p>),</span>
            <span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;hasSumstats&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s2>&quot;Number of studies with harmonized summary stats: </span><span class=si>{</span><span class=n>sumstats</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=n>sumstats</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.ingest_gwas_catalog_studies class="doc doc-heading"> <code class="highlight language-python"><span class=n>ingest_gwas_catalog_studies</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>study_file</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>,</span> <span class=n>summary_stats_list</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>This function ingests study level metadata from the GWAS Catalog.</p> <p>The following information is aggregated/extracted: - All publication related information retained. - Mapped measured and background traits parsed. - Flagged if harmonized summary statistics datasets available. - If available, the ftp path to these files presented. - Ancestries from the discovery and replication stages are structured with sample counts. - Case/control counts extracted. - The number of samples with European ancestry extracted.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETLSession</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>study_file</code></td> <td> <code>str</code> </td> <td><p>path to the GWAS Catalog study table v1.0.3.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>ancestry_file</code></td> <td> <code>str</code> </td> <td><p>path to the GWAS Catalog ancestry table.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>summary_stats_list</code></td> <td> <code>str</code> </td> <td><p>path to the GWAS Catalog harmonized summary statistics list.</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Parsed and annotated GWAS Catalog study table.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>366</span>
<span class=normal>367</span>
<span class=normal>368</span>
<span class=normal>369</span>
<span class=normal>370</span>
<span class=normal>371</span>
<span class=normal>372</span>
<span class=normal>373</span>
<span class=normal>374</span>
<span class=normal>375</span>
<span class=normal>376</span>
<span class=normal>377</span>
<span class=normal>378</span>
<span class=normal>379</span>
<span class=normal>380</span>
<span class=normal>381</span>
<span class=normal>382</span>
<span class=normal>383</span>
<span class=normal>384</span>
<span class=normal>385</span>
<span class=normal>386</span>
<span class=normal>387</span>
<span class=normal>388</span>
<span class=normal>389</span>
<span class=normal>390</span>
<span class=normal>391</span>
<span class=normal>392</span>
<span class=normal>393</span>
<span class=normal>394</span>
<span class=normal>395</span>
<span class=normal>396</span>
<span class=normal>397</span>
<span class=normal>398</span>
<span class=normal>399</span>
<span class=normal>400</span>
<span class=normal>401</span>
<span class=normal>402</span>
<span class=normal>403</span>
<span class=normal>404</span>
<span class=normal>405</span>
<span class=normal>406</span>
<span class=normal>407</span>
<span class=normal>408</span>
<span class=normal>409</span>
<span class=normal>410</span>
<span class=normal>411</span>
<span class=normal>412</span>
<span class=normal>413</span>
<span class=normal>414</span>
<span class=normal>415</span>
<span class=normal>416</span>
<span class=normal>417</span>
<span class=normal>418</span>
<span class=normal>419</span>
<span class=normal>420</span>
<span class=normal>421</span>
<span class=normal>422</span>
<span class=normal>423</span>
<span class=normal>424</span>
<span class=normal>425</span>
<span class=normal>426</span>
<span class=normal>427</span>
<span class=normal>428</span>
<span class=normal>429</span>
<span class=normal>430</span>
<span class=normal>431</span>
<span class=normal>432</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>ingest_gwas_catalog_studies</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>study_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>summary_stats_list</span><span class=p>:</span> <span class=nb>str</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;This function ingests study level metadata from the GWAS Catalog.</span>

<span class=sd>    The following information is aggregated/extracted:</span>
<span class=sd>    - All publication related information retained.</span>
<span class=sd>    - Mapped measured and background traits parsed.</span>
<span class=sd>    - Flagged if harmonized summary statistics datasets available.</span>
<span class=sd>    - If available, the ftp path to these files presented.</span>
<span class=sd>    - Ancestries from the discovery and replication stages are structured with sample counts.</span>
<span class=sd>    - Case/control counts extracted.</span>
<span class=sd>    - The number of samples with European ancestry extracted.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETLSession</span>
<span class=sd>        study_file (str): path to the GWAS Catalog study table v1.0.3.</span>
<span class=sd>        ancestry_file (str): path to the GWAS Catalog ancestry table.</span>
<span class=sd>        summary_stats_list (str): path to the GWAS Catalog harmonized summary statistics list.</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Parsed and annotated GWAS Catalog study table.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Read GWAS Catalogue raw data</span>
    <span class=n>gwas_studies</span> <span class=o>=</span> <span class=n>read_study_table</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>study_file</span><span class=p>)</span>
    <span class=n>gwas_ancestries</span> <span class=o>=</span> <span class=n>parse_ancestries</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>)</span>
    <span class=n>ss_studies</span> <span class=o>=</span> <span class=n>get_sumstats_location</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>summary_stats_list</span><span class=p>)</span>

    <span class=c1># Sample size extraction is a separate process:</span>
    <span class=n>study_size_df</span> <span class=o>=</span> <span class=n>extract_discovery_sample_sizes</span><span class=p>(</span><span class=n>gwas_studies</span><span class=p>)</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>gwas_studies</span>
        <span class=c1># Add study sizes:</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>study_size_df</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>
        <span class=c1># Adding summary stats location:</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>ss_studies</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;hasSumstats&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;hasSumstats&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>False</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>gwas_ancestries</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>
        <span class=c1># Select relevant columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=c1># Publication related fields:</span>
            <span class=s2>&quot;pubmedId&quot;</span><span class=p>,</span>
            <span class=s2>&quot;firstAuthor&quot;</span><span class=p>,</span>
            <span class=s2>&quot;publicationDate&quot;</span><span class=p>,</span>
            <span class=s2>&quot;journal&quot;</span><span class=p>,</span>
            <span class=s2>&quot;study&quot;</span><span class=p>,</span>
            <span class=c1># Disease related fields:</span>
            <span class=s2>&quot;studyDiseaseTrait&quot;</span><span class=p>,</span>
            <span class=n>parse_efos</span><span class=p>(</span><span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;studyEfos&quot;</span><span class=p>),</span>
            <span class=n>parse_efos</span><span class=p>(</span><span class=s2>&quot;mappedBackgroundTraitUri&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;backgroundEfos&quot;</span><span class=p>),</span>
            <span class=c1># Sample related fields:</span>
            <span class=s2>&quot;initialSampleSize&quot;</span><span class=p>,</span>
            <span class=s2>&quot;nCases&quot;</span><span class=p>,</span>
            <span class=s2>&quot;nControls&quot;</span><span class=p>,</span>
            <span class=s2>&quot;nSamples&quot;</span><span class=p>,</span>
            <span class=c1># Ancestry related labels:</span>
            <span class=s2>&quot;discoverySamples&quot;</span><span class=p>,</span>
            <span class=s2>&quot;replicationSamples&quot;</span><span class=p>,</span>
            <span class=c1># &quot;gnomadSamples&quot;,</span>
            <span class=c1># Summary stats fields:</span>
            <span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>,</span>
            <span class=s2>&quot;hasSumstats&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.parse_ancestries class="doc doc-heading"> <code class="highlight language-python"><span class=n>parse_ancestries</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extracting sample sizes and ancestry information.</p> <p>This function parses the ancestry data. Also get counts for the europeans in the same discovery stage.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>ancestry_file</code></td> <td> <code>str</code> </td> <td><p>File name of the ancestry table as downloaded from the GWAS Catalog</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Slimmed and cleaned version of the ancestry annotation.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>278</span>
<span class=normal>279</span>
<span class=normal>280</span>
<span class=normal>281</span>
<span class=normal>282</span>
<span class=normal>283</span>
<span class=normal>284</span>
<span class=normal>285</span>
<span class=normal>286</span>
<span class=normal>287</span>
<span class=normal>288</span>
<span class=normal>289</span>
<span class=normal>290</span>
<span class=normal>291</span>
<span class=normal>292</span>
<span class=normal>293</span>
<span class=normal>294</span>
<span class=normal>295</span>
<span class=normal>296</span>
<span class=normal>297</span>
<span class=normal>298</span>
<span class=normal>299</span>
<span class=normal>300</span>
<span class=normal>301</span>
<span class=normal>302</span>
<span class=normal>303</span>
<span class=normal>304</span>
<span class=normal>305</span>
<span class=normal>306</span>
<span class=normal>307</span>
<span class=normal>308</span>
<span class=normal>309</span>
<span class=normal>310</span>
<span class=normal>311</span>
<span class=normal>312</span>
<span class=normal>313</span>
<span class=normal>314</span>
<span class=normal>315</span>
<span class=normal>316</span>
<span class=normal>317</span>
<span class=normal>318</span>
<span class=normal>319</span>
<span class=normal>320</span>
<span class=normal>321</span>
<span class=normal>322</span>
<span class=normal>323</span>
<span class=normal>324</span>
<span class=normal>325</span>
<span class=normal>326</span>
<span class=normal>327</span>
<span class=normal>328</span>
<span class=normal>329</span>
<span class=normal>330</span>
<span class=normal>331</span>
<span class=normal>332</span>
<span class=normal>333</span>
<span class=normal>334</span>
<span class=normal>335</span>
<span class=normal>336</span>
<span class=normal>337</span>
<span class=normal>338</span>
<span class=normal>339</span>
<span class=normal>340</span>
<span class=normal>341</span>
<span class=normal>342</span>
<span class=normal>343</span>
<span class=normal>344</span>
<span class=normal>345</span>
<span class=normal>346</span>
<span class=normal>347</span>
<span class=normal>348</span>
<span class=normal>349</span>
<span class=normal>350</span>
<span class=normal>351</span>
<span class=normal>352</span>
<span class=normal>353</span>
<span class=normal>354</span>
<span class=normal>355</span>
<span class=normal>356</span>
<span class=normal>357</span>
<span class=normal>358</span>
<span class=normal>359</span>
<span class=normal>360</span>
<span class=normal>361</span>
<span class=normal>362</span>
<span class=normal>363</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>parse_ancestries</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extracting sample sizes and ancestry information.</span>

<span class=sd>    This function parses the ancestry data. Also get counts for the europeans in the same</span>
<span class=sd>    discovery stage.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        ancestry_file (str): File name of the ancestry table as downloaded from the GWAS Catalog</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Slimmed and cleaned version of the ancestry annotation.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Reading ancestry data:</span>
    <span class=n>ancestry</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>ancestry_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
        <span class=c1># Convert column headers to camelcase:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
            <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=o>*</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>expr</span><span class=p>(</span><span class=n>column2camel_case</span><span class=p>(</span><span class=n>x</span><span class=p>))</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=n>df</span><span class=o>.</span><span class=n>columns</span><span class=p>])</span>
        <span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Get a high resolution dataset on experimental stage:</span>
    <span class=n>ancestry_stages</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>ancestry</span><span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>pivot</span><span class=p>(</span><span class=s2>&quot;stage&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>collect_set</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>struct</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;numberOfIndividuals&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;broadAncestralCategory&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;ancestry&quot;</span><span class=p>),</span>
                <span class=p>)</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;initial&quot;</span><span class=p>,</span> <span class=s2>&quot;discoverySamples&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;replication&quot;</span><span class=p>,</span> <span class=s2>&quot;replicationSamples&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=c1># Generate information on the ancestry composition of the discovery stage, and calculate</span>
    <span class=c1># the proportion of the Europeans:</span>
    <span class=n>europeans_deconvoluted</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>ancestry</span>
        <span class=c1># Focus on discovery stage:</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;stage&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;initial&quot;</span><span class=p>)</span>
        <span class=c1># Sorting ancestries if European:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;ancestryFlag&quot;</span><span class=p>,</span>
            <span class=c1># Excluding finnish:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleDescription&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;Finnish&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span>
            <span class=p>)</span>
            <span class=c1># Excluding Icelandic population:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleDescription&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;Icelandic&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span>
            <span class=p>)</span>
            <span class=c1># Including European ancestry:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;broadAncestralCategory&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;European&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>))</span>
            <span class=c1># Exclude all other population:</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Grouping by study accession and initial sample description:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>pivot</span><span class=p>(</span><span class=s2>&quot;ancestryFlag&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=c1># Summarizing sample sizes for all ancestries:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;numberOfIndividuals&quot;</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=c1># Do aritmetics to make sure we have the right proportion of european in the set:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;initialSampleCountEuropean&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;initialSampleCountOther&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;initialSampleCount&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleCountEuropean&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>,</span> <span class=s2>&quot;other&quot;</span><span class=p>,</span> <span class=s2>&quot;initialSampleCountOther&quot;</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>ancestry_stages</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>europeans_deconvoluted</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;outer&quot;</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.parse_efos class="doc doc-heading"> <code class="highlight language-python"><span class=n>parse_efos</span><span class=p>(</span><span class=n>col_name</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extracting EFO identifiers.</p> <p>This function parses EFO identifiers from a comma-separated list of EFO URIs.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>col_name</code></td> <td> <code>str</code> </td> <td><p>name of column with a list of EFO IDs</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>Column</code></td> <td> <code><span title="pyspark.sql.Column">Column</span></code> </td> <td><p>column with a sorted list of parsed EFO IDs</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>236</span>
<span class=normal>237</span>
<span class=normal>238</span>
<span class=normal>239</span>
<span class=normal>240</span>
<span class=normal>241</span>
<span class=normal>242</span>
<span class=normal>243</span>
<span class=normal>244</span>
<span class=normal>245</span>
<span class=normal>246</span>
<span class=normal>247</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>parse_efos</span><span class=p>(</span><span class=n>col_name</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extracting EFO identifiers.</span>

<span class=sd>    This function parses EFO identifiers from a comma-separated list of EFO URIs.</span>

<span class=sd>    Args:</span>
<span class=sd>        col_name (str): name of column with a list of EFO IDs</span>

<span class=sd>    Returns:</span>
<span class=sd>        Column: column with a sorted list of parsed EFO IDs</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>array_sort</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>expr</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;regexp_extract_all(</span><span class=si>{</span><span class=n>col_name</span><span class=si>}</span><span class=s2>, &#39;([A-Z]+_[0-9]+)&#39;)&quot;</span><span class=p>))</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.read_study_table class="doc doc-heading"> <code class="highlight language-python"><span class=n>read_study_table</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_study_file</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Read GWASCatalog study table.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gwas_study_file</code></td> <td> <code>str</code> </td> <td><p>GWAS studies filepath</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Study table.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>read_study_table</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>gwas_study_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Read GWASCatalog study table.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        gwas_study_file (str): GWAS studies filepath</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Study table.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>gwas_study_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=o>*</span><span class=p>[</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>old_name</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=n>new_name</span><span class=p>)</span>
            <span class=k>for</span> <span class=n>old_name</span><span class=p>,</span> <span class=n>new_name</span> <span class=ow>in</span> <span class=n>STUDY_COLUMNS_MAP</span><span class=o>.</span><span class=n>items</span><span class=p>()</span>
        <span class=p>]</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.spliting_gwas_studies class="doc doc-heading"> <code class="highlight language-python"><span class=n>spliting_gwas_studies</span><span class=p>(</span><span class=n>study_association</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Splitting studies and consolidating disease annotation.</p> <p>Processing disease annotation of the joined study/association table. If assigned disease of the study and the association don't agree, we assume the study needs to be split. Then disease EFOs, trait names and study ID are consolidated</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>study_association</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe with the studyAccession, studyId, diseaseTrait, and efos columns.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span>
<span class=normal>194</span>
<span class=normal>195</span>
<span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span>
<span class=normal>209</span>
<span class=normal>210</span>
<span class=normal>211</span>
<span class=normal>212</span>
<span class=normal>213</span>
<span class=normal>214</span>
<span class=normal>215</span>
<span class=normal>216</span>
<span class=normal>217</span>
<span class=normal>218</span>
<span class=normal>219</span>
<span class=normal>220</span>
<span class=normal>221</span>
<span class=normal>222</span>
<span class=normal>223</span>
<span class=normal>224</span>
<span class=normal>225</span>
<span class=normal>226</span>
<span class=normal>227</span>
<span class=normal>228</span>
<span class=normal>229</span>
<span class=normal>230</span>
<span class=normal>231</span>
<span class=normal>232</span>
<span class=normal>233</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>spliting_gwas_studies</span><span class=p>(</span><span class=n>study_association</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Splitting studies and consolidating disease annotation.</span>

<span class=sd>    Processing disease annotation of the joined study/association table. If assigned disease</span>
<span class=sd>    of the study and the association don&#39;t agree, we assume the study needs to be split.</span>
<span class=sd>    Then disease EFOs, trait names and study ID are consolidated</span>

<span class=sd>    Args:</span>
<span class=sd>        study_association (DataFrame): DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        A dataframe with the studyAccession, studyId, diseaseTrait, and efos columns.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># A window to aid study splitting:</span>
    <span class=n>study_split_window</span> <span class=o>=</span> <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>orderBy</span><span class=p>(</span><span class=s2>&quot;splitField&quot;</span><span class=p>)</span>

    <span class=c1># A window to detect ambiguous associations:</span>
    <span class=n>assoc_ambiguity_window</span> <span class=o>=</span> <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=s2>&quot;variantId&quot;</span><span class=p>)</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>study_association</span>
        <span class=c1># As some studies needs to be split by not only the p-value text, but the EFO as well, we need to do this thing:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;splitField&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span>
                <span class=s2>&quot;_&quot;</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pValueText&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array_join</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;associationEfos&quot;</span><span class=p>),</span> <span class=s2>&quot;_&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Windowing over the groups:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;row_number&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>dense_rank</span><span class=p>()</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>study_split_window</span><span class=p>)</span> <span class=o>-</span> <span class=mi>1</span><span class=p>)</span>
        <span class=c1># Study identifiers are split when there are more than one type of associationEfos:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;studyId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;row_number&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=mi>0</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span><span class=s2>&quot;_&quot;</span><span class=p>,</span> <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=s2>&quot;row_number&quot;</span><span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Disese traits are generated based on p-value text when splitting study:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;diseaseTrait&quot;</span><span class=p>,</span>
            <span class=c1># When study is split:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;row_number&quot;</span><span class=p>)</span> <span class=o>!=</span> <span class=mi>0</span><span class=p>,</span>
                <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pValueText&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>(),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>concat</span><span class=p>(</span>
                        <span class=s2>&quot;associationDiseaseTrait&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot; [&quot;</span><span class=p>),</span> <span class=s2>&quot;pValueText&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;]&quot;</span><span class=p>)</span>
                    <span class=p>),</span>
                <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=s2>&quot;associationDiseaseTrait&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=c1># When there&#39;s association disease trait:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;associationDiseaseTrait&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>(),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;associationDiseaseTrait&quot;</span><span class=p>),</span>
            <span class=p>)</span>
            <span class=c1># When no association disease trait is present we get from study:</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyDiseaseTrait&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=c1># The EFO field is also consolidated:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;efos&quot;</span><span class=p>,</span>
            <span class=c1># When available, EFOs are pulled from associations:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;associationEfos&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>(),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;associationEfos&quot;</span><span class=p>))</span>
            <span class=c1># When no association is given, the study level EFOs are used:</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;studyEfos&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=c1># Flagging ambiguous associations:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
            <span class=n>adding_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>),</span>
                <span class=c1># There are more than one variant ID in one study:</span>
                <span class=n>f</span><span class=o>.</span><span class=n>count</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>assoc_ambiguity_window</span><span class=p>)</span> <span class=o>&gt;</span> <span class=mi>1</span><span class=p>,</span>
                <span class=n>AMBIGUOUS_ASSOCIATION</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span><span class=o>.</span><span class=n>drop</span><span class=p>(</span>
            <span class=s2>&quot;row_number&quot;</span><span class=p>,</span>
            <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=s2>&quot;studyEfos&quot;</span><span class=p>,</span>
            <span class=s2>&quot;studyDiseaseTrait&quot;</span><span class=p>,</span>
            <span class=s2>&quot;associationEfos&quot;</span><span class=p>,</span>
            <span class=s2>&quot;associationDiseaseTrait&quot;</span><span class=p>,</span>
            <span class=s2>&quot;pValueText&quot;</span><span class=p>,</span>
            <span class=c1># &#39;full_description&#39;</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest.process_associations></a> <div class="doc doc-contents first"> <p>Process GWAS catalog associations.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.deconvolute_variants class="doc doc-heading"> <code class="highlight language-python"><span class=n>deconvolute_variants</span><span class=p>(</span><span class=n>associations</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Deconvoluting the list of variants attached to GWAS associations.</p> <p>We split the variant notation (chr, pos, snp id) into arrays, and flag associations where the number of genomic location is not consistent with the number of rsids</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>associations</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Flagged associations with inconsistent mappings.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>221</span>
<span class=normal>222</span>
<span class=normal>223</span>
<span class=normal>224</span>
<span class=normal>225</span>
<span class=normal>226</span>
<span class=normal>227</span>
<span class=normal>228</span>
<span class=normal>229</span>
<span class=normal>230</span>
<span class=normal>231</span>
<span class=normal>232</span>
<span class=normal>233</span>
<span class=normal>234</span>
<span class=normal>235</span>
<span class=normal>236</span>
<span class=normal>237</span>
<span class=normal>238</span>
<span class=normal>239</span>
<span class=normal>240</span>
<span class=normal>241</span>
<span class=normal>242</span>
<span class=normal>243</span>
<span class=normal>244</span>
<span class=normal>245</span>
<span class=normal>246</span>
<span class=normal>247</span>
<span class=normal>248</span>
<span class=normal>249</span>
<span class=normal>250</span>
<span class=normal>251</span>
<span class=normal>252</span>
<span class=normal>253</span>
<span class=normal>254</span>
<span class=normal>255</span>
<span class=normal>256</span>
<span class=normal>257</span>
<span class=normal>258</span>
<span class=normal>259</span>
<span class=normal>260</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>deconvolute_variants</span><span class=p>(</span><span class=n>associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Deconvoluting the list of variants attached to GWAS associations.</span>

<span class=sd>    We split the variant notation (chr, pos, snp id) into arrays, and flag associations where the</span>
<span class=sd>    number of genomic location is not consistent with the number of rsids</span>

<span class=sd>    Args:</span>
<span class=sd>        associations (DataFrame): DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Flagged associations with inconsistent mappings.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>associations</span>
        <span class=c1># For further tracking, we save the variant of the association before splitting:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;gwasVariant&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>))</span>
        <span class=c1># Variant notations (chr, pos, snp id) are split into arrays:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span> <span class=s2>&quot;; &quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>),</span> <span class=s2>&quot;; &quot;</span><span class=p>))</span>
        <span class=c1># Flagging associations where the genomic location is not consistent with rsids:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
            <span class=n>adding_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>),</span>
                <span class=c1># Number of chromosome values different from position values:</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>))</span> <span class=o>!=</span> <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>)))</span> <span class=o>|</span>
                <span class=c1># NUmber of chromosome values different from riskAllele values:</span>
                <span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>))</span>
                    <span class=o>!=</span> <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>))</span>
                <span class=p>),</span>
                <span class=n>INCONSISTENCY_FLAG</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.generate_association_table class="doc doc-heading"> <code class="highlight language-python"><span class=n>generate_association_table</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Generating top-loci table.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A DataFrame with the columns specified in the filter_columns list.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>403</span>
<span class=normal>404</span>
<span class=normal>405</span>
<span class=normal>406</span>
<span class=normal>407</span>
<span class=normal>408</span>
<span class=normal>409</span>
<span class=normal>410</span>
<span class=normal>411</span>
<span class=normal>412</span>
<span class=normal>413</span>
<span class=normal>414</span>
<span class=normal>415</span>
<span class=normal>416</span>
<span class=normal>417</span>
<span class=normal>418</span>
<span class=normal>419</span>
<span class=normal>420</span>
<span class=normal>421</span>
<span class=normal>422</span>
<span class=normal>423</span>
<span class=normal>424</span>
<span class=normal>425</span>
<span class=normal>426</span>
<span class=normal>427</span>
<span class=normal>428</span>
<span class=normal>429</span>
<span class=normal>430</span>
<span class=normal>431</span>
<span class=normal>432</span>
<span class=normal>433</span>
<span class=normal>434</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>generate_association_table</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Generating top-loci table.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        A DataFrame with the columns specified in the filter_columns list.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>filter_columns</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;position&quot;</span><span class=p>,</span>
        <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;variantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;studyId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;pValueMantissa&quot;</span><span class=p>,</span>
        <span class=s2>&quot;pValueExponent&quot;</span><span class=p>,</span>
        <span class=s2>&quot;beta&quot;</span><span class=p>,</span>
        <span class=s2>&quot;beta_ci_lower&quot;</span><span class=p>,</span>
        <span class=s2>&quot;beta_ci_upper&quot;</span><span class=p>,</span>
        <span class=s2>&quot;odds_ratio&quot;</span><span class=p>,</span>
        <span class=s2>&quot;odds_ratio_ci_lower&quot;</span><span class=p>,</span>
        <span class=s2>&quot;odds_ratio_ci_upper&quot;</span><span class=p>,</span>
        <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
    <span class=p>]</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=o>*</span><span class=n>filter_columns</span><span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span>
        <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.ingest_gwas_catalog_associations class="doc doc-heading"> <code class="highlight language-python"><span class=n>ingest_gwas_catalog_associations</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_association_path</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Ingest/process/map GWAS Catalog association.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gwas_association_path</code></td> <td> <code>str</code> </td> <td><p>GWAS catalogue dataset path</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation_path</code></td> <td> <code>str</code> </td> <td><p>variant annotation dataset path</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>pvalue_cutoff</code></td> <td> <code>float</code> </td> <td><p>GWAS significance threshold, flagging sub-significant associations</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Post-processed GWASCatalog associations</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>368</span>
<span class=normal>369</span>
<span class=normal>370</span>
<span class=normal>371</span>
<span class=normal>372</span>
<span class=normal>373</span>
<span class=normal>374</span>
<span class=normal>375</span>
<span class=normal>376</span>
<span class=normal>377</span>
<span class=normal>378</span>
<span class=normal>379</span>
<span class=normal>380</span>
<span class=normal>381</span>
<span class=normal>382</span>
<span class=normal>383</span>
<span class=normal>384</span>
<span class=normal>385</span>
<span class=normal>386</span>
<span class=normal>387</span>
<span class=normal>388</span>
<span class=normal>389</span>
<span class=normal>390</span>
<span class=normal>391</span>
<span class=normal>392</span>
<span class=normal>393</span>
<span class=normal>394</span>
<span class=normal>395</span>
<span class=normal>396</span>
<span class=normal>397</span>
<span class=normal>398</span>
<span class=normal>399</span>
<span class=normal>400</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>ingest_gwas_catalog_associations</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>gwas_association_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>variant_annotation_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>pvalue_cutoff</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Ingest/process/map GWAS Catalog association.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        gwas_association_path (str): GWAS catalogue dataset path</span>
<span class=sd>        variant_annotation_path (str): variant annotation dataset path</span>
<span class=sd>        pvalue_cutoff (float): GWAS significance threshold, flagging sub-significant associations</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Post-processed GWASCatalog associations</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=c1># Read associations:</span>
        <span class=n>read_associations_data</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_association_path</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>)</span>
        <span class=c1># Cleaning p-value text:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>_pvalue_text_resolve</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>etl</span><span class=p>))</span>
        <span class=c1># Parsing variants:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>deconvolute_variants</span><span class=p>)</span>
        <span class=c1># Splitting associations by GWAS Catalog variants:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>splitting_association</span><span class=p>)</span>
        <span class=c1># Mapping variants to GnomAD3:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>map_variants</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>,</span> <span class=n>etl</span><span class=p>))</span>
        <span class=c1># Cleaning GnomAD variant mappings to ensure no duplication happens:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>clean_mappings</span><span class=p>)</span>
        <span class=c1># Harmonizing association effect:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>harmonize_effect</span><span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.read_associations_data class="doc doc-heading"> <code class="highlight language-python"><span class=n>read_associations_data</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_association_file</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Read GWASCatalog associations.</p> <p>It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and flags associations that do not reach GWAS significance level.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gwas_association_file</code></td> <td> <code>str</code> </td> <td><p>path to association file CSV</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>pvalue_cutoff</code></td> <td> <code>float</code> </td> <td><p>association p-value cut-off</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p><code>DataFrame</code> with the GWAS Catalog associations</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span>
<span class=normal>194</span>
<span class=normal>195</span>
<span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span>
<span class=normal>209</span>
<span class=normal>210</span>
<span class=normal>211</span>
<span class=normal>212</span>
<span class=normal>213</span>
<span class=normal>214</span>
<span class=normal>215</span>
<span class=normal>216</span>
<span class=normal>217</span>
<span class=normal>218</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>read_associations_data</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>gwas_association_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>:</span> <span class=nb>float</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Read GWASCatalog associations.</span>

<span class=sd>    It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and</span>
<span class=sd>    flags associations that do not reach GWAS significance level.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        gwas_association_file (str): path to association file CSV</span>
<span class=sd>        pvalue_cutoff (float): association p-value cut-off</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: `DataFrame` with the GWAS Catalog associations</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Starting ingesting GWAS Catalog associations...&quot;</span><span class=p>)</span>

    <span class=c1># Reading and filtering associations:</span>
    <span class=n>association_df</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>gwas_association_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
        <span class=c1># Select and rename columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=o>*</span><span class=p>[</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>old_name</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=n>new_name</span><span class=p>)</span>
                <span class=k>for</span> <span class=n>old_name</span><span class=p>,</span> <span class=n>new_name</span> <span class=ow>in</span> <span class=n>ASSOCIATION_COLUMNS_MAP</span><span class=o>.</span><span class=n>items</span><span class=p>()</span>
            <span class=p>],</span>
            <span class=n>f</span><span class=o>.</span><span class=n>monotonically_increasing_id</span><span class=p>()</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;associationId&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Based on pre-defined filters set flag for failing associations:</span>
        <span class=c1># 1. Flagging associations based on variant x variant interactions</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
            <span class=n>adding_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pValueNegLog&quot;</span><span class=p>)</span> <span class=o>&lt;</span> <span class=o>-</span><span class=n>np</span><span class=o>.</span><span class=n>log10</span><span class=p>(</span><span class=n>pvalue_cutoff</span><span class=p>),</span>
                <span class=n>SUBSIGNIFICANT_FLAG</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># 2. Flagging sub-significant associations</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
            <span class=n>adding_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>()</span> <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span>
                <span class=n>INCOMPLETE_MAPPING_FLAG</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># 3. Flagging associations without genomic location</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
            <span class=n>adding_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot; x &quot;</span><span class=p>),</span>
                <span class=n>COMPOSITE_FLAG</span><span class=p>,</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Parsing association EFO:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;associationEfos&quot;</span><span class=p>,</span> <span class=n>parse_efos</span><span class=p>(</span><span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=c1># Providing stats on the filtered association dataset:</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Number of associations: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of studies: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of variants: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of failed associations: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>))</span> <span class=o>!=</span> <span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>association_df</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.splitting_association class="doc doc-heading"> <code class="highlight language-python"><span class=n>splitting_association</span><span class=p>(</span><span class=n>association</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Splitting associations based on the list of parseable variants from the GWAS Catalog.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>association</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe with the same columns as the input dataframe, but with the variants exploded.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>288</span>
<span class=normal>289</span>
<span class=normal>290</span>
<span class=normal>291</span>
<span class=normal>292</span>
<span class=normal>293</span>
<span class=normal>294</span>
<span class=normal>295</span>
<span class=normal>296</span>
<span class=normal>297</span>
<span class=normal>298</span>
<span class=normal>299</span>
<span class=normal>300</span>
<span class=normal>301</span>
<span class=normal>302</span>
<span class=normal>303</span>
<span class=normal>304</span>
<span class=normal>305</span>
<span class=normal>306</span>
<span class=normal>307</span>
<span class=normal>308</span>
<span class=normal>309</span>
<span class=normal>310</span>
<span class=normal>311</span>
<span class=normal>312</span>
<span class=normal>313</span>
<span class=normal>314</span>
<span class=normal>315</span>
<span class=normal>316</span>
<span class=normal>317</span>
<span class=normal>318</span>
<span class=normal>319</span>
<span class=normal>320</span>
<span class=normal>321</span>
<span class=normal>322</span>
<span class=normal>323</span>
<span class=normal>324</span>
<span class=normal>325</span>
<span class=normal>326</span>
<span class=normal>327</span>
<span class=normal>328</span>
<span class=normal>329</span>
<span class=normal>330</span>
<span class=normal>331</span>
<span class=normal>332</span>
<span class=normal>333</span>
<span class=normal>334</span>
<span class=normal>335</span>
<span class=normal>336</span>
<span class=normal>337</span>
<span class=normal>338</span>
<span class=normal>339</span>
<span class=normal>340</span>
<span class=normal>341</span>
<span class=normal>342</span>
<span class=normal>343</span>
<span class=normal>344</span>
<span class=normal>345</span>
<span class=normal>346</span>
<span class=normal>347</span>
<span class=normal>348</span>
<span class=normal>349</span>
<span class=normal>350</span>
<span class=normal>351</span>
<span class=normal>352</span>
<span class=normal>353</span>
<span class=normal>354</span>
<span class=normal>355</span>
<span class=normal>356</span>
<span class=normal>357</span>
<span class=normal>358</span>
<span class=normal>359</span>
<span class=normal>360</span>
<span class=normal>361</span>
<span class=normal>362</span>
<span class=normal>363</span>
<span class=normal>364</span>
<span class=normal>365</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>splitting_association</span><span class=p>(</span><span class=n>association</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Splitting associations based on the list of parseable variants from the GWAS Catalog.</span>

<span class=sd>    Args:</span>
<span class=sd>        association (DataFrame): DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        A dataframe with the same columns as the input dataframe, but with the variants exploded.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>cols</span> <span class=o>=</span> <span class=n>association</span><span class=o>.</span><span class=n>columns</span>
    <span class=n>variant_cols</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;position&quot;</span><span class=p>,</span>
        <span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;snpIds&quot;</span><span class=p>,</span>
    <span class=p>]</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>association</span>
        <span class=c1># Pairing together matching chr:pos:rsid in a list of structs:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;variants&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>array_contains</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>),</span> <span class=s2>&quot;Variant inconsistency&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>arrays_zip</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>),</span>
                <span class=p>),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>None</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Exploding all variants:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=o>*</span><span class=n>cols</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode_outer</span><span class=p>(</span><span class=s2>&quot;variants&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;variant&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Updating chr, pos, rsid, risk-snp-allele columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=o>*</span><span class=p>[</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;variant.</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=n>col</span><span class=p>)</span> <span class=k>if</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>variant_cols</span> <span class=k>else</span> <span class=n>col</span>
                <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>cols</span>
            <span class=p>],</span>
            <span class=c1># Extract p-value exponent:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pvalue&quot;</span><span class=p>),</span> <span class=s2>&quot;E&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span>
            <span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;integer&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pValueExponent&quot;</span><span class=p>),</span>
            <span class=c1># Extract p-value mantissa:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pvalue&quot;</span><span class=p>),</span> <span class=s2>&quot;E&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>)</span>
            <span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;float&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;pValueMantissa&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Extracting risk allele:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;riskAllele&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span> <span class=s2>&quot;-&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span>
        <span class=p>)</span>
        <span class=c1># Creating list of rsIds from gwas catalog dataset:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;rsIdsGwasCatalog&quot;</span><span class=p>,</span>
            <span class=n>_collect_rsids</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Dropping some un-used column:</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span>
            <span class=s2>&quot;pValue&quot;</span><span class=p>,</span>
            <span class=s2>&quot;riskAlleleFrequency&quot;</span><span class=p>,</span>
            <span class=s2>&quot;cnv&quot;</span><span class=p>,</span>
            <span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>,</span>
            <span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>,</span>
            <span class=s2>&quot;snpIds&quot;</span><span class=p>,</span>
            <span class=s2>&quot;pValue&quot;</span><span class=p>,</span>
            <span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>,</span>
            <span class=s2>&quot;pValueNegLog&quot;</span><span class=p>,</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest.effect_harmonization></a> <div class="doc doc-contents first"> <p>Harmonisation of GWAS stats.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.effect_harmonization.harmonize_effect class="doc doc-heading"> <code class="highlight language-python"><span class=n>harmonize_effect</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Harmonisation of effects.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>GWASCatalog stats</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Harmonised GWASCatalog stats</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/effect_harmonization.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>268</span>
<span class=normal>269</span>
<span class=normal>270</span>
<span class=normal>271</span>
<span class=normal>272</span>
<span class=normal>273</span>
<span class=normal>274</span>
<span class=normal>275</span>
<span class=normal>276</span>
<span class=normal>277</span>
<span class=normal>278</span>
<span class=normal>279</span>
<span class=normal>280</span>
<span class=normal>281</span>
<span class=normal>282</span>
<span class=normal>283</span>
<span class=normal>284</span>
<span class=normal>285</span>
<span class=normal>286</span>
<span class=normal>287</span>
<span class=normal>288</span>
<span class=normal>289</span>
<span class=normal>290</span>
<span class=normal>291</span>
<span class=normal>292</span>
<span class=normal>293</span>
<span class=normal>294</span>
<span class=normal>295</span>
<span class=normal>296</span>
<span class=normal>297</span>
<span class=normal>298</span>
<span class=normal>299</span>
<span class=normal>300</span>
<span class=normal>301</span>
<span class=normal>302</span>
<span class=normal>303</span>
<span class=normal>304</span>
<span class=normal>305</span>
<span class=normal>306</span>
<span class=normal>307</span>
<span class=normal>308</span>
<span class=normal>309</span>
<span class=normal>310</span>
<span class=normal>311</span>
<span class=normal>312</span>
<span class=normal>313</span>
<span class=normal>314</span>
<span class=normal>315</span>
<span class=normal>316</span>
<span class=normal>317</span>
<span class=normal>318</span>
<span class=normal>319</span>
<span class=normal>320</span>
<span class=normal>321</span>
<span class=normal>322</span>
<span class=normal>323</span>
<span class=normal>324</span>
<span class=normal>325</span>
<span class=normal>326</span>
<span class=normal>327</span>
<span class=normal>328</span>
<span class=normal>329</span>
<span class=normal>330</span>
<span class=normal>331</span>
<span class=normal>332</span>
<span class=normal>333</span>
<span class=normal>334</span>
<span class=normal>335</span>
<span class=normal>336</span>
<span class=normal>337</span>
<span class=normal>338</span>
<span class=normal>339</span>
<span class=normal>340</span>
<span class=normal>341</span>
<span class=normal>342</span>
<span class=normal>343</span>
<span class=normal>344</span>
<span class=normal>345</span>
<span class=normal>346</span>
<span class=normal>347</span>
<span class=normal>348</span>
<span class=normal>349</span>
<span class=normal>350</span>
<span class=normal>351</span>
<span class=normal>352</span>
<span class=normal>353</span>
<span class=normal>354</span>
<span class=normal>355</span>
<span class=normal>356</span>
<span class=normal>357</span>
<span class=normal>358</span>
<span class=normal>359</span>
<span class=normal>360</span>
<span class=normal>361</span>
<span class=normal>362</span>
<span class=normal>363</span>
<span class=normal>364</span>
<span class=normal>365</span>
<span class=normal>366</span>
<span class=normal>367</span>
<span class=normal>368</span>
<span class=normal>369</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>harmonize_effect</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Harmonisation of effects.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): GWASCatalog stats</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Harmonised GWASCatalog stats</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span>
        <span class=c1># If the alleles are palindrome - the reference and alt alleles are reverse complement of each other:</span>
        <span class=c1># eg. T -&gt; A: in such cases we cannot disambiguate the effect, which means we cannot be sure if</span>
        <span class=c1># the effect is given to the alt allele on the positive strand or the ref allele on</span>
        <span class=c1># The negative strand. We assume, we don&#39;t need to harminze.</span>
        <span class=c1># Adding a flag indicating if effect harmonization is required:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;isPalindrome&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>)</span>
                <span class=o>==</span> <span class=n>_get_reverse_complement</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>)),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># If a variant is palindrome, we drop the effect size, so no harmonization can happen:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;effectSize&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;isPalindrome&quot;</span><span class=p>),</span> <span class=kc>None</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;effectSize&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=c1># As we are calculating effect on the alternate allele, we have to harmonise effect</span>
        <span class=c1># if the risk allele is reference allele or the reverse complement of the reference allele</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;needsHarmonisation&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>))</span>
                <span class=o>|</span> <span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)</span>
                    <span class=o>==</span> <span class=n>_get_reverse_complement</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>))</span>
                <span class=p>),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Z-score is needed to calculate 95% confidence interval:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;zscore&quot;</span><span class=p>,</span>
            <span class=n>_pval_to_zscore</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>concat_ws</span><span class=p>(</span><span class=s2>&quot;E&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pValueMantissa&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pValueExponent&quot;</span><span class=p>))</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Harmonizing betas + calculate the corresponding confidence interval:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta&quot;</span><span class=p>,</span>
            <span class=n>_harmonize_beta</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;effectSize&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;confidenceInterval&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;needsHarmonisation&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta_ci_upper&quot;</span><span class=p>,</span>
            <span class=n>_calculate_beta_ci</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;upper&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta_ci_lower&quot;</span><span class=p>,</span>
            <span class=n>_calculate_beta_ci</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;lower&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=c1># Harmonizing odds-ratios + calculate the corresponding confidence interval:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio&quot;</span><span class=p>,</span>
            <span class=n>_harmonize_odds_ratio</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;effectSize&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;confidenceInterval&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;needsHarmonisation&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio_ci_upper&quot;</span><span class=p>,</span>
            <span class=n>_calculate_or_ci</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;upper&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio_ci_lower&quot;</span><span class=p>,</span>
            <span class=n>_calculate_or_ci</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;lower&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=c1># Adding QC flag to variants with palindrome alleles:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;qualityControl&quot;</span><span class=p>,</span>
            <span class=n>adding_quality_flag</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;qualityControl&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;isPalindrome&quot;</span><span class=p>),</span> <span class=n>PALINDROME_FLAG</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Dropping unused columns:</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span>
            <span class=s2>&quot;zscore&quot;</span><span class=p>,</span>
            <span class=s2>&quot;effectSize&quot;</span><span class=p>,</span>
            <span class=s2>&quot;confidenceInterval&quot;</span><span class=p>,</span>
            <span class=s2>&quot;needsHarmonisation&quot;</span><span class=p>,</span>
            <span class=s2>&quot;isPalindrome&quot;</span><span class=p>,</span>
            <span class=s2>&quot;zscore&quot;</span><span class=p>,</span>
            <span class=s2>&quot;riskAllele&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">October 1, 2022</span> </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../colocalisation/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Modules" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Modules </div> </div> </a> <a href=../schemas/ class="md-footer__link md-footer__link--next" aria-label="Next: Schemas" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Schemas </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.instant", "navigation.top", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../../assets/javascripts/bundle.5a2dcb6a.min.js></script> </body> </html>