<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=../../assets/favicon.png><meta name=generator content="mkdocs-1.4.0, mkdocs-material-8.4.4"><title>Gwas ingest - Genetics Portal Pipeline</title><link rel=stylesheet href=../../assets/stylesheets/main.6b80c2a2.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.cbb835fc.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#etl.gwas_ingest class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="Genetics Portal Pipeline" class="md-header__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Genetics Portal Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Gwas ingest </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="Genetics Portal Pipeline" class="md-nav__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> Genetics Portal Pipeline </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Home </a> </li> <li class=md-nav__item> <a href=../../roadmap/ class=md-nav__link> Roadmap </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <div class="md-nav__link md-nav__link--index "> <a href=../colocalisation/ >Modules</a> <label for=__nav_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav aria-label=Modules data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Modules </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> Gwas ingest <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> Gwas ingest </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#etl.gwas_ingest class=md-nav__link> etl.gwas_ingest </a> <nav class=md-nav aria-label=etl.gwas_ingest> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.gwas_ingest--summary-of-the-logic class=md-nav__link> Summary of the logic </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion class=md-nav__link> etl.gwas_ingest.study_ingestion </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.column2camel_case class=md-nav__link> column2camel_case() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.extract_discovery_sample_sizes class=md-nav__link> extract_discovery_sample_sizes() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.get_sumstats_location class=md-nav__link> get_sumstats_location() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.ingest_gwas_catalog_studies class=md-nav__link> ingest_gwas_catalog_studies() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_ancestries class=md-nav__link> parse_ancestries() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_efos class=md-nav__link> parse_efos() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.read_study_table class=md-nav__link> read_study_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations class=md-nav__link> etl.gwas_ingest.process_associations </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.concordance_filter class=md-nav__link> concordance_filter() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.deduplicate class=md-nav__link> deduplicate() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.filter_assoc_by_maf class=md-nav__link> filter_assoc_by_maf() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.filter_assoc_by_rsid class=md-nav__link> filter_assoc_by_rsid() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.ingest_gwas_catalog_associations class=md-nav__link> ingest_gwas_catalog_associations() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.map_variants class=md-nav__link> map_variants() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.process_associations class=md-nav__link> process_associations() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.read_associations_data class=md-nav__link> read_associations_data() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization class=md-nav__link> etl.gwas_ingest.effect_harmonization </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.get_reverse_complement class=md-nav__link> get_reverse_complement() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonise_beta class=md-nav__link> harmonise_beta() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonise_odds_ratio class=md-nav__link> harmonise_odds_ratio() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonize_effect class=md-nav__link> harmonize_effect() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.pval_to_zscore class=md-nav__link> pval_to_zscore() </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../intervals/ class=md-nav__link> Intervals </a> </li> <li class=md-nav__item> <a href=../schemas/ class=md-nav__link> Schemas </a> </li> <li class=md-nav__item> <a href=../variants/ class=md-nav__link> Variants </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#etl.gwas_ingest class=md-nav__link> etl.gwas_ingest </a> <nav class=md-nav aria-label=etl.gwas_ingest> <ul class=md-nav__list> <li class=md-nav__item> <a href=#etl.gwas_ingest--summary-of-the-logic class=md-nav__link> Summary of the logic </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion class=md-nav__link> etl.gwas_ingest.study_ingestion </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.column2camel_case class=md-nav__link> column2camel_case() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.extract_discovery_sample_sizes class=md-nav__link> extract_discovery_sample_sizes() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.get_sumstats_location class=md-nav__link> get_sumstats_location() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.ingest_gwas_catalog_studies class=md-nav__link> ingest_gwas_catalog_studies() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_ancestries class=md-nav__link> parse_ancestries() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.parse_efos class=md-nav__link> parse_efos() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.study_ingestion.read_study_table class=md-nav__link> read_study_table() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations class=md-nav__link> etl.gwas_ingest.process_associations </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.concordance_filter class=md-nav__link> concordance_filter() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.deduplicate class=md-nav__link> deduplicate() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.filter_assoc_by_maf class=md-nav__link> filter_assoc_by_maf() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.filter_assoc_by_rsid class=md-nav__link> filter_assoc_by_rsid() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.ingest_gwas_catalog_associations class=md-nav__link> ingest_gwas_catalog_associations() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.map_variants class=md-nav__link> map_variants() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.process_associations class=md-nav__link> process_associations() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.process_associations.read_associations_data class=md-nav__link> read_associations_data() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization class=md-nav__link> etl.gwas_ingest.effect_harmonization </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.get_reverse_complement class=md-nav__link> get_reverse_complement() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonise_beta class=md-nav__link> harmonise_beta() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonise_odds_ratio class=md-nav__link> harmonise_odds_ratio() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.harmonize_effect class=md-nav__link> harmonize_effect() </a> </li> <li class=md-nav__item> <a href=#etl.gwas_ingest.effect_harmonization.pval_to_zscore class=md-nav__link> pval_to_zscore() </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/opentargets/genetics_etl_python/edit/master/docs/modules/gwas_ingest.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1>Gwas ingest</h1> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest></a> <div class="doc doc-contents first"> <p>This pipeline ingest the curated GWAS Catalog associations and studies. Prepares top-loci table and study table.</p> <h3 id=etl.gwas_ingest--summary-of-the-logic>Summary of the logic</h3> <ol> <li>Reading association data set. Applying some filters and do basic processing.</li> <li>Map associated variants to GnomAD variants based on the annotated chromosome:position (on GRCh38).</li> <li>Harmonize effect size, calculate Z-score + direction of effect.</li> <li>Read and process study table (parse ancestry, sample size etc).</li> <li>Join study table with associations on <code>study_accession</code>. Split studies when necessary on the basis that one study describes one trait only.</li> <li>Split data into top-loci and study tables and save.</li> </ol> <div class="doc doc-children"> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest.study_ingestion></a> <div class="doc doc-contents first"> <p>GWAS Catalog study ingestion.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.column2camel_case class="doc doc-heading"> <code class="highlight language-python"><span class=n>column2camel_case</span><span class=p>(</span><span class=n>s</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>A helper function to convert column names to camel cases.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>s</code></td> <td> <code>str</code> </td> <td><p>a single column name</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>str</code></td> <td> <code>str</code> </td> <td><p>spark expression to select and rename the column</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>column2camel_case</span><span class=p>(</span><span class=n>s</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=nb>str</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;A helper function to convert column names to camel cases.</span>

<span class=sd>    Args:</span>
<span class=sd>        s (str): a single column name</span>

<span class=sd>    Returns:</span>
<span class=sd>        str: spark expression to select and rename the column</span>
<span class=sd>    &quot;&quot;&quot;</span>

    <span class=k>def</span> <span class=nf>string2camelcase</span><span class=p>(</span><span class=n>s</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=nb>str</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Converting a string to camelcase.</span>

<span class=sd>        Args:</span>
<span class=sd>            s (str): a random string</span>

<span class=sd>        Returns:</span>
<span class=sd>            str: Camel cased string</span>
<span class=sd>        &quot;&quot;&quot;</span>
        <span class=c1># Removing a bunch of unwanted characters from the column names:</span>
        <span class=n>s</span> <span class=o>=</span> <span class=n>re</span><span class=o>.</span><span class=n>sub</span><span class=p>(</span><span class=sa>r</span><span class=s2>&quot;[\/\(\)\-]+&quot;</span><span class=p>,</span> <span class=s2>&quot; &quot;</span><span class=p>,</span> <span class=n>s</span><span class=p>)</span>

        <span class=n>first</span><span class=p>,</span> <span class=o>*</span><span class=n>rest</span> <span class=o>=</span> <span class=n>s</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=s2>&quot; &quot;</span><span class=p>)</span>
        <span class=k>return</span> <span class=s2>&quot;&quot;</span><span class=o>.</span><span class=n>join</span><span class=p>([</span><span class=n>first</span><span class=o>.</span><span class=n>lower</span><span class=p>(),</span> <span class=o>*</span><span class=nb>map</span><span class=p>(</span><span class=nb>str</span><span class=o>.</span><span class=n>capitalize</span><span class=p>,</span> <span class=n>rest</span><span class=p>)])</span>

    <span class=k>return</span> <span class=sa>f</span><span class=s2>&quot;`</span><span class=si>{</span><span class=n>s</span><span class=si>}</span><span class=s2>` as </span><span class=si>{</span><span class=n>string2camelcase</span><span class=p>(</span><span class=n>s</span><span class=p>)</span><span class=si>}</span><span class=s2>&quot;</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.extract_discovery_sample_sizes class="doc doc-heading"> <code class="highlight language-python"><span class=n>extract_discovery_sample_sizes</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extract the sample size of the discovery stage of the study as annotated in the GWAS Catalog.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>gwas studies table with a column called <code>initialSampleSize</code></p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>df with columns <code>nCases</code>, <code>nControls</code>, and <code>nSamples</code> per <code>studyAccession</code></p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>extract_discovery_sample_sizes</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extract the sample size of the discovery stage of the study as annotated in the GWAS Catalog.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): gwas studies table with a column called `initialSampleSize`</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: df with columns `nCases`, `nControls`, and `nSamples` per `studyAccession`</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode_outer</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleSize&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;,\s+&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;samples&quot;</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Extracting the sample size from the string:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;sampleSize&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_extract</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;samples&quot;</span><span class=p>),</span> <span class=s2>&quot;,&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;[0-9,]+&quot;</span><span class=p>,</span> <span class=mi>0</span>
            <span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>IntegerType</span><span class=p>()),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=s2>&quot;sampleSize&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;samples&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;cases&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nCases&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;samples&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;controls&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nControls&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Aggregating sample sizes for all ancestries:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=s2>&quot;nCases&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nCases&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=s2>&quot;nControls&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nControls&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;nSamples&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.get_sumstats_location class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_sumstats_location</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>summarystats_list</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Get summary stat locations.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>summarystats_list</code></td> <td> <code>str</code> </td> <td><p>filepath of table listing summary stats</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>dataframe with each GCST with summary stats and its location</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span>
<span class=normal>76</span>
<span class=normal>77</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_sumstats_location</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>summarystats_list</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get summary stat locations.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        summarystats_list (str): filepath of table listing summary stats</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: dataframe with each GCST with summary stats and its location</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>gwas_sumstats_base_uri</span> <span class=o>=</span> <span class=s2>&quot;ftp://ftp.ebi.ac.uk/pub/databases/gwas/summary_statistics&quot;</span>

    <span class=n>sumstats</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>summarystats_list</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>False</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>concat</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>gwas_sumstats_base_uri</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>regexp_replace</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;_c0&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;^\.\/&quot;</span><span class=p>,</span> <span class=s2>&quot;&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>regexp_extract</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>),</span> <span class=sa>r</span><span class=s2>&quot;\/(GCST\d+)\/&quot;</span><span class=p>,</span> <span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;studyAccession&quot;</span>
            <span class=p>),</span>
            <span class=s2>&quot;summarystatsLocation&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;hasSumstats&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s2>&quot;Number of studies with harmonized summary stats: </span><span class=si>{</span><span class=n>sumstats</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=n>sumstats</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.ingest_gwas_catalog_studies class="doc doc-heading"> <code class="highlight language-python"><span class=n>ingest_gwas_catalog_studies</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>study_file</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>,</span> <span class=n>summary_stats_list</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>This function ingests study level metadata from the GWAS Catalog.</p> <p>The following information is aggregated/extracted: - All publication related information retained. - Mapped measured and background traits parsed. - Flagged if harmonized summary statistics datasets available. - If available, the ftp path to these files presented. - Ancestries from the discovery and replication stages are structured with sample counts. - Case/control counts extracted. - The number of samples with European ancestry extracted.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETLSession</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>study_file</code></td> <td> <code>str</code> </td> <td><p>path to the GWAS Catalog study table v1.0.3.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>ancestry_file</code></td> <td> <code>str</code> </td> <td><p>path to the GWAS Catalog ancestry table.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>summary_stats_list</code></td> <td> <code>str</code> </td> <td><p>path to the GWAS Catalog harmonized summary statistics list.</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Parsed and annotated GWAS Catalog study table.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>271</span>
<span class=normal>272</span>
<span class=normal>273</span>
<span class=normal>274</span>
<span class=normal>275</span>
<span class=normal>276</span>
<span class=normal>277</span>
<span class=normal>278</span>
<span class=normal>279</span>
<span class=normal>280</span>
<span class=normal>281</span>
<span class=normal>282</span>
<span class=normal>283</span>
<span class=normal>284</span>
<span class=normal>285</span>
<span class=normal>286</span>
<span class=normal>287</span>
<span class=normal>288</span>
<span class=normal>289</span>
<span class=normal>290</span>
<span class=normal>291</span>
<span class=normal>292</span>
<span class=normal>293</span>
<span class=normal>294</span>
<span class=normal>295</span>
<span class=normal>296</span>
<span class=normal>297</span>
<span class=normal>298</span>
<span class=normal>299</span>
<span class=normal>300</span>
<span class=normal>301</span>
<span class=normal>302</span>
<span class=normal>303</span>
<span class=normal>304</span>
<span class=normal>305</span>
<span class=normal>306</span>
<span class=normal>307</span>
<span class=normal>308</span>
<span class=normal>309</span>
<span class=normal>310</span>
<span class=normal>311</span>
<span class=normal>312</span>
<span class=normal>313</span>
<span class=normal>314</span>
<span class=normal>315</span>
<span class=normal>316</span>
<span class=normal>317</span>
<span class=normal>318</span>
<span class=normal>319</span>
<span class=normal>320</span>
<span class=normal>321</span>
<span class=normal>322</span>
<span class=normal>323</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>ingest_gwas_catalog_studies</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>study_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>summary_stats_list</span><span class=p>:</span> <span class=nb>str</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;This function ingests study level metadata from the GWAS Catalog.</span>

<span class=sd>    The following information is aggregated/extracted:</span>
<span class=sd>    - All publication related information retained.</span>
<span class=sd>    - Mapped measured and background traits parsed.</span>
<span class=sd>    - Flagged if harmonized summary statistics datasets available.</span>
<span class=sd>    - If available, the ftp path to these files presented.</span>
<span class=sd>    - Ancestries from the discovery and replication stages are structured with sample counts.</span>
<span class=sd>    - Case/control counts extracted.</span>
<span class=sd>    - The number of samples with European ancestry extracted.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETLSession</span>
<span class=sd>        study_file (str): path to the GWAS Catalog study table v1.0.3.</span>
<span class=sd>        ancestry_file (str): path to the GWAS Catalog ancestry table.</span>
<span class=sd>        summary_stats_list (str): path to the GWAS Catalog harmonized summary statistics list.</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Parsed and annotated GWAS Catalog study table.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Read GWAS Catalogue raw data</span>
    <span class=n>gwas_studies</span> <span class=o>=</span> <span class=n>read_study_table</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>study_file</span><span class=p>)</span>
    <span class=n>gwas_ancestries</span> <span class=o>=</span> <span class=n>parse_ancestries</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>)</span>

    <span class=n>study_size_df</span> <span class=o>=</span> <span class=n>extract_discovery_sample_sizes</span><span class=p>(</span><span class=n>gwas_studies</span><span class=p>)</span>
    <span class=n>ss_studies</span> <span class=o>=</span> <span class=n>get_sumstats_location</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>summary_stats_list</span><span class=p>)</span>

    <span class=n>studies</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>gwas_studies</span>
        <span class=c1># Add study sizes:</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>study_size_df</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>
        <span class=c1># Adding summary stats location:</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>ss_studies</span><span class=p>,</span>
            <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;hasSumstats&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;hasSumstats&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=kc>False</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span><span class=n>gwas_ancestries</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;*&quot;</span><span class=p>,</span>
            <span class=n>parse_efos</span><span class=p>(</span><span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;efos&quot;</span><span class=p>),</span>
            <span class=n>parse_efos</span><span class=p>(</span><span class=s2>&quot;mappedBackgroundTraitUri&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;backgroundEfos&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;initialSampleSize&quot;</span><span class=p>,</span> <span class=s2>&quot;mappedBackgroundTraitUri&quot;</span><span class=p>,</span> <span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=n>validate_df_schema</span><span class=p>(</span><span class=n>studies</span><span class=p>,</span> <span class=s2>&quot;studies.json&quot;</span><span class=p>)</span>

    <span class=k>return</span> <span class=n>studies</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.parse_ancestries class="doc doc-heading"> <code class="highlight language-python"><span class=n>parse_ancestries</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extracting sample sizes and ancestry information.</p> <p>This function parses the ancestry data. Also get counts for the europeans in the same discovery stage.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>ancestry_file</code></td> <td> <code>str</code> </td> <td><p>File name of the ancestry table as downloaded from the GWAS Catalog</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Slimmed and cleaned version of the ancestry annotation.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span>
<span class=normal>194</span>
<span class=normal>195</span>
<span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span>
<span class=normal>209</span>
<span class=normal>210</span>
<span class=normal>211</span>
<span class=normal>212</span>
<span class=normal>213</span>
<span class=normal>214</span>
<span class=normal>215</span>
<span class=normal>216</span>
<span class=normal>217</span>
<span class=normal>218</span>
<span class=normal>219</span>
<span class=normal>220</span>
<span class=normal>221</span>
<span class=normal>222</span>
<span class=normal>223</span>
<span class=normal>224</span>
<span class=normal>225</span>
<span class=normal>226</span>
<span class=normal>227</span>
<span class=normal>228</span>
<span class=normal>229</span>
<span class=normal>230</span>
<span class=normal>231</span>
<span class=normal>232</span>
<span class=normal>233</span>
<span class=normal>234</span>
<span class=normal>235</span>
<span class=normal>236</span>
<span class=normal>237</span>
<span class=normal>238</span>
<span class=normal>239</span>
<span class=normal>240</span>
<span class=normal>241</span>
<span class=normal>242</span>
<span class=normal>243</span>
<span class=normal>244</span>
<span class=normal>245</span>
<span class=normal>246</span>
<span class=normal>247</span>
<span class=normal>248</span>
<span class=normal>249</span>
<span class=normal>250</span>
<span class=normal>251</span>
<span class=normal>252</span>
<span class=normal>253</span>
<span class=normal>254</span>
<span class=normal>255</span>
<span class=normal>256</span>
<span class=normal>257</span>
<span class=normal>258</span>
<span class=normal>259</span>
<span class=normal>260</span>
<span class=normal>261</span>
<span class=normal>262</span>
<span class=normal>263</span>
<span class=normal>264</span>
<span class=normal>265</span>
<span class=normal>266</span>
<span class=normal>267</span>
<span class=normal>268</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>parse_ancestries</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>ancestry_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extracting sample sizes and ancestry information.</span>

<span class=sd>    This function parses the ancestry data. Also get counts for the europeans in the same</span>
<span class=sd>    discovery stage.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        ancestry_file (str): File name of the ancestry table as downloaded from the GWAS Catalog</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Slimmed and cleaned version of the ancestry annotation.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Reading ancestry data:</span>
    <span class=n>ancestry</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>ancestry_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
        <span class=c1># Convert column headers to camelcase:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span>
            <span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=o>*</span><span class=p>[</span><span class=n>f</span><span class=o>.</span><span class=n>expr</span><span class=p>(</span><span class=n>column2camel_case</span><span class=p>(</span><span class=n>x</span><span class=p>))</span> <span class=k>for</span> <span class=n>x</span> <span class=ow>in</span> <span class=n>df</span><span class=o>.</span><span class=n>columns</span><span class=p>])</span>
        <span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Get a high resolution dataset on experimental stage:</span>
    <span class=n>ancestry_stages</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>ancestry</span><span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>pivot</span><span class=p>(</span><span class=s2>&quot;stage&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>collect_set</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>struct</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;numberOfIndividuals&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;sampleSize&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;broadAncestralCategory&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;ancestry&quot;</span><span class=p>),</span>
                <span class=p>)</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;initial&quot;</span><span class=p>,</span> <span class=s2>&quot;discoverySamples&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>withColumnRenamed</span><span class=p>(</span><span class=s2>&quot;replication&quot;</span><span class=p>,</span> <span class=s2>&quot;replicationSamples&quot;</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Generate information on the ancestry composition of the discovery stage, and calculate</span>
    <span class=c1># the proportion of the Europeans:</span>
    <span class=n>europeans_deconvoluted</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>ancestry</span>
        <span class=c1># Focus on discovery stage:</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;stage&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;initial&quot;</span><span class=p>)</span>
        <span class=c1># Sorting ancestries if European:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;ancestryFlag&quot;</span><span class=p>,</span>
            <span class=c1># Excluding finnish:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleDescription&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;Finnish&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span>
            <span class=p>)</span>
            <span class=c1># Excluding Icelandic population:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleDescription&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;Icelandic&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span>
            <span class=p>)</span>
            <span class=c1># Including European ancestry:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;broadAncestralCategory&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;European&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>))</span>
            <span class=c1># Exclude all other population:</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Grouping by study accession and initial sample description:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>pivot</span><span class=p>(</span><span class=s2>&quot;ancestryFlag&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=c1># Summarizing sample sizes for all ancestries:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;numberOfIndividuals&quot;</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=c1># Do aritmetics to make sure we have the right proportion of european in the set:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;initialSampleCountEuropean&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;initialSampleCountOther&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>(),</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;initialSampleCount&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;initialSampleCountEuropean&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;other&quot;</span><span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;european&quot;</span><span class=p>,</span> <span class=s2>&quot;other&quot;</span><span class=p>,</span> <span class=s2>&quot;initialSampleCountOther&quot;</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>ancestry_stages</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>europeans_deconvoluted</span><span class=p>,</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;outer&quot;</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.parse_efos class="doc doc-heading"> <code class="highlight language-python"><span class=n>parse_efos</span><span class=p>(</span><span class=n>c</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Extracting EFO identifiers.</p> <p>This function parses EFO identifiers from a comma-separated list of EFO URIs.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>c</code></td> <td> <code>str</code> </td> <td><p>name of column with a list of EFO IDs</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>Column</code></td> <td> <code><span title="pyspark.sql.Column">Column</span></code> </td> <td><p>column with a list of parsed EFO IDs</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>parse_efos</span><span class=p>(</span><span class=n>c</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Extracting EFO identifiers.</span>

<span class=sd>    This function parses EFO identifiers from a comma-separated list of EFO URIs.</span>

<span class=sd>    Args:</span>
<span class=sd>        c (str): name of column with a list of EFO IDs</span>

<span class=sd>    Returns:</span>
<span class=sd>        Column: column with a list of parsed EFO IDs</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>expr</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;regexp_extract_all(</span><span class=si>{</span><span class=n>c</span><span class=si>}</span><span class=s2>, &#39;([A-Z]+_[0-9]+)&#39;)&quot;</span><span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.study_ingestion.read_study_table class="doc doc-heading"> <code class="highlight language-python"><span class=n>read_study_table</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_study_file</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Read GWASCatalog study table.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gwas_study_file</code></td> <td> <code>str</code> </td> <td><p>GWAS studies filepath</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Study table.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/study_ingestion.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>80</span>
<span class=normal>81</span>
<span class=normal>82</span>
<span class=normal>83</span>
<span class=normal>84</span>
<span class=normal>85</span>
<span class=normal>86</span>
<span class=normal>87</span>
<span class=normal>88</span>
<span class=normal>89</span>
<span class=normal>90</span>
<span class=normal>91</span>
<span class=normal>92</span>
<span class=normal>93</span>
<span class=normal>94</span>
<span class=normal>95</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>read_study_table</span><span class=p>(</span><span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>gwas_study_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Read GWASCatalog study table.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): ETL session</span>
<span class=sd>        gwas_study_file (str): GWAS studies filepath</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Study table.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>gwas_study_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=o>*</span><span class=p>[</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>old_name</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=n>new_name</span><span class=p>)</span>
            <span class=k>for</span> <span class=n>old_name</span><span class=p>,</span> <span class=n>new_name</span> <span class=ow>in</span> <span class=n>STUDY_COLUMNS_MAP</span><span class=o>.</span><span class=n>items</span><span class=p>()</span>
        <span class=p>]</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest.process_associations></a> <div class="doc doc-contents first"> <p>Process GWAS catalog associations.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.concordance_filter class="doc doc-heading"> <code class="highlight language-python"><span class=n>concordance_filter</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Concordance filter.</p> <p>This function filters for variants with concordant alleles with the association's reported risk allele. A risk allele is considered concordant if: - equal to the alt allele - equal to the ref allele - equal to the revese complement of the alt allele. - equal to the revese complement of the ref allele. - Or the risk allele is ambigious, noted by '?'</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations filtered for variants with concordant alleles with the risk allele.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>concordance_filter</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Concordance filter.</span>

<span class=sd>    This function filters for variants with concordant alleles with the association&#39;s reported risk allele.</span>
<span class=sd>    A risk allele is considered concordant if:</span>
<span class=sd>    - equal to the alt allele</span>
<span class=sd>    - equal to the ref allele</span>
<span class=sd>    - equal to the revese complement of the alt allele.</span>
<span class=sd>    - equal to the revese complement of the ref allele.</span>
<span class=sd>    - Or the risk allele is ambigious, noted by &#39;?&#39;</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): associations</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: associations filtered for variants with concordant alleles with the risk allele.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span>
        <span class=c1># Adding column with the reverse-complement of the risk allele:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;riskAlleleReverseComplement&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=sa>r</span><span class=s2>&quot;^[ACTG]+$&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>reverse</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>translate</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>),</span> <span class=s2>&quot;ACTG&quot;</span><span class=p>,</span> <span class=s2>&quot;TGAC&quot;</span><span class=p>)),</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=c1># Adding columns flagging concordance:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;isConcordant&quot;</span><span class=p>,</span>
            <span class=c1># If risk allele is found on the positive strand:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>))</span>
                <span class=o>|</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>)),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># If risk allele is found on the negative strand:</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAlleleReverseComplement&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;referenceAllele&quot;</span><span class=p>))</span>
                <span class=o>|</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAlleleReverseComplement&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;alternateAllele&quot;</span><span class=p>)),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span>
            <span class=c1># If risk allele is ambiguous, still accepted: &lt; This condition could be reconsidered</span>
            <span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;?&quot;</span><span class=p>,</span> <span class=kc>True</span><span class=p>)</span>
            <span class=c1># Allele is discordant:</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Dropping discordant associations:</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;isConcordant&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;isConcordant&quot;</span><span class=p>,</span> <span class=s2>&quot;riskAlleleReverseComplement&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.deduplicate class="doc doc-heading"> <code class="highlight language-python"><span class=n>deduplicate</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Deduplicate DataFrame (not implemented).</p> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>299</span>
<span class=normal>300</span>
<span class=normal>301</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>deduplicate</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Deduplicate DataFrame (not implemented).&quot;&quot;&quot;</span>
    <span class=k>raise</span> <span class=ne>NotImplementedError</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.filter_assoc_by_maf class="doc doc-heading"> <code class="highlight language-python"><span class=n>filter_assoc_by_maf</span><span class=p>(</span><span class=n>associations</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Filter associations by Minor Allele Frequency.</p> <p>If an association is mapped to multiple concordant (or at least not discordant) variants, we only keep the one that has the highest minor allele frequency.</p> <p>This filter is based on the assumption that GWAS Studies are focusing on common variants mostly. This filter is especially designed to filter out rare alleles of multiallelics with matching rsIDs.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>associations</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations filtered by allelic frequency</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span>
<span class=normal>76</span>
<span class=normal>77</span>
<span class=normal>78</span>
<span class=normal>79</span>
<span class=normal>80</span>
<span class=normal>81</span>
<span class=normal>82</span>
<span class=normal>83</span>
<span class=normal>84</span>
<span class=normal>85</span>
<span class=normal>86</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>filter_assoc_by_maf</span><span class=p>(</span><span class=n>associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Filter associations by Minor Allele Frequency.</span>

<span class=sd>    If an association is mapped to multiple concordant (or at least not discordant) variants,</span>
<span class=sd>    we only keep the one that has the highest minor allele frequency.</span>

<span class=sd>    This filter is based on the assumption that GWAS Studies are focusing on common variants mostly.</span>
<span class=sd>    This filter is especially designed to filter out rare alleles of multiallelics with matching rsIDs.</span>

<span class=sd>    Args:</span>
<span class=sd>        associations (DataFrame): associations</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: associations filtered by allelic frequency</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># parsing population names from schema:</span>
    <span class=k>for</span> <span class=n>field</span> <span class=ow>in</span> <span class=n>associations</span><span class=o>.</span><span class=n>schema</span><span class=o>.</span><span class=n>fields</span><span class=p>:</span>
        <span class=k>if</span> <span class=n>field</span><span class=o>.</span><span class=n>name</span> <span class=o>==</span> <span class=s2>&quot;alleleFrequencies&quot;</span> <span class=ow>and</span> <span class=nb>isinstance</span><span class=p>(</span>
            <span class=n>field</span><span class=o>.</span><span class=n>dataType</span><span class=p>,</span> <span class=n>t</span><span class=o>.</span><span class=n>StructType</span>
        <span class=p>):</span>
            <span class=n>pop_names</span> <span class=o>=</span> <span class=n>field</span><span class=o>.</span><span class=n>dataType</span><span class=o>.</span><span class=n>fieldNames</span><span class=p>()</span>
            <span class=k>break</span>

    <span class=k>def</span> <span class=nf>af2maf</span><span class=p>(</span><span class=n>c</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
        <span class=sd>&quot;&quot;&quot;Column function to calculate minor allele frequency from allele frequency.&quot;&quot;&quot;</span>
        <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>c</span> <span class=o>&gt;</span> <span class=mf>0.5</span><span class=p>,</span> <span class=mi>1</span> <span class=o>-</span> <span class=n>c</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>c</span><span class=p>)</span>

    <span class=c1># Windowing through all associations. Within an association, rows are ordered by the maximum MAF:</span>
    <span class=n>w</span> <span class=o>=</span> <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=s2>&quot;associationId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>orderBy</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>desc</span><span class=p>(</span><span class=s2>&quot;maxMAF&quot;</span><span class=p>))</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>associations</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;maxMAF&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array_max</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span>
                    <span class=o>*</span><span class=p>[</span><span class=n>af2maf</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;alleleFrequencies.</span><span class=si>{</span><span class=n>pop</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>))</span> <span class=k>for</span> <span class=n>pop</span> <span class=ow>in</span> <span class=n>pop_names</span><span class=p>]</span>
                <span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;row_number&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>row_number</span><span class=p>()</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>w</span><span class=p>))</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;row_number&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=mi>1</span><span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;row_number&quot;</span><span class=p>,</span> <span class=s2>&quot;alleleFrequencies&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.filter_assoc_by_rsid class="doc doc-heading"> <code class="highlight language-python"><span class=n>filter_assoc_by_rsid</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Filter associations by rsid.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations requiring: - associationId - rsIdsGwasCatalog - rsIdsGnomad</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>filtered associations</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>304</span>
<span class=normal>305</span>
<span class=normal>306</span>
<span class=normal>307</span>
<span class=normal>308</span>
<span class=normal>309</span>
<span class=normal>310</span>
<span class=normal>311</span>
<span class=normal>312</span>
<span class=normal>313</span>
<span class=normal>314</span>
<span class=normal>315</span>
<span class=normal>316</span>
<span class=normal>317</span>
<span class=normal>318</span>
<span class=normal>319</span>
<span class=normal>320</span>
<span class=normal>321</span>
<span class=normal>322</span>
<span class=normal>323</span>
<span class=normal>324</span>
<span class=normal>325</span>
<span class=normal>326</span>
<span class=normal>327</span>
<span class=normal>328</span>
<span class=normal>329</span>
<span class=normal>330</span>
<span class=normal>331</span>
<span class=normal>332</span>
<span class=normal>333</span>
<span class=normal>334</span>
<span class=normal>335</span>
<span class=normal>336</span>
<span class=normal>337</span>
<span class=normal>338</span>
<span class=normal>339</span>
<span class=normal>340</span>
<span class=normal>341</span>
<span class=normal>342</span>
<span class=normal>343</span>
<span class=normal>344</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>filter_assoc_by_rsid</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Filter associations by rsid.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): associations requiring:</span>
<span class=sd>            - associationId</span>
<span class=sd>            - rsIdsGwasCatalog</span>
<span class=sd>            - rsIdsGnomad</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: filtered associations</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>w</span> <span class=o>=</span> <span class=n>Window</span><span class=o>.</span><span class=n>partitionBy</span><span class=p>(</span><span class=s2>&quot;associationId&quot;</span><span class=p>)</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span>
        <span class=c1># See if the GnomAD variant that was mapped to a given association has a matching rsId:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;matchingRsId&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>size</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>array_intersect</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIdsGwasCatalog&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;rsIdsGnomad&quot;</span><span class=p>))</span>
                <span class=p>)</span>
                <span class=o>&gt;</span> <span class=mi>0</span><span class=p>,</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;successfulMappingExists&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array_contains</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>collect_set</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;matchingRsId&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>over</span><span class=p>(</span><span class=n>w</span><span class=p>),</span> <span class=kc>True</span><span class=p>),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;matchingRsId&quot;</span><span class=p>)</span> <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;successfulMappingExists&quot;</span><span class=p>))</span>
            <span class=o>|</span> <span class=p>(</span><span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;matchingRsId&quot;</span><span class=p>)</span> <span class=o>&amp;</span> <span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;successfulMappingExists&quot;</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;successfulMappingExists&quot;</span><span class=p>,</span> <span class=s2>&quot;matchingRsId&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.ingest_gwas_catalog_associations class="doc doc-heading"> <code class="highlight language-python"><span class=n>ingest_gwas_catalog_associations</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_association_path</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Ingest/process/map GWAS Catalog association.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gwas_association_path</code></td> <td> <code>str</code> </td> <td><p>GWAS catalogue dataset path</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation_path</code></td> <td> <code>str</code> </td> <td><p>variant annotation dataset path</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>pvalue_cutoff</code></td> <td> <code>float</code> </td> <td><p>GWAS significance threshold</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Post-processed GWASCatalog associations</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>382</span>
<span class=normal>383</span>
<span class=normal>384</span>
<span class=normal>385</span>
<span class=normal>386</span>
<span class=normal>387</span>
<span class=normal>388</span>
<span class=normal>389</span>
<span class=normal>390</span>
<span class=normal>391</span>
<span class=normal>392</span>
<span class=normal>393</span>
<span class=normal>394</span>
<span class=normal>395</span>
<span class=normal>396</span>
<span class=normal>397</span>
<span class=normal>398</span>
<span class=normal>399</span>
<span class=normal>400</span>
<span class=normal>401</span>
<span class=normal>402</span>
<span class=normal>403</span>
<span class=normal>404</span>
<span class=normal>405</span>
<span class=normal>406</span>
<span class=normal>407</span>
<span class=normal>408</span>
<span class=normal>409</span>
<span class=normal>410</span>
<span class=normal>411</span>
<span class=normal>412</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>ingest_gwas_catalog_associations</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span>
    <span class=n>gwas_association_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>variant_annotation_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span>
    <span class=n>pvalue_cutoff</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Ingest/process/map GWAS Catalog association.</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        gwas_association_path (str): GWAS catalogue dataset path</span>
<span class=sd>        variant_annotation_path (str): variant annotation dataset path</span>
<span class=sd>        pvalue_cutoff (float): GWAS significance threshold</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Post-processed GWASCatalog associations</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=c1># 1. Read associations:</span>
        <span class=n>read_associations_data</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_association_path</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>)</span>
        <span class=c1># 2. Process -&gt; apply filter:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>process_associations</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>etl</span><span class=p>))</span>
        <span class=c1># 3. Map variants to GnomAD3:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>map_variants</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>,</span> <span class=n>etl</span><span class=p>))</span>
        <span class=c1># 4. Remove discordants:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>concordance_filter</span><span class=p>)</span>
        <span class=c1># 5. deduplicate associations by matching rsIDs:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>filter_assoc_by_rsid</span><span class=p>)</span>
        <span class=c1># 6. deduplication by MAF:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>filter_assoc_by_maf</span><span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.map_variants class="doc doc-heading"> <code class="highlight language-python"><span class=n>map_variants</span><span class=p>(</span><span class=n>parsed_associations</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>,</span> <span class=n>etl</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Add variant metadata in associations.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>parsed_associations</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>variant_annotation_path</code></td> <td> <code>str</code> </td> <td><p>variant annotation path</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations with variant metadata</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>347</span>
<span class=normal>348</span>
<span class=normal>349</span>
<span class=normal>350</span>
<span class=normal>351</span>
<span class=normal>352</span>
<span class=normal>353</span>
<span class=normal>354</span>
<span class=normal>355</span>
<span class=normal>356</span>
<span class=normal>357</span>
<span class=normal>358</span>
<span class=normal>359</span>
<span class=normal>360</span>
<span class=normal>361</span>
<span class=normal>362</span>
<span class=normal>363</span>
<span class=normal>364</span>
<span class=normal>365</span>
<span class=normal>366</span>
<span class=normal>367</span>
<span class=normal>368</span>
<span class=normal>369</span>
<span class=normal>370</span>
<span class=normal>371</span>
<span class=normal>372</span>
<span class=normal>373</span>
<span class=normal>374</span>
<span class=normal>375</span>
<span class=normal>376</span>
<span class=normal>377</span>
<span class=normal>378</span>
<span class=normal>379</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>map_variants</span><span class=p>(</span>
    <span class=n>parsed_associations</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>variant_annotation_path</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Add variant metadata in associations.</span>

<span class=sd>    Args:</span>
<span class=sd>        parsed_associations (DataFrame): associations</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        variant_annotation_path (str): variant annotation path</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: associations with variant metadata</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>variants</span> <span class=o>=</span> <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>parquet</span><span class=p>(</span><span class=n>variant_annotation_path</span><span class=p>)</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>),</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;position&quot;</span><span class=p>,</span>
        <span class=s2>&quot;rsIdsGnomad&quot;</span><span class=p>,</span>
        <span class=s2>&quot;referenceAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;alternateAllele&quot;</span><span class=p>,</span>
        <span class=s2>&quot;alleleFrequencies&quot;</span><span class=p>,</span>
    <span class=p>)</span>

    <span class=n>mapped_associations</span> <span class=o>=</span> <span class=n>variants</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>broadcast</span><span class=p>(</span><span class=n>parsed_associations</span><span class=p>),</span> <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;chromosome&quot;</span><span class=p>,</span> <span class=s2>&quot;position&quot;</span><span class=p>],</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;right&quot;</span>
    <span class=p>)</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=n>assoc_without_variant</span> <span class=o>=</span> <span class=n>mapped_associations</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;variantId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNull</span><span class=p>()</span>
    <span class=p>)</span><span class=o>.</span><span class=n>count</span><span class=p>()</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s2>&quot;Loading variant annotation and joining with associations... </span><span class=si>{</span><span class=n>assoc_without_variant</span><span class=si>}</span><span class=s2> associations outside gnomAD&quot;</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=n>mapped_associations</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.process_associations class="doc doc-heading"> <code class="highlight language-python"><span class=n>process_associations</span><span class=p>(</span><span class=n>association_df</span><span class=p>,</span> <span class=n>etl</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Post-process associations DataFrame.</p> <ul> <li>The function does the following:<ul> <li>Adds a unique identifier to each association.</li> <li>Processes the variant related columns.</li> <li>Processes the EFO terms.</li> <li>Splits the p-value into exponent and mantissa.</li> <li>Drops some columns.</li> <li>Provides some stats on the filtered association dataset.</li> </ul> </li> </ul> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>association_df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>associations including the next columns: - associationId - snpIdCurrent - chrId - chrPos - snpIds - riskAllele - rsidGwasCatalog - efo - exponent - mantissa</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span>
<span class=normal>209</span>
<span class=normal>210</span>
<span class=normal>211</span>
<span class=normal>212</span>
<span class=normal>213</span>
<span class=normal>214</span>
<span class=normal>215</span>
<span class=normal>216</span>
<span class=normal>217</span>
<span class=normal>218</span>
<span class=normal>219</span>
<span class=normal>220</span>
<span class=normal>221</span>
<span class=normal>222</span>
<span class=normal>223</span>
<span class=normal>224</span>
<span class=normal>225</span>
<span class=normal>226</span>
<span class=normal>227</span>
<span class=normal>228</span>
<span class=normal>229</span>
<span class=normal>230</span>
<span class=normal>231</span>
<span class=normal>232</span>
<span class=normal>233</span>
<span class=normal>234</span>
<span class=normal>235</span>
<span class=normal>236</span>
<span class=normal>237</span>
<span class=normal>238</span>
<span class=normal>239</span>
<span class=normal>240</span>
<span class=normal>241</span>
<span class=normal>242</span>
<span class=normal>243</span>
<span class=normal>244</span>
<span class=normal>245</span>
<span class=normal>246</span>
<span class=normal>247</span>
<span class=normal>248</span>
<span class=normal>249</span>
<span class=normal>250</span>
<span class=normal>251</span>
<span class=normal>252</span>
<span class=normal>253</span>
<span class=normal>254</span>
<span class=normal>255</span>
<span class=normal>256</span>
<span class=normal>257</span>
<span class=normal>258</span>
<span class=normal>259</span>
<span class=normal>260</span>
<span class=normal>261</span>
<span class=normal>262</span>
<span class=normal>263</span>
<span class=normal>264</span>
<span class=normal>265</span>
<span class=normal>266</span>
<span class=normal>267</span>
<span class=normal>268</span>
<span class=normal>269</span>
<span class=normal>270</span>
<span class=normal>271</span>
<span class=normal>272</span>
<span class=normal>273</span>
<span class=normal>274</span>
<span class=normal>275</span>
<span class=normal>276</span>
<span class=normal>277</span>
<span class=normal>278</span>
<span class=normal>279</span>
<span class=normal>280</span>
<span class=normal>281</span>
<span class=normal>282</span>
<span class=normal>283</span>
<span class=normal>284</span>
<span class=normal>285</span>
<span class=normal>286</span>
<span class=normal>287</span>
<span class=normal>288</span>
<span class=normal>289</span>
<span class=normal>290</span>
<span class=normal>291</span>
<span class=normal>292</span>
<span class=normal>293</span>
<span class=normal>294</span>
<span class=normal>295</span>
<span class=normal>296</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>process_associations</span><span class=p>(</span><span class=n>association_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Post-process associations DataFrame.</span>

<span class=sd>    - The function does the following:</span>
<span class=sd>        - Adds a unique identifier to each association.</span>
<span class=sd>        - Processes the variant related columns.</span>
<span class=sd>        - Processes the EFO terms.</span>
<span class=sd>        - Splits the p-value into exponent and mantissa.</span>
<span class=sd>        - Drops some columns.</span>
<span class=sd>        - Provides some stats on the filtered association dataset.</span>

<span class=sd>    Args:</span>
<span class=sd>        association_df (DataFrame): associations</span>
<span class=sd>        etl (ETLSession): current ETL session</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: associations including the next columns:</span>
<span class=sd>            - associationId</span>
<span class=sd>            - snpIdCurrent</span>
<span class=sd>            - chrId</span>
<span class=sd>            - chrPos</span>
<span class=sd>            - snpIds</span>
<span class=sd>            - riskAllele</span>
<span class=sd>            - rsidGwasCatalog</span>
<span class=sd>            - efo</span>
<span class=sd>            - exponent</span>
<span class=sd>            - mantissa</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Processing associations:</span>
    <span class=n>parsed_associations</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>association_df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=c1># Adding association identifier for future deduplication:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>monotonically_increasing_id</span><span class=p>()</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;associationId&quot;</span><span class=p>),</span>
            <span class=c1># Processing variant related columns:</span>
            <span class=c1>#   - Sorting out current rsID field: &lt;- why do we need this? rs identifiers should always come from the GnomAD dataset.</span>
            <span class=c1>#   - Removing variants with no genomic mappings -&gt; losing ~3% of all associations</span>
            <span class=c1>#   - Multiple variants can correspond to a single association.</span>
            <span class=c1>#   - Variant identifiers are stored in the SNPS column, while the mapped coordinates are stored in the CHR_ID and CHR_POS columns.</span>
            <span class=c1>#   - All these fields are split into arrays, then they are paired with the same index eg. first ID is paired with first coordinate, and so on</span>
            <span class=c1>#   - Then the association is exploded to all variants.</span>
            <span class=c1>#   - The risk allele is extracted from the &#39;STRONGEST SNP-RISK ALLELE&#39; column.</span>
            <span class=c1># The current snp id field is just a number at the moment (stored as a string). Adding &#39;rs&#39; prefix if looks good.</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=s2>&quot;^[0-9]*$&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>format_string</span><span class=p>(</span><span class=s2>&quot;rs</span><span class=si>%s</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>)),</span>
            <span class=p>)</span>
            <span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>))</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>),</span>
            <span class=c1># Variant fields are joined together in a matching list, then extracted into a separate rows again:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>explode</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>arrays_zip</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span> <span class=s2>&quot;;&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span> <span class=s2>&quot;; &quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>),</span> <span class=s2>&quot;; &quot;</span><span class=p>),</span>
                <span class=p>)</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;VARIANT&quot;</span><span class=p>),</span>
            <span class=c1># Extracting variant fields:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;VARIANT.chromosome&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;chromosome&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;VARIANT.position&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;position&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;VARIANT.snpIds&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;VARIANT.strongestSnpRiskAllele&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=s2>&quot;*&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span> <span class=s2>&quot;-&quot;</span><span class=p>)</span>
            <span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span>
            <span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;riskAllele&quot;</span><span class=p>),</span>
            <span class=c1># Create a unique set of SNPs linked to the assocition:</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array_distinct</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>),</span> <span class=s2>&quot;-&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIdCurrent&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>),</span>
                <span class=p>)</span>
            <span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;rsIdsGwasCatalog&quot;</span><span class=p>),</span>
            <span class=c1># Processing EFO terms:</span>
            <span class=c1>#   - Multiple EFO terms can correspond to a single association.</span>
            <span class=c1>#   - EFO terms are stored as full URIS, separated by semicolons.</span>
            <span class=c1>#   - Associations are exploded to all EFO terms.</span>
            <span class=c1>#   - EFO terms in the study table is not considered as association level EFO annotation has priority (via p-value text)</span>
            <span class=c1># Process EFO URIs: -&gt; why do we explode?</span>
            <span class=n>parse_efos</span><span class=p>(</span><span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;efos&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pvalue&quot;</span><span class=p>),</span> <span class=s2>&quot;E&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;integer&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;exponent&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>split</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pvalue&quot;</span><span class=p>),</span> <span class=s2>&quot;E&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=s2>&quot;float&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;mantissa&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Cleaning up:</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;mappedTraitUri&quot;</span><span class=p>,</span> <span class=s2>&quot;strongestSnpRiskAllele&quot;</span><span class=p>,</span> <span class=s2>&quot;VARIANT&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=c1># Providing stats on the filtered association dataset:</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Number of associations: </span><span class=si>{</span><span class=n>parsed_associations</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of studies: </span><span class=si>{</span><span class=n>parsed_associations</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of variants: </span><span class=si>{</span><span class=n>parsed_associations</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>parsed_associations</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.process_associations.read_associations_data class="doc doc-heading"> <code class="highlight language-python"><span class=n>read_associations_data</span><span class=p>(</span><span class=n>etl</span><span class=p>,</span> <span class=n>gwas_association_file</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Read GWASCatalog associations.</p> <p>It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and applies some pre-defined filters on the data</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>etl</code></td> <td> <code><span title="etl.common.ETLSession.ETLSession">ETLSession</span></code> </td> <td><p>current ETL session</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>gwas_association_file</code></td> <td> <code>str</code> </td> <td><p>path to association file CSV</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>pvalue_cutoff</code></td> <td> <code>float</code> </td> <td><p>association p-value cut-off</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p><code>DataFrame</code> with the GWAS Catalog associations</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/process_associations.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>read_associations_data</span><span class=p>(</span>
    <span class=n>etl</span><span class=p>:</span> <span class=n>ETLSession</span><span class=p>,</span> <span class=n>gwas_association_file</span><span class=p>:</span> <span class=nb>str</span><span class=p>,</span> <span class=n>pvalue_cutoff</span><span class=p>:</span> <span class=nb>float</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Read GWASCatalog associations.</span>

<span class=sd>    It reads the GWAS Catalog association dataset, selects and renames columns, casts columns, and</span>
<span class=sd>    applies some pre-defined filters on the data</span>

<span class=sd>    Args:</span>
<span class=sd>        etl (ETLSession): current ETL session</span>
<span class=sd>        gwas_association_file (str): path to association file CSV</span>
<span class=sd>        pvalue_cutoff (float): association p-value cut-off</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: `DataFrame` with the GWAS Catalog associations</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=s2>&quot;Starting ingesting GWAS Catalog associations...&quot;</span><span class=p>)</span>

    <span class=c1># Reading and filtering associations:</span>
    <span class=n>association_df</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>etl</span><span class=o>.</span><span class=n>spark</span><span class=o>.</span><span class=n>read</span><span class=o>.</span><span class=n>csv</span><span class=p>(</span><span class=n>gwas_association_file</span><span class=p>,</span> <span class=n>sep</span><span class=o>=</span><span class=s2>&quot;</span><span class=se>\t</span><span class=s2>&quot;</span><span class=p>,</span> <span class=n>header</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
        <span class=c1># Select and rename columns:</span>
        <span class=o>.</span><span class=n>select</span><span class=p>(</span>
            <span class=o>*</span><span class=p>[</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>old_name</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=n>new_name</span><span class=p>)</span>
                <span class=k>for</span> <span class=n>old_name</span><span class=p>,</span> <span class=n>new_name</span> <span class=ow>in</span> <span class=n>ASSOCIATION_COLUMNS_MAP</span><span class=o>.</span><span class=n>items</span><span class=p>()</span>
            <span class=p>]</span>
        <span class=p>)</span>
        <span class=c1># Cast minus log p-value as float:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;pvalueMlog&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pvalueMlog&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>()))</span>
        <span class=c1># Apply some pre-defined filters on the data:</span>
        <span class=c1># 1. Dropping associations based on variant x variant interactions</span>
        <span class=c1># 2. Dropping sub-significant associations</span>
        <span class=c1># 3. Dropping associations without genomic location</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot; x &quot;</span><span class=p>)</span>
            <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pvalueMlog&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=o>-</span><span class=n>np</span><span class=o>.</span><span class=n>log10</span><span class=p>(</span><span class=n>pvalue_cutoff</span><span class=p>))</span>
            <span class=o>&amp;</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrPos&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span> <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;chrId&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span>
        <span class=p>)</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=c1># Providing stats on the filtered association dataset:</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;Number of associations: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of studies: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;studyAccession&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>
    <span class=n>etl</span><span class=o>.</span><span class=n>logger</span><span class=o>.</span><span class=n>info</span><span class=p>(</span>
        <span class=sa>f</span><span class=s1>&#39;Number of variants: </span><span class=si>{</span><span class=n>association_df</span><span class=o>.</span><span class=n>select</span><span class=p>(</span><span class=s2>&quot;snpIds&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>distinct</span><span class=p>()</span><span class=o>.</span><span class=n>count</span><span class=p>()</span><span class=si>}</span><span class=s1>&#39;</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>association_df</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <div class="doc doc-object doc-module"> <a id=etl.gwas_ingest.effect_harmonization></a> <div class="doc doc-contents first"> <p>Harmonisation of GWAS stats.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.effect_harmonization.get_reverse_complement class="doc doc-heading"> <code class="highlight language-python"><span class=n>get_reverse_complement</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=n>allele_col</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Get reverse complement allele of a specified allele column.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>input DataFrame</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>allele_col</code></td> <td> <code>str</code> </td> <td><p>the name of the column containing the allele</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>A dataframe with a new column called revcomp_{allele_col}</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/effect_harmonization.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>get_reverse_complement</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>allele_col</span><span class=p>:</span> <span class=nb>str</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Get reverse complement allele of a specified allele column.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): input DataFrame</span>
<span class=sd>        allele_col (str): the name of the column containing the allele</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: A dataframe with a new column called revcomp_{allele_col}</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
        <span class=sa>f</span><span class=s2>&quot;revcomp_</span><span class=si>{</span><span class=n>allele_col</span><span class=si>}</span><span class=s2>&quot;</span><span class=p>,</span>
        <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>allele_col</span><span class=p>)</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=s2>&quot;[ACTG]+&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>reverse</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>translate</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=n>allele_col</span><span class=p>),</span> <span class=s2>&quot;ACTG&quot;</span><span class=p>,</span> <span class=s2>&quot;TGAC&quot;</span><span class=p>)),</span>
        <span class=p>),</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.effect_harmonization.harmonise_beta class="doc doc-heading"> <code class="highlight language-python"><span class=n>harmonise_beta</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Harmonise betas.</p> <p>The harmonization of the beta follows the logic: - The beta is flipped (multiplied by -1) if: 1) the effect needs harmonization and 2) the annotation of the effect is annotated as decrease - The 95% confidence interval of the effect is calculated using the z-score - Irrelevant columns are dropped.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>summary stat DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>input DataFrame with harmonised beta columns: - beta - beta_ci_lower - beta_ci_upper - beta_direction</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/effect_harmonization.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 70</span>
<span class=normal> 71</span>
<span class=normal> 72</span>
<span class=normal> 73</span>
<span class=normal> 74</span>
<span class=normal> 75</span>
<span class=normal> 76</span>
<span class=normal> 77</span>
<span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>harmonise_beta</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Harmonise betas.</span>

<span class=sd>    The harmonization of the beta follows the logic:</span>
<span class=sd>    - The beta is flipped (multiplied by -1) if:</span>
<span class=sd>        1) the effect needs harmonization and</span>
<span class=sd>        2) the annotation of the effect is annotated as decrease</span>
<span class=sd>    - The 95% confidence interval of the effect is calculated using the z-score</span>
<span class=sd>    - Irrelevant columns are dropped.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): summary stat DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: input DataFrame with harmonised beta columns:</span>
<span class=sd>            - beta</span>
<span class=sd>            - beta_ci_lower</span>
<span class=sd>            - beta_ci_upper</span>
<span class=sd>            - beta_direction</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># The z-score corresponding to p-value: 0.05</span>
    <span class=n>zscore_95</span> <span class=o>=</span> <span class=mf>1.96</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;confidence_interval&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;increase&quot;</span><span class=p>)</span>
                    <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;needs_harmonization&quot;</span><span class=p>)</span>
                <span class=p>)</span>
                <span class=o>|</span> <span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;confidence_interval&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>contains</span><span class=p>(</span><span class=s2>&quot;decrease&quot;</span><span class=p>)</span>
                    <span class=o>&amp;</span> <span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;needs_harmonization&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>*</span> <span class=o>-</span><span class=mi>1</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta_conf_intervals&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>zscore_95</span><span class=p>)</span> <span class=o>*</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>zscore_95</span><span class=p>)</span> <span class=o>*</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;beta_ci_lower&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>array_min</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta_conf_intervals&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;beta_ci_upper&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>array_max</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta_conf_intervals&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta_direction&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;+&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta&quot;</span><span class=p>)</span> <span class=o>&lt;</span> <span class=mi>0</span><span class=p>,</span> <span class=s2>&quot;-&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;beta_conf_intervals&quot;</span><span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.effect_harmonization.harmonise_odds_ratio class="doc doc-heading"> <code class="highlight language-python"><span class=n>harmonise_odds_ratio</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Harmonise odds ratio.</p> <p>The harmonization of the odds ratios follows the logic: - The effect is flipped (reciprocal value is calculated) if the effect needs harmonization - The 95% confidence interval is calculated using the z-score - Irrelevant columns are dropped.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>summary stat DataFrame</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>odds ratio with harmonised OR in columns: - odds_ratio - odds_ratio_ci_lower - odds_ratio_ci_upper - odds_ratio_direction</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/effect_harmonization.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span>
<span class=normal>128</span>
<span class=normal>129</span>
<span class=normal>130</span>
<span class=normal>131</span>
<span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>harmonise_odds_ratio</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Harmonise odds ratio.</span>

<span class=sd>    The harmonization of the odds ratios follows the logic:</span>
<span class=sd>    - The effect is flipped (reciprocal value is calculated) if the effect needs harmonization</span>
<span class=sd>    - The 95% confidence interval is calculated using the z-score</span>
<span class=sd>    - Irrelevant columns are dropped.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): summary stat DataFrame</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: odds ratio with harmonised OR in columns:</span>
<span class=sd>            - odds_ratio</span>
<span class=sd>            - odds_ratio_ci_lower</span>
<span class=sd>            - odds_ratio_ci_upper</span>
<span class=sd>            - odds_ratio_direction</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># The z-score corresponding to p-value: 0.05</span>
    <span class=n>zscore_95</span> <span class=o>=</span> <span class=mf>1.96</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;needs_harmonization&quot;</span><span class=p>),</span> <span class=mi>1</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;odds_ratio_estimate&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;odds_ratio_se&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_estimate&quot;</span><span class=p>)</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio_direction&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>)</span> <span class=o>&gt;=</span> <span class=mi>1</span><span class=p>,</span> <span class=s2>&quot;+&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio&quot;</span><span class=p>)</span> <span class=o>&lt;</span> <span class=mi>1</span><span class=p>,</span> <span class=s2>&quot;-&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio_conf_intervals&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>exp</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_estimate&quot;</span><span class=p>)</span>
                    <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>zscore_95</span><span class=p>)</span> <span class=o>*</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_se&quot;</span><span class=p>)</span>
                <span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>exp</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_estimate&quot;</span><span class=p>)</span>
                    <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>zscore_95</span><span class=p>)</span> <span class=o>*</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_se&quot;</span><span class=p>)</span>
                <span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio_ci_lower&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>array_min</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_conf_intervals&quot;</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio_ci_upper&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>array_max</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_conf_intervals&quot;</span><span class=p>))</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;odds_ratio_conf_intervals&quot;</span><span class=p>,</span> <span class=s2>&quot;odds_ratio_se&quot;</span><span class=p>,</span> <span class=s2>&quot;odds_ratio_estimate&quot;</span><span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.effect_harmonization.harmonize_effect class="doc doc-heading"> <code class="highlight language-python"><span class=n>harmonize_effect</span><span class=p>(</span><span class=n>df</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Harmonisation of effects.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>df</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>GWASCatalog stats</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Harmonised GWASCatalog stats</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/effect_harmonization.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span>
<span class=normal>194</span>
<span class=normal>195</span>
<span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span>
<span class=normal>209</span>
<span class=normal>210</span>
<span class=normal>211</span>
<span class=normal>212</span>
<span class=normal>213</span>
<span class=normal>214</span>
<span class=normal>215</span>
<span class=normal>216</span>
<span class=normal>217</span>
<span class=normal>218</span>
<span class=normal>219</span>
<span class=normal>220</span>
<span class=normal>221</span>
<span class=normal>222</span>
<span class=normal>223</span>
<span class=normal>224</span>
<span class=normal>225</span>
<span class=normal>226</span>
<span class=normal>227</span>
<span class=normal>228</span>
<span class=normal>229</span>
<span class=normal>230</span>
<span class=normal>231</span>
<span class=normal>232</span>
<span class=normal>233</span>
<span class=normal>234</span>
<span class=normal>235</span>
<span class=normal>236</span>
<span class=normal>237</span>
<span class=normal>238</span>
<span class=normal>239</span>
<span class=normal>240</span>
<span class=normal>241</span>
<span class=normal>242</span>
<span class=normal>243</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>harmonize_effect</span><span class=p>(</span><span class=n>df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Harmonisation of effects.</span>

<span class=sd>    Args:</span>
<span class=sd>        df (DataFrame): GWASCatalog stats</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Harmonised GWASCatalog stats</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>df</span>
        <span class=c1># Get reverse complement of the alleles of the mapped variants:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>get_reverse_complement</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;alt&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=k>lambda</span> <span class=n>df</span><span class=p>:</span> <span class=n>get_reverse_complement</span><span class=p>(</span><span class=n>df</span><span class=p>,</span> <span class=s2>&quot;ref&quot;</span><span class=p>))</span>
        <span class=c1># A variant is palindromic if the reference and alt alleles are reverse complement of each other:</span>
        <span class=c1># eg. T -&gt; A: in such cases we cannot disambigate the effect, which means we cannot be sure if</span>
        <span class=c1># the effect is given to the alt allele on the positive strand or the ref allele on</span>
        <span class=c1># The negative strand.</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;is_palindrome&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ref&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;revcomp_alt&quot;</span><span class=p>),</span> <span class=kc>True</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># We are harmonizing the effect on the alternative allele:</span>
        <span class=c1># Adding a flag to trigger harmonization if: risk == ref or risk == revcomp(ref):</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;needs_harmonization&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;risk_allele&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;ref&quot;</span><span class=p>))</span>
                <span class=o>|</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;risk_allele&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;revcomp_ref&quot;</span><span class=p>)),</span>
                <span class=kc>True</span><span class=p>,</span>
            <span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span><span class=kc>False</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Z-score is needed to calculate 95% confidence interval:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>,</span> <span class=n>pval_to_zscore</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;p_value&quot;</span><span class=p>)))</span>
        <span class=c1># Annotation provides information if the effect is odds-ratio or beta:</span>
        <span class=c1># Effect is lost for variants with palindromic alleles.</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;beta&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;confidence_interval&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=sa>r</span><span class=s2>&quot;[increase|decrease]&quot;</span><span class=p>)</span>
                <span class=o>&amp;</span> <span class=p>(</span><span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;is_palindrome&quot;</span><span class=p>)),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;or_beta&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;odds_ratio&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span>
                <span class=p>(</span><span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;confidence_interval&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>rlike</span><span class=p>(</span><span class=sa>r</span><span class=s2>&quot;[increase|decrease]&quot;</span><span class=p>))</span>
                <span class=o>&amp;</span> <span class=p>(</span><span class=o>~</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;is_palindrome&quot;</span><span class=p>)),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;or_beta&quot;</span><span class=p>),</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=c1># Harmonize beta:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>harmonise_beta</span><span class=p>)</span>
        <span class=c1># Harmonize odds-ratio:</span>
        <span class=o>.</span><span class=n>transform</span><span class=p>(</span><span class=n>harmonise_odds_ratio</span><span class=p>)</span>
        <span class=c1># Coalesce effect direction:</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;direction&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>coalesce</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;beta_direction&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;odds_ratio_direction&quot;</span><span class=p>)),</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.gwas_ingest.effect_harmonization.pval_to_zscore class="doc doc-heading"> <code class="highlight language-python"><span class=n>pval_to_zscore</span><span class=p>(</span><span class=n>pvalcol</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Convert p-value column to z-score column.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>pvalcol</code></td> <td> <code><span title="pyspark.sql.Column">Column</span></code> </td> <td><p>pvalues to be casted to floats.</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>Column</code></td> <td> <code><span title="pyspark.sql.Column">Column</span></code> </td> <td><p>p-values transformed to z-scores</p></td> </tr> </tbody> </table> <p><strong>Examples:</strong></p> <div class=highlight><pre><span></span><code><span class=gp>&gt;&gt;&gt; </span><span class=n>d</span> <span class=o>=</span> <span class=n>d</span> <span class=o>=</span> <span class=p>[{</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=s2>&quot;t1&quot;</span><span class=p>,</span> <span class=s2>&quot;pval&quot;</span><span class=p>:</span> <span class=s2>&quot;1&quot;</span><span class=p>},</span> <span class=p>{</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=s2>&quot;t2&quot;</span><span class=p>,</span> <span class=s2>&quot;pval&quot;</span><span class=p>:</span> <span class=s2>&quot;0.9&quot;</span><span class=p>},</span> <span class=p>{</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=s2>&quot;t3&quot;</span><span class=p>,</span> <span class=s2>&quot;pval&quot;</span><span class=p>:</span> <span class=s2>&quot;0.05&quot;</span><span class=p>},</span> <span class=p>{</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=s2>&quot;t4&quot;</span><span class=p>,</span> <span class=s2>&quot;pval&quot;</span><span class=p>:</span> <span class=s2>&quot;1e-300&quot;</span><span class=p>},</span> <span class=p>{</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=s2>&quot;t5&quot;</span><span class=p>,</span> <span class=s2>&quot;pval&quot;</span><span class=p>:</span> <span class=s2>&quot;1e-1000&quot;</span><span class=p>},</span> <span class=p>{</span><span class=s2>&quot;id&quot;</span><span class=p>:</span> <span class=s2>&quot;t6&quot;</span><span class=p>,</span> <span class=s2>&quot;pval&quot;</span><span class=p>:</span> <span class=s2>&quot;NA&quot;</span><span class=p>}]</span>
<span class=gp>&gt;&gt;&gt; </span><span class=n>df</span> <span class=o>=</span> <span class=n>spark</span><span class=o>.</span><span class=n>createDataFrame</span><span class=p>(</span><span class=n>d</span><span class=p>)</span>
<span class=gp>&gt;&gt;&gt; </span><span class=n>df</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;zscore&quot;</span><span class=p>,</span> <span class=n>pval_to_zscore</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;pval&quot;</span><span class=p>)))</span><span class=o>.</span><span class=n>show</span><span class=p>()</span>
<span class=go>+---+-------+----------+</span>
<span class=go>| id|   pval|    zscore|</span>
<span class=go>+---+-------+----------+</span>
<span class=go>| t1|      1|       0.0|</span>
<span class=go>| t2|    0.9|0.12566137|</span>
<span class=go>| t3|   0.05|  1.959964|</span>
<span class=go>| t4| 1e-300| 37.537838|</span>
<span class=go>| t5|1e-1000| 37.537838|</span>
<span class=go>| t6|     NA|      null|</span>
<span class=go>+---+-------+----------+</span>
</code></pre></div> <details class=quote> <summary>Source code in <code>etl/gwas_ingest/effect_harmonization.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>15</span>
<span class=normal>16</span>
<span class=normal>17</span>
<span class=normal>18</span>
<span class=normal>19</span>
<span class=normal>20</span>
<span class=normal>21</span>
<span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>pval_to_zscore</span><span class=p>(</span><span class=n>pvalcol</span><span class=p>:</span> <span class=n>Column</span><span class=p>)</span> <span class=o>-&gt;</span> <span class=n>Column</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Convert p-value column to z-score column.</span>

<span class=sd>    Args:</span>
<span class=sd>        pvalcol (Column): pvalues to be casted to floats.</span>

<span class=sd>    Returns:</span>
<span class=sd>        Column: p-values transformed to z-scores</span>

<span class=sd>    Examples:</span>
<span class=sd>        &gt;&gt;&gt; d = d = [{&quot;id&quot;: &quot;t1&quot;, &quot;pval&quot;: &quot;1&quot;}, {&quot;id&quot;: &quot;t2&quot;, &quot;pval&quot;: &quot;0.9&quot;}, {&quot;id&quot;: &quot;t3&quot;, &quot;pval&quot;: &quot;0.05&quot;}, {&quot;id&quot;: &quot;t4&quot;, &quot;pval&quot;: &quot;1e-300&quot;}, {&quot;id&quot;: &quot;t5&quot;, &quot;pval&quot;: &quot;1e-1000&quot;}, {&quot;id&quot;: &quot;t6&quot;, &quot;pval&quot;: &quot;NA&quot;}]</span>
<span class=sd>        &gt;&gt;&gt; df = spark.createDataFrame(d)</span>
<span class=sd>        &gt;&gt;&gt; df.withColumn(&quot;zscore&quot;, pval_to_zscore(f.col(&quot;pval&quot;))).show()</span>
<span class=sd>        +---+-------+----------+</span>
<span class=sd>        | id|   pval|    zscore|</span>
<span class=sd>        +---+-------+----------+</span>
<span class=sd>        | t1|      1|       0.0|</span>
<span class=sd>        | t2|    0.9|0.12566137|</span>
<span class=sd>        | t3|   0.05|  1.959964|</span>
<span class=sd>        | t4| 1e-300| 37.537838|</span>
<span class=sd>        | t5|1e-1000| 37.537838|</span>
<span class=sd>        | t6|     NA|      null|</span>
<span class=sd>        +---+-------+----------+</span>
<span class=sd>        &lt;BLANKLINE&gt;</span>

<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>pvalue_float</span> <span class=o>=</span> <span class=n>pvalcol</span><span class=o>.</span><span class=n>cast</span><span class=p>(</span><span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>())</span>
    <span class=n>pvalue_nozero</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>when</span><span class=p>(</span><span class=n>pvalue_float</span> <span class=o>==</span> <span class=mi>0</span><span class=p>,</span> <span class=n>sys</span><span class=o>.</span><span class=n>float_info</span><span class=o>.</span><span class=n>min</span><span class=p>)</span><span class=o>.</span><span class=n>otherwise</span><span class=p>(</span>
        <span class=n>pvalue_float</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=n>f</span><span class=o>.</span><span class=n>udf</span><span class=p>(</span>
        <span class=k>lambda</span> <span class=n>pv</span><span class=p>:</span> <span class=nb>float</span><span class=p>(</span><span class=nb>abs</span><span class=p>(</span><span class=n>norm</span><span class=o>.</span><span class=n>ppf</span><span class=p>((</span><span class=nb>float</span><span class=p>(</span><span class=n>pv</span><span class=p>))</span> <span class=o>/</span> <span class=mi>2</span><span class=p>)))</span> <span class=k>if</span> <span class=n>pv</span> <span class=k>else</span> <span class=kc>None</span><span class=p>,</span>
        <span class=n>t</span><span class=o>.</span><span class=n>FloatType</span><span class=p>(),</span>
    <span class=p>)(</span><span class=n>pvalue_nozero</span><span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">October 1, 2022</span> </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../colocalisation/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Modules" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Modules </div> </div> </a> <a href=../intervals/ class="md-footer__link md-footer__link--next" aria-label="Next: Intervals" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Intervals </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.instant", "navigation.top", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../assets/javascripts/workers/search.ecf98df9.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../../assets/javascripts/bundle.2ab50757.min.js></script> </body> </html>