---
title: Data Source
---

# Data Source

Open Targets Genetics is designed to be scalable and readily extensible. New data sources will continually be curated, reviewed, incorporated, and documented here and in the release notes. Please get in touch with any suggestions.

This section contains information about the data sources used in Open Targets Genetics.

We use GnomAD v4.0 as a source for variant annotation and GnomAD v2.1.1 as a source for linkage disequilibrium (LD) information (described in the **GnomAD** section).

We rely on Open Targets as a source for the list of targets and the Gold Standard training set (described in the **Open Targets** section).

## Study Sources

1. GWAS catalog
2. FinnGen

## Molecular QTLs

1. eQTL catalogue

## Interaction / Interval-based Experiments

1. **Promoter Capture Hi-C (Javierre et al., 2016):** Evidence linking genetic variation to genes using Promoter Capture Hi-C in each of the 17 human primary hematopoietic cell types.

2. **Enhancer-TSS Correlation (Andersson et al., 2014):** Evidence linking genetic variation to genes using the correlation between the transcriptional activity of enhancers and transcription start sites, as documented in the FANTOM5 CAGE expression atlas.

3. **DHS-Promoter Correlation (Thurman et al., 2012):** Evidence linking genetic variation to genes using the correlation of DNase I hypersensitive site and gene promoters across 125 cell and tissue types from ENCODE.

4. **Promoter Capture Hi-C (Jung et al., 2019):** A compendium of promoter-centered long-range chromatin interactions in the human genome.
