<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><link rel=icon href=../../assets/favicon.png><meta name=generator content="mkdocs-1.4.2, mkdocs-material-8.5.10"><title>Modules - Genetics Portal Pipeline</title><link rel=stylesheet href=../../assets/stylesheets/main.975780f9.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.2505c338.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback"><style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style><link rel=stylesheet href=../../assets/_mkdocstrings.css><script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue> <script>var palette=__md_get("__palette");if(palette&&"object"==typeof palette.color)for(var key of Object.keys(palette.color))document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#summary-of-the-logic class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title="Genetics Portal Pipeline" class="md-header__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> Genetics Portal Pipeline </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> Modules </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme=default data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6zm0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4zM7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"/></svg> </label> <input class=md-option data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme=slate data-md-color-primary=blue-grey data-md-color-accent=light-blue aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" title=Clear aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg> </button> </nav> <div class=md-search__suggest data-md-component=search-suggest></div> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title="Genetics Portal Pipeline" class="md-nav__button md-logo" aria-label="Genetics Portal Pipeline" data-md-component=logo> <img src=../../assets/otlogo.png alt=logo> </a> Genetics Portal Pipeline </label> <div class=md-nav__source> <a href=https://github.com/opentargets/genetics_etl_python title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> opentargets/genetics_etl_python </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> Home </a> </li> <li class=md-nav__item> <a href=../../roadmap/ class=md-nav__link> Roadmap </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3 checked> <div class="md-nav__link md-nav__link--index md-nav__link--active"> <a href=./ >Modules</a> <label for=__nav_3> <span class="md-nav__icon md-icon"></span> </label> </div> <nav class=md-nav aria-label=Modules data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> Modules </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../gwas_ingest/ class=md-nav__link> Gwas ingest </a> </li> <li class=md-nav__item> <a href=../schemas/ class=md-nav__link> Schemas </a> </li> <li class=md-nav__item> <a href=../variant_to_gene/ class=md-nav__link> Variant to gene </a> </li> <li class=md-nav__item> <a href=../variants/ class=md-nav__link> Variants </a> </li> </ul> </nav> </li> </ul> </nav> </div> </div> </div> <div class="md-sidebar md-sidebar--secondary" data-md-component=sidebar data-md-type=toc> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#summary-of-the-logic class=md-nav__link> Summary of the logic </a> </li> <li class=md-nav__item> <a href=#etl.coloc.overlaps class=md-nav__link> etl.coloc.overlaps </a> </li> <li class=md-nav__item> <a href=#etl.coloc.overlaps.find_all_vs_all_overlapping_signals class=md-nav__link> find_all_vs_all_overlapping_signals() </a> </li> <li class=md-nav__item> <a href=#etl.coloc.overlaps.find_gwas_vs_all_overlapping_peaks class=md-nav__link> find_gwas_vs_all_overlapping_peaks() </a> </li> <li class=md-nav__item> <a href=#etl.coloc.coloc class=md-nav__link> etl.coloc.coloc </a> </li> <li class=md-nav__item> <a href=#etl.coloc.coloc.colocalisation class=md-nav__link> colocalisation() </a> </li> <li class=md-nav__item> <a href=#etl.coloc.coloc.ecaviar_colocalisation class=md-nav__link> ecaviar_colocalisation() </a> </li> <li class=md-nav__item> <a href=#etl.coloc.coloc.run_colocalisation class=md-nav__link> run_colocalisation() </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/opentargets/genetics_etl_python/edit/master/docs/modules/colocalisation.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg> </a> <h1>Modules</h1> <p>This workflow runs colocalization analyses that assess the degree to which independent signals of the association share the same causal variant in a region of the genome, typically limited by linkage disequilibrium (LD).</p> <p>The colocalisation test is performed using two methods:</p> <ol> <li> <p>Based on the <a href=https://github.com/chr1swallace/coloc/blob/main/R/claudia.R>R COLOC package</a>, which uses the Bayes factors from the credible set to estimate the posterior probability of colocalisation. This method makes the simplifying assumption that <strong>only a single causal variant</strong> exists for any given trait in any genomic region.</p> <p>Using GWAS summary statistics, and without information about LD, we start by enumerating all variant-level hypotheses:</p> <table> <thead> <tr> <th>Hypothesis</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td>H_0</td> <td>no association with either trait in the region</td> </tr> <tr> <td>H_1</td> <td>association with trait 1 only</td> </tr> <tr> <td>H_2</td> <td>association with trait 2 only</td> </tr> <tr> <td>H_3</td> <td>both traits are associated, but have different single causal variants</td> </tr> <tr> <td>H_4</td> <td>both traits are associated and share the same single causal variant</td> </tr> </tbody> </table> </li> <li> <p>Based on eCAVIAR. It extends the <a href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5142122/#bib18>CAVIAR</a> framework to explicitly estimate the posterior probability that the same variant is causal in 2 studies while accounting for the uncertainty of LD.</p> <p>eCAVIAR computes the colocalization posterior probability (<strong>CLPP</strong>) by utilizing the marginal posterior probabilities derived from PICS. This framework allows for <strong>multiple variants to be causal</strong> in a single locus.</p> </li> </ol> <h2 id=summary-of-the-logic>Summary of the logic</h2> <p>The workflow is divided into 2 steps for both methods:</p> <p><strong>1. Find all vs all pairs of independent signals of association in the region of interest.</strong> </p> <div class="doc doc-object doc-module"> <a id=etl.coloc.overlaps></a> <div class="doc doc-contents first"> <p>Find overlapping signals susceptible of colocalisation analysis.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.coloc.overlaps.find_all_vs_all_overlapping_signals class="doc doc-heading"> <code class="highlight language-python"><span class=n>find_all_vs_all_overlapping_signals</span><span class=p>(</span><span class=n>credible_sets_enriched</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Find overlapping signals.</p> <p>Find overlapping signals between all pairs of cred sets (exploded at the tag variant level) Any study-lead variant pair with at least one overlapping tag variant is considered.</p> <p>All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always appearing on the right side.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>credible_sets_enriched</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame containing the credible sets to be analysed</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>overlapping peaks to be compared</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/coloc/overlaps.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>13</span>
<span class=normal>14</span>
<span class=normal>15</span>
<span class=normal>16</span>
<span class=normal>17</span>
<span class=normal>18</span>
<span class=normal>19</span>
<span class=normal>20</span>
<span class=normal>21</span>
<span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>find_all_vs_all_overlapping_signals</span><span class=p>(</span>
    <span class=n>credible_sets_enriched</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Find overlapping signals.</span>

<span class=sd>    Find overlapping signals between all pairs of cred sets (exploded at the tag variant level)</span>
<span class=sd>    Any study-lead variant pair with at least one overlapping tag variant is considered.</span>

<span class=sd>    All GWAS-GWAS and all GWAS-Molecular traits are computed with the Molecular traits always</span>
<span class=sd>    appearing on the right side.</span>

<span class=sd>    Args:</span>
<span class=sd>        credible_sets_enriched (DataFrame): DataFrame containing the credible sets to be analysed</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: overlapping peaks to be compared</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=c1># Columnns to be used as left and right</span>
    <span class=n>id_cols</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;studyId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;leadVariantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;type&quot;</span><span class=p>,</span>
    <span class=p>]</span>
    <span class=n>metadata_cols</span> <span class=o>=</span> <span class=p>[</span><span class=s2>&quot;phenotype&quot;</span><span class=p>,</span> <span class=s2>&quot;biofeature&quot;</span><span class=p>,</span> <span class=s2>&quot;gene_id&quot;</span><span class=p>]</span>

    <span class=c1># Self join with complex condition. Left it&#39;s all gwas and right can be gwas or molecular trait</span>
    <span class=c1># This function includes some metadata about the overlap that needs to be dropped to avoid duplicates</span>
    <span class=n>cols_to_drop</span> <span class=o>=</span> <span class=p>[</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>,</span> <span class=s2>&quot;right_logABF&quot;</span><span class=p>,</span> <span class=s2>&quot;tagVariantId&quot;</span><span class=p>]</span>
    <span class=n>overlapping_peaks</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>find_gwas_vs_all_overlapping_peaks</span><span class=p>(</span><span class=n>credible_sets_enriched</span><span class=p>,</span> <span class=s2>&quot;logABF&quot;</span><span class=p>)</span>
        <span class=c1># Keep overlapping peaks where logABF is at either side</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span> <span class=o>&amp;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right_logABF&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>()</span>
        <span class=p>)</span><span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=o>*</span><span class=n>cols_to_drop</span><span class=p>)</span>
    <span class=p>)</span>

    <span class=c1># Bring metadata from left and right for all variants that tag the peak</span>
    <span class=n>overlapping_left</span> <span class=o>=</span> <span class=n>credible_sets_enriched</span><span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span>
        <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2> as left_</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>id_cols</span> <span class=o>+</span> <span class=n>metadata_cols</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;logABF&quot;</span><span class=p>]]</span>
        <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;tagVariantId&quot;</span><span class=p>]</span>
    <span class=p>)</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>overlapping_peaks</span><span class=o>.</span><span class=n>sortWithinPartitions</span><span class=p>([</span><span class=sa>f</span><span class=s2>&quot;left_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>]),</span>
        <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=sa>f</span><span class=s2>&quot;left_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>],</span>
        <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
    <span class=p>)</span>

    <span class=n>overlapping_right</span> <span class=o>=</span> <span class=n>credible_sets_enriched</span><span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span>
        <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2> as right_</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>id_cols</span> <span class=o>+</span> <span class=n>metadata_cols</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;logABF&quot;</span><span class=p>]]</span>
        <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;tagVariantId&quot;</span><span class=p>]</span>
    <span class=p>)</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>overlapping_peaks</span><span class=o>.</span><span class=n>sortWithinPartitions</span><span class=p>([</span><span class=sa>f</span><span class=s2>&quot;right_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>]),</span>
        <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=sa>f</span><span class=s2>&quot;right_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>],</span>
        <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=n>overlapping_left</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>overlapping_right</span><span class=p>,</span>
        <span class=n>on</span><span class=o>=</span><span class=p>[</span><span class=sa>f</span><span class=s2>&quot;right_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>]</span>
        <span class=o>+</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;left_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>]</span>
        <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;tagVariantId&quot;</span><span class=p>],</span>
        <span class=n>how</span><span class=o>=</span><span class=s2>&quot;outer&quot;</span><span class=p>,</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.coloc.overlaps.find_gwas_vs_all_overlapping_peaks class="doc doc-heading"> <code class="highlight language-python"><span class=n>find_gwas_vs_all_overlapping_peaks</span><span class=p>(</span><span class=n>credible_sets_enriched</span><span class=p>,</span> <span class=n>causality_statistic</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Find overlapping signals between GWAS (left) and GWAS or Molecular traits (right).</p> <p>The main principle is that here we extract which are the peaks that share a tagging variant for the same trait.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>credible_sets_enriched</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame containing the credible sets to be analysed</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>causality_statistic</code></td> <td> <code>str</code> </td> <td><p>Causality statistic to be used for the analysis. Can be either logABF or posteriorProbability</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>overlapping peaks to be compared</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/coloc/overlaps.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal> 78</span>
<span class=normal> 79</span>
<span class=normal> 80</span>
<span class=normal> 81</span>
<span class=normal> 82</span>
<span class=normal> 83</span>
<span class=normal> 84</span>
<span class=normal> 85</span>
<span class=normal> 86</span>
<span class=normal> 87</span>
<span class=normal> 88</span>
<span class=normal> 89</span>
<span class=normal> 90</span>
<span class=normal> 91</span>
<span class=normal> 92</span>
<span class=normal> 93</span>
<span class=normal> 94</span>
<span class=normal> 95</span>
<span class=normal> 96</span>
<span class=normal> 97</span>
<span class=normal> 98</span>
<span class=normal> 99</span>
<span class=normal>100</span>
<span class=normal>101</span>
<span class=normal>102</span>
<span class=normal>103</span>
<span class=normal>104</span>
<span class=normal>105</span>
<span class=normal>106</span>
<span class=normal>107</span>
<span class=normal>108</span>
<span class=normal>109</span>
<span class=normal>110</span>
<span class=normal>111</span>
<span class=normal>112</span>
<span class=normal>113</span>
<span class=normal>114</span>
<span class=normal>115</span>
<span class=normal>116</span>
<span class=normal>117</span>
<span class=normal>118</span>
<span class=normal>119</span>
<span class=normal>120</span>
<span class=normal>121</span>
<span class=normal>122</span>
<span class=normal>123</span>
<span class=normal>124</span>
<span class=normal>125</span>
<span class=normal>126</span>
<span class=normal>127</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>find_gwas_vs_all_overlapping_peaks</span><span class=p>(</span>
    <span class=n>credible_sets_enriched</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>causality_statistic</span><span class=p>:</span> <span class=nb>str</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Find overlapping signals between GWAS (left) and GWAS or Molecular traits (right).</span>

<span class=sd>    The main principle is that here we extract which are the peaks that share a tagging variant for the same trait.</span>

<span class=sd>    Args:</span>
<span class=sd>        credible_sets_enriched (DataFrame): DataFrame containing the credible sets to be analysed</span>
<span class=sd>        causality_statistic (str): Causality statistic to be used for the analysis. Can be either logABF or posteriorProbability</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: overlapping peaks to be compared</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>id_cols</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;studyId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;leadVariantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;type&quot;</span><span class=p>,</span>
    <span class=p>]</span>
    <span class=n>cols_to_rename</span> <span class=o>=</span> <span class=n>id_cols</span> <span class=o>+</span> <span class=p>[</span><span class=n>causality_statistic</span><span class=p>]</span>
    <span class=n>credset_to_self_join</span> <span class=o>=</span> <span class=n>credible_sets_enriched</span><span class=o>.</span><span class=n>select</span><span class=p>(</span>
        <span class=n>id_cols</span> <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;tagVariantId&quot;</span><span class=p>,</span> <span class=n>causality_statistic</span><span class=p>]</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=p>(</span>
        <span class=n>credset_to_self_join</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;left&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;type&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=s2>&quot;gwas&quot;</span><span class=p>)</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>credset_to_self_join</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;right&quot;</span><span class=p>),</span>
            <span class=n>on</span><span class=o>=</span><span class=p>[</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left.chromosome&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right.chromosome&quot;</span><span class=p>),</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left.tagVariantId&quot;</span><span class=p>)</span> <span class=o>==</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right.tagVariantId&quot;</span><span class=p>),</span>
                <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right.type&quot;</span><span class=p>)</span> <span class=o>!=</span> <span class=s2>&quot;gwas&quot;</span><span class=p>)</span>
                <span class=o>|</span> <span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left.studyId&quot;</span><span class=p>)</span> <span class=o>&gt;</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right.studyId&quot;</span><span class=p>)),</span>
            <span class=p>],</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;inner&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=c1># Rename columns to make them unambiguous</span>
        <span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span>
            <span class=o>*</span><span class=p>(</span>
                <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;left.</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2> as left_</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>cols_to_rename</span><span class=p>]</span>
                <span class=o>+</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;right.</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2> as right_</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>cols_to_rename</span><span class=p>]</span>
                <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;left.tagVariantId as tagVariantId&quot;</span><span class=p>],</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>dropDuplicates</span><span class=p>(</span>
            <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;left_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>]</span> <span class=o>+</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;right_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>id_cols</span><span class=p>]</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>cache</span><span class=p>()</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div><p><strong>2. For each pair of signals, run the colocalisation test.</strong> </p> <div class="doc doc-object doc-module"> <a id=etl.coloc.coloc></a> <div class="doc doc-contents first"> <p>Utilities to perform colocalisation analysis.</p> <div class="doc doc-children"> <div class="doc doc-object doc-function"> <h2 id=etl.coloc.coloc.colocalisation class="doc doc-heading"> <code class="highlight language-python"><span class=n>colocalisation</span><span class=p>(</span><span class=n>overlapping_signals</span><span class=p>,</span> <span class=n>priorc1</span><span class=p>,</span> <span class=n>priorc2</span><span class=p>,</span> <span class=n>priorc12</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Calculate bayesian colocalisation based on overlapping signals.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>overlapping_signals</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>overlapping peaks</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>priorc1</code></td> <td> <code>float</code> </td> <td><p>p1 prior</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>priorc2</code></td> <td> <code>float</code> </td> <td><p>p2 prior</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>priorc12</code></td> <td> <code>float</code> </td> <td><p>p12 prior</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>Colocalisation results</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/coloc/coloc.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>132</span>
<span class=normal>133</span>
<span class=normal>134</span>
<span class=normal>135</span>
<span class=normal>136</span>
<span class=normal>137</span>
<span class=normal>138</span>
<span class=normal>139</span>
<span class=normal>140</span>
<span class=normal>141</span>
<span class=normal>142</span>
<span class=normal>143</span>
<span class=normal>144</span>
<span class=normal>145</span>
<span class=normal>146</span>
<span class=normal>147</span>
<span class=normal>148</span>
<span class=normal>149</span>
<span class=normal>150</span>
<span class=normal>151</span>
<span class=normal>152</span>
<span class=normal>153</span>
<span class=normal>154</span>
<span class=normal>155</span>
<span class=normal>156</span>
<span class=normal>157</span>
<span class=normal>158</span>
<span class=normal>159</span>
<span class=normal>160</span>
<span class=normal>161</span>
<span class=normal>162</span>
<span class=normal>163</span>
<span class=normal>164</span>
<span class=normal>165</span>
<span class=normal>166</span>
<span class=normal>167</span>
<span class=normal>168</span>
<span class=normal>169</span>
<span class=normal>170</span>
<span class=normal>171</span>
<span class=normal>172</span>
<span class=normal>173</span>
<span class=normal>174</span>
<span class=normal>175</span>
<span class=normal>176</span>
<span class=normal>177</span>
<span class=normal>178</span>
<span class=normal>179</span>
<span class=normal>180</span>
<span class=normal>181</span>
<span class=normal>182</span>
<span class=normal>183</span>
<span class=normal>184</span>
<span class=normal>185</span>
<span class=normal>186</span>
<span class=normal>187</span>
<span class=normal>188</span>
<span class=normal>189</span>
<span class=normal>190</span>
<span class=normal>191</span>
<span class=normal>192</span>
<span class=normal>193</span>
<span class=normal>194</span>
<span class=normal>195</span>
<span class=normal>196</span>
<span class=normal>197</span>
<span class=normal>198</span>
<span class=normal>199</span>
<span class=normal>200</span>
<span class=normal>201</span>
<span class=normal>202</span>
<span class=normal>203</span>
<span class=normal>204</span>
<span class=normal>205</span>
<span class=normal>206</span>
<span class=normal>207</span>
<span class=normal>208</span>
<span class=normal>209</span>
<span class=normal>210</span>
<span class=normal>211</span>
<span class=normal>212</span>
<span class=normal>213</span>
<span class=normal>214</span>
<span class=normal>215</span>
<span class=normal>216</span>
<span class=normal>217</span>
<span class=normal>218</span>
<span class=normal>219</span>
<span class=normal>220</span>
<span class=normal>221</span>
<span class=normal>222</span>
<span class=normal>223</span>
<span class=normal>224</span>
<span class=normal>225</span>
<span class=normal>226</span>
<span class=normal>227</span>
<span class=normal>228</span>
<span class=normal>229</span>
<span class=normal>230</span>
<span class=normal>231</span>
<span class=normal>232</span>
<span class=normal>233</span>
<span class=normal>234</span>
<span class=normal>235</span>
<span class=normal>236</span>
<span class=normal>237</span>
<span class=normal>238</span>
<span class=normal>239</span>
<span class=normal>240</span>
<span class=normal>241</span>
<span class=normal>242</span>
<span class=normal>243</span>
<span class=normal>244</span>
<span class=normal>245</span>
<span class=normal>246</span>
<span class=normal>247</span>
<span class=normal>248</span>
<span class=normal>249</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>colocalisation</span><span class=p>(</span>
    <span class=n>overlapping_signals</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>priorc1</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span> <span class=n>priorc2</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span> <span class=n>priorc12</span><span class=p>:</span> <span class=nb>float</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Calculate bayesian colocalisation based on overlapping signals.</span>

<span class=sd>    Args:</span>
<span class=sd>        overlapping_signals (DataFrame): overlapping peaks</span>
<span class=sd>        priorc1 (float): p1 prior</span>
<span class=sd>        priorc2 (float): p2 prior</span>
<span class=sd>        priorc12 (float): p12 prior</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: Colocalisation results</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>signal_pairs_cols</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;studyId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;leadVariantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;type&quot;</span><span class=p>,</span>
    <span class=p>]</span>

    <span class=c1># register udfs</span>
    <span class=n>logsum</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>udf</span><span class=p>(</span><span class=n>_get_logsum</span><span class=p>,</span> <span class=n>DoubleType</span><span class=p>())</span>
    <span class=n>posteriors</span> <span class=o>=</span> <span class=n>f</span><span class=o>.</span><span class=n>udf</span><span class=p>(</span><span class=n>_get_posteriors</span><span class=p>,</span> <span class=n>VectorUDT</span><span class=p>())</span>
    <span class=n>coloc</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>overlapping_signals</span>
        <span class=c1># Before summing log_abf columns nulls need to be filled with 0:</span>
        <span class=o>.</span><span class=n>fillna</span><span class=p>(</span><span class=mi>0</span><span class=p>,</span> <span class=n>subset</span><span class=o>=</span><span class=p>[</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>,</span> <span class=s2>&quot;right_logABF&quot;</span><span class=p>])</span>
        <span class=c1># Sum of log_abfs for each pair of signals</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;sum_log_abf&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right_logABF&quot;</span><span class=p>))</span>
        <span class=c1># Group by overlapping peak and generating dense vectors of log_abf:</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span>
            <span class=o>*</span><span class=p>[</span><span class=s2>&quot;left_&quot;</span> <span class=o>+</span> <span class=n>col</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>signal_pairs_cols</span><span class=p>]</span>
            <span class=o>+</span> <span class=p>[</span><span class=s2>&quot;right_&quot;</span> <span class=o>+</span> <span class=n>col</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>signal_pairs_cols</span><span class=p>]</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>count</span><span class=p>(</span><span class=s2>&quot;*&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;coloc_n_vars&quot;</span><span class=p>),</span>
            <span class=n>fml</span><span class=o>.</span><span class=n>array_to_vector</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>collect_list</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>)))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;left_logABF&quot;</span>
            <span class=p>),</span>
            <span class=n>fml</span><span class=o>.</span><span class=n>array_to_vector</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>collect_list</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right_logABF&quot;</span><span class=p>)))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;right_logABF&quot;</span>
            <span class=p>),</span>
            <span class=n>fml</span><span class=o>.</span><span class=n>array_to_vector</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>collect_list</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sum_log_abf&quot;</span><span class=p>)))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span>
                <span class=s2>&quot;sum_log_abf&quot;</span>
            <span class=p>),</span>
            <span class=c1># carrying over information and renaming columns (backwards compatible)</span>
            <span class=n>f</span><span class=o>.</span><span class=n>first</span><span class=p>(</span><span class=s2>&quot;left_phenotype&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;left_phenotype&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>first</span><span class=p>(</span><span class=s2>&quot;left_biofeature&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;left_biofeature&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>first</span><span class=p>(</span><span class=s2>&quot;left_gene_id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;left_gene_id&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>first</span><span class=p>(</span><span class=s2>&quot;right_phenotype&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;right_phenotype&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>first</span><span class=p>(</span><span class=s2>&quot;right_biofeature&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;right_biofeature&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>first</span><span class=p>(</span><span class=s2>&quot;right_gene_id&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;right_gene_id&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;logsum1&quot;</span><span class=p>,</span> <span class=n>logsum</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;logsum2&quot;</span><span class=p>,</span> <span class=n>logsum</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right_logABF&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;logsum12&quot;</span><span class=p>,</span> <span class=n>logsum</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sum_log_abf&quot;</span><span class=p>)))</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;left_logABF&quot;</span><span class=p>,</span> <span class=s2>&quot;right_logABF&quot;</span><span class=p>,</span> <span class=s2>&quot;sum_log_abf&quot;</span><span class=p>)</span>
        <span class=c1># Add priors</span>
        <span class=c1># priorc1 Prior on variant being causal for trait 1</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;priorc1&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>priorc1</span><span class=p>))</span>
        <span class=c1># priorc2 Prior on variant being causal for trait 2</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;priorc2&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>priorc2</span><span class=p>))</span>
        <span class=c1># priorc12 Prior on variant being causal for traits 1 and 2</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;priorc12&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=n>priorc12</span><span class=p>))</span>
        <span class=c1># h0-h2</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;lH0abf&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>lit</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;lH1abf&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;priorc1&quot;</span><span class=p>))</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum1&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;lH2abf&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;priorc2&quot;</span><span class=p>))</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum2&quot;</span><span class=p>))</span>
        <span class=c1># h3</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;sumlogsum&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum1&quot;</span><span class=p>)</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum2&quot;</span><span class=p>))</span>
        <span class=c1># exclude null H3/H4s: due to sumlogsum == logsum12</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sumlogsum&quot;</span><span class=p>)</span> <span class=o>!=</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum12&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;max&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>greatest</span><span class=p>(</span><span class=s2>&quot;sumlogsum&quot;</span><span class=p>,</span> <span class=s2>&quot;logsum12&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;logdiff&quot;</span><span class=p>,</span>
            <span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;max&quot;</span><span class=p>)</span>
                <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>exp</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;sumlogsum&quot;</span><span class=p>)</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;max&quot;</span><span class=p>))</span>
                    <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>exp</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum12&quot;</span><span class=p>)</span> <span class=o>-</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;max&quot;</span><span class=p>))</span>
                <span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;lH3abf&quot;</span><span class=p>,</span>
            <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;priorc1&quot;</span><span class=p>))</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;priorc2&quot;</span><span class=p>))</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logdiff&quot;</span><span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;right_logsum&quot;</span><span class=p>,</span> <span class=s2>&quot;left_logsum&quot;</span><span class=p>,</span> <span class=s2>&quot;sumlogsum&quot;</span><span class=p>,</span> <span class=s2>&quot;max&quot;</span><span class=p>,</span> <span class=s2>&quot;logdiff&quot;</span><span class=p>)</span>
        <span class=c1># h4</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;lH4abf&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>log</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;priorc12&quot;</span><span class=p>))</span> <span class=o>+</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logsum12&quot;</span><span class=p>))</span>
        <span class=c1># cleaning</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;priorc1&quot;</span><span class=p>,</span> <span class=s2>&quot;priorc2&quot;</span><span class=p>,</span> <span class=s2>&quot;priorc12&quot;</span><span class=p>,</span> <span class=s2>&quot;logsum1&quot;</span><span class=p>,</span> <span class=s2>&quot;logsum2&quot;</span><span class=p>,</span> <span class=s2>&quot;logsum12&quot;</span><span class=p>)</span>
        <span class=c1># posteriors</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;allABF&quot;</span><span class=p>,</span>
            <span class=n>fml</span><span class=o>.</span><span class=n>array_to_vector</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>array</span><span class=p>(</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;lH0abf&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;lH1abf&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;lH2abf&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;lH3abf&quot;</span><span class=p>),</span>
                    <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;lH4abf&quot;</span><span class=p>),</span>
                <span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>,</span> <span class=n>fml</span><span class=o>.</span><span class=n>vector_to_array</span><span class=p>(</span><span class=n>posteriors</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;allABF&quot;</span><span class=p>))))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_h0&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>0</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_h1&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>1</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_h2&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>2</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_h3&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>3</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_h4&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>getItem</span><span class=p>(</span><span class=mi>4</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_h4_h3&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;coloc_h4&quot;</span><span class=p>)</span> <span class=o>/</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;coloc_h3&quot;</span><span class=p>))</span>
        <span class=o>.</span><span class=n>withColumn</span><span class=p>(</span><span class=s2>&quot;coloc_log2_h4_h3&quot;</span><span class=p>,</span> <span class=n>f</span><span class=o>.</span><span class=n>log2</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;coloc_h4_h3&quot;</span><span class=p>)))</span>
        <span class=c1># clean up</span>
        <span class=o>.</span><span class=n>drop</span><span class=p>(</span><span class=s2>&quot;posteriors&quot;</span><span class=p>,</span> <span class=s2>&quot;allABF&quot;</span><span class=p>,</span> <span class=s2>&quot;lH0abf&quot;</span><span class=p>,</span> <span class=s2>&quot;lH1abf&quot;</span><span class=p>,</span> <span class=s2>&quot;lH2abf&quot;</span><span class=p>,</span> <span class=s2>&quot;lH3abf&quot;</span><span class=p>,</span> <span class=s2>&quot;lH4abf&quot;</span><span class=p>)</span>
    <span class=p>)</span>
    <span class=k>return</span> <span class=n>coloc</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.coloc.coloc.ecaviar_colocalisation class="doc doc-heading"> <code class="highlight language-python"><span class=n>ecaviar_colocalisation</span><span class=p>(</span><span class=n>overlapping_signals</span><span class=p>,</span> <span class=n>clpp_threshold</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Calculate bayesian colocalisation based on overlapping signals.</p> <p><strong>Parameters:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> <th>Default</th> </tr> </thead> <tbody> <tr> <td><code>overlapping_signals</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame with overlapping signals.</p></td> <td> <em>required</em> </td> </tr> <tr> <td><code>clpp_threshold</code></td> <td> <code>float</code> </td> <td><p>Colocalization cutoff threshold as described in the paper.</p></td> <td> <em>required</em> </td> </tr> </tbody> </table> <p><strong>Returns:</strong></p> <table> <thead> <tr> <th>Name</th> <th>Type</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>DataFrame</code></td> <td> <code><span title="pyspark.sql.DataFrame">DataFrame</span></code> </td> <td><p>DataFrame with colocalisation results.</p></td> </tr> </tbody> </table> <details class=quote> <summary>Source code in <code>etl/coloc/coloc.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>252</span>
<span class=normal>253</span>
<span class=normal>254</span>
<span class=normal>255</span>
<span class=normal>256</span>
<span class=normal>257</span>
<span class=normal>258</span>
<span class=normal>259</span>
<span class=normal>260</span>
<span class=normal>261</span>
<span class=normal>262</span>
<span class=normal>263</span>
<span class=normal>264</span>
<span class=normal>265</span>
<span class=normal>266</span>
<span class=normal>267</span>
<span class=normal>268</span>
<span class=normal>269</span>
<span class=normal>270</span>
<span class=normal>271</span>
<span class=normal>272</span>
<span class=normal>273</span>
<span class=normal>274</span>
<span class=normal>275</span>
<span class=normal>276</span>
<span class=normal>277</span>
<span class=normal>278</span>
<span class=normal>279</span>
<span class=normal>280</span>
<span class=normal>281</span>
<span class=normal>282</span>
<span class=normal>283</span>
<span class=normal>284</span>
<span class=normal>285</span>
<span class=normal>286</span>
<span class=normal>287</span>
<span class=normal>288</span>
<span class=normal>289</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>ecaviar_colocalisation</span><span class=p>(</span>
    <span class=n>overlapping_signals</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span> <span class=n>clpp_threshold</span><span class=p>:</span> <span class=nb>float</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Calculate bayesian colocalisation based on overlapping signals.</span>

<span class=sd>    Args:</span>
<span class=sd>        overlapping_signals (DataFrame): DataFrame with overlapping signals.</span>
<span class=sd>        clpp_threshold (float): Colocalization cutoff threshold as described in the paper.</span>

<span class=sd>    Returns:</span>
<span class=sd>        DataFrame: DataFrame with colocalisation results.</span>
<span class=sd>    &quot;&quot;&quot;</span>
    <span class=n>signal_pairs_cols</span> <span class=o>=</span> <span class=p>[</span>
        <span class=s2>&quot;chromosome&quot;</span><span class=p>,</span>
        <span class=s2>&quot;studyId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;leadVariantId&quot;</span><span class=p>,</span>
        <span class=s2>&quot;type&quot;</span><span class=p>,</span>
    <span class=p>]</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=n>overlapping_signals</span><span class=o>.</span><span class=n>withColumn</span><span class=p>(</span>
            <span class=s2>&quot;clpp&quot;</span><span class=p>,</span>
            <span class=n>_get_clpp</span><span class=p>(</span>
                <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;left_posteriorProbability&quot;</span><span class=p>),</span> <span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;right_posteriorProbability&quot;</span><span class=p>)</span>
            <span class=p>),</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;clpp&quot;</span><span class=p>)</span> <span class=o>&gt;</span> <span class=n>clpp_threshold</span><span class=p>)</span>
        <span class=o>.</span><span class=n>groupBy</span><span class=p>(</span>
            <span class=o>*</span><span class=p>(</span>
                <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;left_</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>signal_pairs_cols</span><span class=p>]</span>
                <span class=o>+</span> <span class=p>[</span><span class=sa>f</span><span class=s2>&quot;right_</span><span class=si>{</span><span class=n>col</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>col</span> <span class=ow>in</span> <span class=n>signal_pairs_cols</span><span class=p>]</span>
            <span class=p>)</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>agg</span><span class=p>(</span>
            <span class=n>f</span><span class=o>.</span><span class=n>count</span><span class=p>(</span><span class=s2>&quot;*&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;coloc_n_vars&quot;</span><span class=p>),</span>
            <span class=n>f</span><span class=o>.</span><span class=n>sum</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;clpp&quot;</span><span class=p>))</span><span class=o>.</span><span class=n>alias</span><span class=p>(</span><span class=s2>&quot;clpp&quot;</span><span class=p>),</span>
        <span class=p>)</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> <div class="doc doc-object doc-function"> <h2 id=etl.coloc.coloc.run_colocalisation class="doc doc-heading"> <code class="highlight language-python"><span class=n>run_colocalisation</span><span class=p>(</span><span class=n>credible_sets</span><span class=p>,</span> <span class=n>study_df</span><span class=p>,</span> <span class=n>priorc1</span><span class=p>,</span> <span class=n>priorc2</span><span class=p>,</span> <span class=n>prioc12</span><span class=p>,</span> <span class=n>pp_threshold</span><span class=p>,</span> <span class=n>sumstats</span><span class=p>)</span></code> </h2> <div class="doc doc-contents "> <p>Run colocalisation analysis.</p> <details class=quote> <summary>Source code in <code>etl/coloc/coloc.py</code></summary> <div class=highlight><table class=highlighttable><tr><td class=linenos><div class=linenodiv><pre><span></span><span class=normal>22</span>
<span class=normal>23</span>
<span class=normal>24</span>
<span class=normal>25</span>
<span class=normal>26</span>
<span class=normal>27</span>
<span class=normal>28</span>
<span class=normal>29</span>
<span class=normal>30</span>
<span class=normal>31</span>
<span class=normal>32</span>
<span class=normal>33</span>
<span class=normal>34</span>
<span class=normal>35</span>
<span class=normal>36</span>
<span class=normal>37</span>
<span class=normal>38</span>
<span class=normal>39</span>
<span class=normal>40</span>
<span class=normal>41</span>
<span class=normal>42</span>
<span class=normal>43</span>
<span class=normal>44</span>
<span class=normal>45</span>
<span class=normal>46</span>
<span class=normal>47</span>
<span class=normal>48</span>
<span class=normal>49</span>
<span class=normal>50</span>
<span class=normal>51</span>
<span class=normal>52</span>
<span class=normal>53</span>
<span class=normal>54</span>
<span class=normal>55</span>
<span class=normal>56</span>
<span class=normal>57</span>
<span class=normal>58</span>
<span class=normal>59</span>
<span class=normal>60</span>
<span class=normal>61</span>
<span class=normal>62</span>
<span class=normal>63</span>
<span class=normal>64</span>
<span class=normal>65</span>
<span class=normal>66</span>
<span class=normal>67</span>
<span class=normal>68</span>
<span class=normal>69</span>
<span class=normal>70</span>
<span class=normal>71</span>
<span class=normal>72</span>
<span class=normal>73</span>
<span class=normal>74</span>
<span class=normal>75</span>
<span class=normal>76</span>
<span class=normal>77</span></pre></div></td><td class=code><div><pre><span></span><code><span class=k>def</span> <span class=nf>run_colocalisation</span><span class=p>(</span>
    <span class=n>credible_sets</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>study_df</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
    <span class=n>priorc1</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
    <span class=n>priorc2</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
    <span class=n>prioc12</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
    <span class=n>pp_threshold</span><span class=p>:</span> <span class=nb>float</span><span class=p>,</span>
    <span class=n>sumstats</span><span class=p>:</span> <span class=n>DataFrame</span><span class=p>,</span>
<span class=p>)</span> <span class=o>-&gt;</span> <span class=n>DataFrame</span><span class=p>:</span>
    <span class=sd>&quot;&quot;&quot;Run colocalisation analysis.&quot;&quot;&quot;</span>
    <span class=n>credible_sets_enriched</span> <span class=o>=</span> <span class=n>credible_sets</span><span class=o>.</span><span class=n>join</span><span class=p>(</span>
        <span class=n>f</span><span class=o>.</span><span class=n>broadcast</span><span class=p>(</span><span class=n>study_df</span><span class=p>),</span> <span class=n>on</span><span class=o>=</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span>
    <span class=p>)</span><span class=o>.</span><span class=n>persist</span><span class=p>()</span>
    <span class=c1># 1. Standard colocalisation</span>
    <span class=c1># Looking for overlapping signals</span>
    <span class=n>overlapping_signals</span> <span class=o>=</span> <span class=n>find_all_vs_all_overlapping_signals</span><span class=p>(</span>
        <span class=c1># We keep only the credible sets where probability is given as a bayes factor</span>
        <span class=n>credible_sets_enriched</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;logABF&quot;</span><span class=p>)</span><span class=o>.</span><span class=n>isNotNull</span><span class=p>())</span>
    <span class=p>)</span>
    <span class=n>fm_coloc</span> <span class=o>=</span> <span class=n>colocalisation</span><span class=p>(</span><span class=n>overlapping_signals</span><span class=p>,</span> <span class=n>priorc1</span><span class=p>,</span> <span class=n>priorc2</span><span class=p>,</span> <span class=n>prioc12</span><span class=p>)</span>

    <span class=c1># 2. LD-expanded colocalisation</span>
    <span class=c1># Looking for overlapping signals</span>
    <span class=n>overlapping_signals</span> <span class=o>=</span> <span class=n>find_gwas_vs_all_overlapping_peaks</span><span class=p>(</span>
        <span class=c1># We keep the credible sets where probability is given as a posterior probability (resulted from PICS and SuSIE for FINNGEN)</span>
        <span class=n>credible_sets_enriched</span><span class=o>.</span><span class=n>filter</span><span class=p>(</span><span class=n>f</span><span class=o>.</span><span class=n>col</span><span class=p>(</span><span class=s2>&quot;posteriorProbability&quot;</span><span class=p>)</span> <span class=o>&gt;</span> <span class=n>pp_threshold</span><span class=p>),</span>
        <span class=s2>&quot;posteriorProbability&quot;</span><span class=p>,</span>
    <span class=p>)</span>

    <span class=c1># IDEA: pass a parameter with the type of coloc: all vs all, gwas vs all</span>
    <span class=n>metadata_cols</span> <span class=o>=</span> <span class=p>[</span><span class=s2>&quot;studyId&quot;</span><span class=p>,</span> <span class=s2>&quot;phenotype&quot;</span><span class=p>,</span> <span class=s2>&quot;biofeature&quot;</span><span class=p>]</span>
    <span class=n>ecaviar_coloc</span> <span class=o>=</span> <span class=p>(</span>
        <span class=n>ecaviar_colocalisation</span><span class=p>(</span><span class=n>overlapping_signals</span><span class=p>,</span> <span class=n>pp_threshold</span><span class=p>)</span>
        <span class=c1># Add study metadata - to be deprecated</span>
        <span class=c1># the resulting table has more rows because of the studies that have multiple mapped traits</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>study_df</span><span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span><span class=o>*</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2> as left_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>metadata_cols</span><span class=p>)),</span>
            <span class=n>on</span><span class=o>=</span><span class=s2>&quot;left_studyId&quot;</span><span class=p>,</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>join</span><span class=p>(</span>
            <span class=n>study_df</span><span class=o>.</span><span class=n>selectExpr</span><span class=p>(</span><span class=o>*</span><span class=p>(</span><span class=sa>f</span><span class=s2>&quot;</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2> as right_</span><span class=si>{</span><span class=n>i</span><span class=si>}</span><span class=s2>&quot;</span> <span class=k>for</span> <span class=n>i</span> <span class=ow>in</span> <span class=n>metadata_cols</span><span class=p>)),</span>
            <span class=n>on</span><span class=o>=</span><span class=s2>&quot;right_studyId&quot;</span><span class=p>,</span>
            <span class=n>how</span><span class=o>=</span><span class=s2>&quot;left&quot;</span><span class=p>,</span>
        <span class=p>)</span>
        <span class=o>.</span><span class=n>distinct</span><span class=p>()</span>
    <span class=p>)</span>

    <span class=k>return</span> <span class=p>(</span>
        <span class=c1># 3. Join colocalisation results</span>
        <span class=n>fm_coloc</span><span class=o>.</span><span class=n>unionByName</span><span class=p>(</span><span class=n>ecaviar_coloc</span><span class=p>,</span> <span class=n>allowMissingColumns</span><span class=o>=</span><span class=kc>True</span><span class=p>)</span>
        <span class=c1># 4. Add betas from sumstats</span>
        <span class=c1># Adds backwards compatibility with production schema</span>
        <span class=c1># Note: First implementation in _add_coloc_sumstats_info hasn&#39;t been fully tested</span>
        <span class=c1># .transform(lambda df: _add_coloc_sumstats_info(df, sumstats))</span>
    <span class=p>)</span>
</code></pre></div></td></tr></table></div> </details> </div> </div> </div> </div> </div> <hr> <div class=md-source-file> <small> Last update: <span class="git-revision-date-localized-plugin git-revision-date-localized-plugin-date">January 3, 2023</span> </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../../roadmap/ class="md-footer__link md-footer__link--prev" aria-label="Previous: Roadmap" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> Roadmap </div> </div> </a> <a href=../gwas_ingest/ class="md-footer__link md-footer__link--next" aria-label="Next: Gwas ingest" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> Gwas ingest </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4Z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-copyright> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-social> <a href=https://github.com/genetics_etl_python target=_blank rel=noopener title=github.com class=md-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><!-- Font Awesome Free 6.2.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.instant", "navigation.top", "content.code.annotate", "search.suggest", "search.highlight"], "search": "../../assets/javascripts/workers/search.16e2a7d4.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version.title": "Select version"}}</script> <script src=../../assets/javascripts/bundle.5a2dcb6a.min.js></script> </body> </html>